{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This task uses GREG Bigbin as database.\n",
    "At first, we still need to start neo4j/GREG, if you don't know how to do it, please refer to [my workshop2 about graph Algorithms](https://nbviewer.jupyter.org/github/xiaowei3223/GREGgraphAlgorithms/blob/master/workshop2/GREG_Algorithms.ipynb) to learn for start neo4j/GREG."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[1. Number of nodes of each type (TF, chr1, etc)](#1)\n",
    "\n",
    "[2. Number of relationships of each type (bind, etc)](#2)\n",
    "\n",
    "[3. List of hubs for chr12 (a hub is a node with a high degree)](#3)\n",
    "\n",
    "[4. Degree distribution for chr12 (GREG-min)](#4)\n",
    "\n",
    "[5. Modules for chromosome 12](#5)\n",
    "\n",
    "[6. Annotate all modules with the genes they contain in their bins](#6)\n",
    "\n",
    "[6.1  learn how to get the genes ](#6.1)\n",
    "\n",
    "[6.2  Get genes in each modules ](#6.2)\n",
    "\n",
    "[6.3  enriched modules in some KEGG pathway or GO term.](#6.3)\n",
    "\n",
    "[ 6.3.1 Enrich GO ](#6.3.1)\n",
    "\n",
    "[ 6.3.2 Enrich KEGG](#6.3.2)\n",
    "\n",
    "[7. Get All genes names](#7)\n",
    "\n",
    "[ 8. Find the modules that contain the gene \"NANOG\" and report all the nodes and relationships inside such modules. Also, report if they are enriched in some KEGG pathway or GO term . ](#8)\n",
    "\n",
    "[ Step1： Find out those nodes about target genes](#8.1)\n",
    "\n",
    "[ Step2: To get modules which includes target genes](#8.2)\n",
    "\n",
    "[ Step3: build a function to get relationships during nodes in each community(module)](#8.3)\n",
    "\n",
    "[ step4: get relationships during nodes in each community(module)](#8.4)\n",
    "\n",
    "[9. Select the modules that include NANOG AND CTCF AND EP300 AND POLR2A AND YY1 AND cohesin (RAD21, SMC3, OR STAG1) AND mediator (MED1 or MED12).](#9)\n",
    "\n",
    "[ Step1: Get nodes in all gene names in chromosomes,TF and LncRNA.](#9.1)\n",
    "\n",
    "[ Step2: A vector about out genes.](#9.2)\n",
    "\n",
    "[ Step3: Create a function `geneMatchIndex` for find out those nodes that genes we want.](#9.3)\n",
    "\n",
    "[ Step4: Find out those nodes that genes we want.](#9.4)\n",
    "\n",
    "[ Step5: To get modules which includes target genes](#9.5)\n",
    "\n",
    "[ Step6: Build a function to judge which modules is we want.](#9.6)\n",
    "\n",
    "[ Step7: Select those mudules as we want](#9.7)\n",
    "\n",
    "[ Step8: get relationships during nodes in each community(module)](#9.8)\n",
    "\n",
    "[10  save all results](#10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we use `RNeo4j` package to get all the results of the task."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#connect neo4j with RNeo4j\n",
    "library(RNeo4j)\n",
    "\n",
    "graph = startGraph(\"http://localhost:7474/db/data/\", username=\"neo4j\", password=\"xiaowei\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windowsRegistered S3 method overwritten by 'grid':\n",
      "  method     from  \n",
      "  print.path RNeo4j\n"
     ]
    }
   ],
   "source": [
    "#---define the backgroup of plot\n",
    "library(ggplot2)\n",
    "backgroup <-   theme(axis.text.x = element_text(vjust = 0.5, hjust = 1, angle = 90 ),\n",
    "                     plot.title = element_text(face =\"bold\" ,hjust = 0.5),\n",
    "                     panel.grid.major = element_blank(), #删除网格线\n",
    "                     panel.grid.minor = element_blank(), #删除网格线\n",
    "                     panel.background=element_rect(fill='transparent',color =\"gray\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 id='1'>1. Number of nodes of each type (TF, chr1, etc)<h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Number of the allnode is 1267023."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 1 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>AllNodes</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1267023</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 1\n",
       "\\begin{tabular}{r|l}\n",
       " AllNodes\\\\\n",
       " <int>\\\\\n",
       "\\hline\n",
       "\t 1267023\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 1\n",
       "\n",
       "| AllNodes &lt;int&gt; |\n",
       "|---|\n",
       "| 1267023 |\n",
       "\n"
      ],
      "text/plain": [
       "  AllNodes\n",
       "1 1267023 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cypher(graph, query =\"MATCH (n) RETURN count(n) as AllNodes\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"MATCH (n) RETURN distinct labels(n) as label, count(labels(n)) as count\"\n",
    "\n",
    "labels <- cypher(graph,query) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 50 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>label</th><th scope=col>count</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1       </td><td>  1245</td></tr>\n",
       "\t<tr><td>chr2       </td><td>  1211</td></tr>\n",
       "\t<tr><td>chr3       </td><td>   992</td></tr>\n",
       "\t<tr><td>chr4       </td><td>   951</td></tr>\n",
       "\t<tr><td>chr5       </td><td>   908</td></tr>\n",
       "\t<tr><td>chr6       </td><td>   854</td></tr>\n",
       "\t<tr><td>chr7       </td><td>   797</td></tr>\n",
       "\t<tr><td>chr8       </td><td>   726</td></tr>\n",
       "\t<tr><td>chr9       </td><td>   692</td></tr>\n",
       "\t<tr><td>chr10      </td><td>   669</td></tr>\n",
       "\t<tr><td>chr11      </td><td>   676</td></tr>\n",
       "\t<tr><td>chr12      </td><td>   667</td></tr>\n",
       "\t<tr><td>chr13      </td><td>   572</td></tr>\n",
       "\t<tr><td>chr14      </td><td>   535</td></tr>\n",
       "\t<tr><td>chr15      </td><td>   510</td></tr>\n",
       "\t<tr><td>chr16      </td><td>   452</td></tr>\n",
       "\t<tr><td>chr17      </td><td>   417</td></tr>\n",
       "\t<tr><td>chr18      </td><td>   402</td></tr>\n",
       "\t<tr><td>chr19      </td><td>   293</td></tr>\n",
       "\t<tr><td>chr20      </td><td>   322</td></tr>\n",
       "\t<tr><td>chr21      </td><td>   234</td></tr>\n",
       "\t<tr><td>chr22      </td><td>   255</td></tr>\n",
       "\t<tr><td>chrX       </td><td>   781</td></tr>\n",
       "\t<tr><td>chrY       </td><td>   287</td></tr>\n",
       "\t<tr><td>TF         </td><td>   261</td></tr>\n",
       "\t<tr><td>chr1_Range </td><td>127334</td></tr>\n",
       "\t<tr><td>chr2_Range </td><td>106646</td></tr>\n",
       "\t<tr><td>chr3_Range </td><td> 77198</td></tr>\n",
       "\t<tr><td>chr4_Range </td><td> 47295</td></tr>\n",
       "\t<tr><td>chr5_Range </td><td> 66077</td></tr>\n",
       "\t<tr><td>chr6_Range </td><td> 76812</td></tr>\n",
       "\t<tr><td>chr7_Range </td><td> 58335</td></tr>\n",
       "\t<tr><td>chr8_Range </td><td> 56539</td></tr>\n",
       "\t<tr><td>chr9_Range </td><td> 54826</td></tr>\n",
       "\t<tr><td>chr10_Range</td><td> 61085</td></tr>\n",
       "\t<tr><td>chr11_Range</td><td> 66916</td></tr>\n",
       "\t<tr><td>chr12_Range</td><td> 62708</td></tr>\n",
       "\t<tr><td>chr13_Range</td><td> 26761</td></tr>\n",
       "\t<tr><td>chr14_Range</td><td> 40299</td></tr>\n",
       "\t<tr><td>chr15_Range</td><td> 47826</td></tr>\n",
       "\t<tr><td>chr16_Range</td><td> 45602</td></tr>\n",
       "\t<tr><td>chr17_Range</td><td> 64562</td></tr>\n",
       "\t<tr><td>chr18_Range</td><td> 25941</td></tr>\n",
       "\t<tr><td>chr19_Range</td><td> 37520</td></tr>\n",
       "\t<tr><td>chr20_Range</td><td> 37620</td></tr>\n",
       "\t<tr><td>chr21_Range</td><td> 18740</td></tr>\n",
       "\t<tr><td>chr22_Range</td><td> 26476</td></tr>\n",
       "\t<tr><td>chrX_Range </td><td> 17992</td></tr>\n",
       "\t<tr><td>chrY_Range </td><td>   169</td></tr>\n",
       "\t<tr><td>LncRNA     </td><td>    35</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 50 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       " label & count\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t chr1        &   1245\\\\\n",
       "\t chr2        &   1211\\\\\n",
       "\t chr3        &    992\\\\\n",
       "\t chr4        &    951\\\\\n",
       "\t chr5        &    908\\\\\n",
       "\t chr6        &    854\\\\\n",
       "\t chr7        &    797\\\\\n",
       "\t chr8        &    726\\\\\n",
       "\t chr9        &    692\\\\\n",
       "\t chr10       &    669\\\\\n",
       "\t chr11       &    676\\\\\n",
       "\t chr12       &    667\\\\\n",
       "\t chr13       &    572\\\\\n",
       "\t chr14       &    535\\\\\n",
       "\t chr15       &    510\\\\\n",
       "\t chr16       &    452\\\\\n",
       "\t chr17       &    417\\\\\n",
       "\t chr18       &    402\\\\\n",
       "\t chr19       &    293\\\\\n",
       "\t chr20       &    322\\\\\n",
       "\t chr21       &    234\\\\\n",
       "\t chr22       &    255\\\\\n",
       "\t chrX        &    781\\\\\n",
       "\t chrY        &    287\\\\\n",
       "\t TF          &    261\\\\\n",
       "\t chr1\\_Range  & 127334\\\\\n",
       "\t chr2\\_Range  & 106646\\\\\n",
       "\t chr3\\_Range  &  77198\\\\\n",
       "\t chr4\\_Range  &  47295\\\\\n",
       "\t chr5\\_Range  &  66077\\\\\n",
       "\t chr6\\_Range  &  76812\\\\\n",
       "\t chr7\\_Range  &  58335\\\\\n",
       "\t chr8\\_Range  &  56539\\\\\n",
       "\t chr9\\_Range  &  54826\\\\\n",
       "\t chr10\\_Range &  61085\\\\\n",
       "\t chr11\\_Range &  66916\\\\\n",
       "\t chr12\\_Range &  62708\\\\\n",
       "\t chr13\\_Range &  26761\\\\\n",
       "\t chr14\\_Range &  40299\\\\\n",
       "\t chr15\\_Range &  47826\\\\\n",
       "\t chr16\\_Range &  45602\\\\\n",
       "\t chr17\\_Range &  64562\\\\\n",
       "\t chr18\\_Range &  25941\\\\\n",
       "\t chr19\\_Range &  37520\\\\\n",
       "\t chr20\\_Range &  37620\\\\\n",
       "\t chr21\\_Range &  18740\\\\\n",
       "\t chr22\\_Range &  26476\\\\\n",
       "\t chrX\\_Range  &  17992\\\\\n",
       "\t chrY\\_Range  &    169\\\\\n",
       "\t LncRNA      &     35\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 50 × 2\n",
       "\n",
       "| label &lt;chr&gt; | count &lt;int&gt; |\n",
       "|---|---|\n",
       "| chr1        |   1245 |\n",
       "| chr2        |   1211 |\n",
       "| chr3        |    992 |\n",
       "| chr4        |    951 |\n",
       "| chr5        |    908 |\n",
       "| chr6        |    854 |\n",
       "| chr7        |    797 |\n",
       "| chr8        |    726 |\n",
       "| chr9        |    692 |\n",
       "| chr10       |    669 |\n",
       "| chr11       |    676 |\n",
       "| chr12       |    667 |\n",
       "| chr13       |    572 |\n",
       "| chr14       |    535 |\n",
       "| chr15       |    510 |\n",
       "| chr16       |    452 |\n",
       "| chr17       |    417 |\n",
       "| chr18       |    402 |\n",
       "| chr19       |    293 |\n",
       "| chr20       |    322 |\n",
       "| chr21       |    234 |\n",
       "| chr22       |    255 |\n",
       "| chrX        |    781 |\n",
       "| chrY        |    287 |\n",
       "| TF          |    261 |\n",
       "| chr1_Range  | 127334 |\n",
       "| chr2_Range  | 106646 |\n",
       "| chr3_Range  |  77198 |\n",
       "| chr4_Range  |  47295 |\n",
       "| chr5_Range  |  66077 |\n",
       "| chr6_Range  |  76812 |\n",
       "| chr7_Range  |  58335 |\n",
       "| chr8_Range  |  56539 |\n",
       "| chr9_Range  |  54826 |\n",
       "| chr10_Range |  61085 |\n",
       "| chr11_Range |  66916 |\n",
       "| chr12_Range |  62708 |\n",
       "| chr13_Range |  26761 |\n",
       "| chr14_Range |  40299 |\n",
       "| chr15_Range |  47826 |\n",
       "| chr16_Range |  45602 |\n",
       "| chr17_Range |  64562 |\n",
       "| chr18_Range |  25941 |\n",
       "| chr19_Range |  37520 |\n",
       "| chr20_Range |  37620 |\n",
       "| chr21_Range |  18740 |\n",
       "| chr22_Range |  26476 |\n",
       "| chrX_Range  |  17992 |\n",
       "| chrY_Range  |    169 |\n",
       "| LncRNA      |     35 |\n",
       "\n"
      ],
      "text/plain": [
       "   label       count \n",
       "1  chr1          1245\n",
       "2  chr2          1211\n",
       "3  chr3           992\n",
       "4  chr4           951\n",
       "5  chr5           908\n",
       "6  chr6           854\n",
       "7  chr7           797\n",
       "8  chr8           726\n",
       "9  chr9           692\n",
       "10 chr10          669\n",
       "11 chr11          676\n",
       "12 chr12          667\n",
       "13 chr13          572\n",
       "14 chr14          535\n",
       "15 chr15          510\n",
       "16 chr16          452\n",
       "17 chr17          417\n",
       "18 chr18          402\n",
       "19 chr19          293\n",
       "20 chr20          322\n",
       "21 chr21          234\n",
       "22 chr22          255\n",
       "23 chrX           781\n",
       "24 chrY           287\n",
       "25 TF             261\n",
       "26 chr1_Range  127334\n",
       "27 chr2_Range  106646\n",
       "28 chr3_Range   77198\n",
       "29 chr4_Range   47295\n",
       "30 chr5_Range   66077\n",
       "31 chr6_Range   76812\n",
       "32 chr7_Range   58335\n",
       "33 chr8_Range   56539\n",
       "34 chr9_Range   54826\n",
       "35 chr10_Range  61085\n",
       "36 chr11_Range  66916\n",
       "37 chr12_Range  62708\n",
       "38 chr13_Range  26761\n",
       "39 chr14_Range  40299\n",
       "40 chr15_Range  47826\n",
       "41 chr16_Range  45602\n",
       "42 chr17_Range  64562\n",
       "43 chr18_Range  25941\n",
       "44 chr19_Range  37520\n",
       "45 chr20_Range  37620\n",
       "46 chr21_Range  18740\n",
       "47 chr22_Range  26476\n",
       "48 chrX_Range   17992\n",
       "49 chrY_Range     169\n",
       "50 LncRNA          35"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAOVBMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6et2OaysrK9vb2+vr7Hx8fQ0NDZ2dnh4eHp6enw8PD///9TaHlrAAAACXBI\nWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO2d6WKrrBpG3WY67deM93+xJ+KEA0P0JYCu9aNp\n80RAZEXEtC1eALCaInYDALYAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCDASKRLURTn\nOC0ByJiRSEXFM05TAvDemWv7jd/LV1Z4Kfs3Ip/Kl9XYbHW2lPB3egenvxWVCGJqgLFh0Vv8\nOcMGX5VIf5HaIs97Z8r2G7+Xr6vvXBQDkZyVrxDpVhbGEu5lUVNeF1ciyP5EOhbFe3J3iNQW\neaqx9Nt84/fy1fU9Pqp8hUjNpnMl3IueewLDcnciPSqJDqrzt0E3U/2eSJ9VvqJGi0jvQ3h4\nn4qef+8T0zGBYbk7kX6K4qf6conVGmnUWL682iPTHp/uHf33UBx+X4/3jOx4a55/X+Ucmrnt\n9X2CPl6bDR6Hakzez++xebg8tDrUU+dbW1s/BEaV669883coyksb9TVNangn1RbP5pW3So3q\n+7auStbL25xz/+537SYV98PPY/SSblf05sz0xbC5g1a5NtT2puuJ/odH1ZTi5Orvma5ImoFI\nZfUO+myn9hvgfUAO2uxmLNJRDcZTfUFx055S7ySXeqz+dOWcqmHcTZgamldVL5uKpFeuv7L9\n4VhHWk2TGv7q1rwf1NLFpbqEHYlUjtp0blc5+ob0L2l3ZdCcmb4Y5INWeW74M2hA9/2jLepq\n7++ZrkgaXaSr6uDXaXQcMqY6XmqninmRnq/famDdq4d6DlTe1ZX6TV1ovPNjO/re3z+aae+f\nenHNrdvk/hq99Y4qH7yy/6EY1jSpoXlfqxbhqlUM9V7Xtr+ppbypUdjdtihHC6+Dl7S7MmjO\nTF8Mcr1Vrg31vekb0H1/VGtZ7v6e6Yqk0UVq3skan7ZAdUAO1V7Ni9RML27aU00HnFVv1If3\n0q9kT98ez/WCwu9wEjdX+eCVfVcXk5pGNRzVAkZz/nn0ZvYiVSU9i8GpcNwL/UvaXRk0Z6Yv\nxrl+vnNt2O2NoUE+/T3XFSmj7WE3pxu/o+VLdUDu1RXDvEiv1/Ch+b7uiEMb1G+dqkeq2f1h\ncHeg6atnfVkyEUmrfPDKsqupGNY0reGvGrhXNSG7VsP3Or9qZxdpsNvPScNn+mKQ661ybajv\nzXyDnrefo7O/57oiZbQ9/C16fsxb5IQ6IO9Z0e9HInU/NPRBM22/PIc1dI8TkRyVT2ua1vCs\nhtb77fpevW8f+6LMIs1M7Qw1+/aF3iqPDbu9GTWg7sPjpFvn+3vaFSmj7eFBE2kjt5KK5hKk\nFBLpdT/XT90GNXSPU5Hslc8MvUkNlUSPRqJHfSXkEGl8kbteJL1VK0WqZrPnv4dPf0+7ImH6\nPdTv4mXSeCf1Abm0h9VjDAymM6Nyaq7V8m33RuOY2mmVj+dKzQ/FsKZpDeqatVLjTz3qF3R6\n07Sq/6bL35Pd/mxqp7fKc2o3PQw1B21Zxt3fo65ImL7xl/7ycDO3kuoD8tREqg7czSJSdR19\nVbt/UYN2MF6GxSociw1a5YNXXuqTxp+K9JqmNTTv1U1Jpd5Y/ZX6FmV9S+f5W2qrhiORxmsG\n474Y7pjWKteGjr2pv3/49/eo71Nl0OB2NvrMpPFOmv24NGP5pI7YrbSIVN7ald16OfbcX92/\n1Ara33BF9mpd/tYrH7zyPl3+rmua1lB/gu9c116P6vaUodUyqLq7A1O0K379S9ofrqNV7HFf\nDHK9Va4N9b3pe6JBnXh+X/duscHS33NdkS5d7w8WvY8b+eSqNo+pvmkG2GDJaCSSdoPwp+i+\nb8t5NDc2y/5u++A265xIbeWWG7JaTTM1qMuK9vx170pu9JoTqX6vqDjoeg9EGt9XHffFIB+0\nyrWhtjd9T3QiNcdAzfDs/T3XFenS9f5Rfwu5ZvI24KIdNc3CmXonPP5aL7C1j6zc3meww/Bj\np+qjNuX0I0KnmV+YGFc+eOXwI0J9TXM1aIvlB63k03Te1vP73tPyPGzVUCS9OXM+DJo7aJVr\nQ21vup7oRFKrdserWjZx9PdcVyTLNqZwAJFBJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACR\nAARAJAABEAlAAEQCEACRAASQEuk/gB0iL5JQOQAZgUgAAiASgACIBCAAIgEIgEgAAiASgACI\nBCAAIgEIgEgAAiASgACIBCBAIJH+Z0SoFoCkQCQAARAJQABEAhAAkQAE+Eyksv76Zu5RKxGR\nYF98JFLjTfNl/KiXiEiwLz4RqXzJiIRlsDmWTO2a7xAJoEVWpH9vEAl2yEKRyhdnJIAeRAIQ\nYJlIs4sOeomIBPtikUhl/xWRAF4Lb8hqPyASwGuRSGXZfJRh6ScbEAk2R4zP2iESbA5EAhAA\nkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJ\nQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAE\nQCQAARAJQABEAhAAkQAEQCQAAZITCcsgRxAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQA\nAfISCcsgURAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAATYkEpZBPBAJQABEAhAAkQAE\nQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhBgLyJZQ4C1IJLQbsO+QSSh3YZ9g0hCuw37BpGE\ndhv2DSIJ7TbsG0QS2m3YN0tEKt/MPWolJueKNQRYywKRyubL+FEvMTlXrCHAWhBpdRcCIBIi\ngQiyIv17g0iwQ1YsNmznjGQNATxgaucKATxAJFcI4AEiuULIjSiHE5FcIeRGlMPJJxtcIeRG\nlMPJZ+1cIeRGlMOJSK4QciPK4UQkVwi5EeVwIpIrhNyIcjgRyRVCbkQ5nIjkCiE3ohxORHKF\nkBtRDiciuULIjSiHE5FcIeRGlMOJSKtCSJAoR8woUtH8rH/6x7/EGIM6RggJEuWIzYtUFhpL\nSowxqGOEkCBRjti8SL+aR79LSowxqGOEkCBRjphzaresxBiDOkYICRLliLHYsCqEBIlyxMwi\nXUqukRApR6IcMaNIFxYb1oYQhygHxShS+fEqg15ijHGbXAhxiHJQWGwIF0IcohwUo0in4rmi\nxBjjNrnQhnlDFFxLlK41ivQoj4/lJQYamnmFNswbItJaonStZWrHYsO60JwtD5ce5l0RpfsQ\nKVhozhApKFG6jxuywUJzhkhBidJ9iBQsNGeIFJQo3cfULlhozhApKFG6D5GCheYMkYISpfsc\nU7vH8WdRidZ92UlozhApKFG6z3WN9Cw+NQmRmtCcIVJQonSfc7GBqd3S0JwhUlCidJ9LpL+C\nv9mwMDRniBSUKN3nXmy4LCnRui87Cc0ZIgUlSve5RCo/9QiR2tCcIVJQonQfN2SDheYMkYIS\npfsQKVhozhApKFG6zyzS83IoisPl499KQqQmNGeIFJQo3Wf5faTmIunT30pCpCY0Z4gUlCjd\nZxTpXFS/2Pc4FuclJVr3ZSehOUOkoETpPuffbOCG7NLQnCFSUKJ0HyIFC80ZIgUlSvcxtQsW\nmjNECkqU7mOxIVhozlaJZA3hlZpILH+vDc0ZIgUlSg9xQzZYaM4QKShRegiRgoXmDJGCEqWH\nzCKd1BPFgWukhaE5Q6SgROkh23+jUM+yarc0NGeIFJQoPWT5bxS36uHOfaSloTlDpKBE6SFu\nyAYLzRkiBSVKD1n+G8X5Wa2BF8clJS4fCtsJzVk4kazhXojSCe4bsvclJQYaClmF5gyRghKl\nE5w3ZD/+3y6I1ITmDJGCEqUTuI8ULDRniBSUKJ2ASMFCc4ZIQYnSCYgULDRnkUSyhhsiyn4i\nUrDQnCFSUKLsJyIFC80ZIgUlyn4iUrDQnCFSUKLsJyIFC80ZIgUlyn4iUrDQnCFSUKLsJyIF\nC80ZIgUlyn4iUrDQnCFSUKLsJyIFC80ZIgUlyn4iUrDQnCFSUKLsJyIFC80ZIgUlyn4iUrDQ\nnCFSUKLsJyIFC81ZiiJZw7yIsiuIFCw0Z4gUlCi7gkjBQnOGSEGJsiuIFCw0Z4gUlCi7gkjB\nQnOGSEGJsiuIFCw0Z4gUlCi7gkjBQnOGSEGJsisLRSrVlzf6o1ZijKOdWmjOshPJGiZHlNYu\nE0mJU0vUP+olxjigqYXmDJGCEqW1i0QqX4jkDs0ZIgUlSmuXiNTIg0iIhEgtsiL9e4NITWjO\ntiWSNYxBlAYtEKl8cUbyCc0ZIgUlSoM+F6nzBpEQCZFaFohUg0iI5BHGIEqDVtxHQiREQqQW\nRAoWmjNECkqUBvHJhmChOUOkoERpEJ+1CxaaM0QKSpQGIVKw0JwhUlCiNAiRgoXmDJGCEqVB\niBQsNGeIFJQoDUKkYKE5Q6SgRGkQIgULzRkiBSVKgxApWGjOECkoURqESMFCc4ZIQYnSIEQK\nFpozRApKlAYhUrDQnCFSzfItYxRrB5GCheYMkWqWbxmjWDuIFCw0Z4hUs3zLGMXaQaRgoTlD\npJrlW8Yo1g4iBQvNGSLVLN8yRrF2EClYaM4QqWb5ljGKtYNIwUJztiORlofLCVSsHUQKFpoz\nRPIIlxOoWDuIFCw0Z4jkES4nRp2IhEiJhsuJUSciIVKi4XJi1IlIiJRouJwYdSISIiUaLidG\nnYiESImGy4nSIEQKFpozRPIIrSTXIERCpERDK8k1CJEQKdHQSnINQiRESjS0klyDEAmREg2t\nJNcgREIkQr/QY9grEEk4NGeIlGPoMewViCQcmjNEyjH0GPYKRBIOzRki5Rh6DHsFIgmH5gyR\ncgw9hr0CkYRDc4ZIOYYew16BSMKhOUOkHEOPYa9AJOHQnCFSjqHHsFcgknBozhApx9Bj2CsQ\nSTg0Z4iUY+gx7BWIJByaM0TKMfQY9gpEEg7NGSLlGHoMewUiCYfmDJFyDD2GvQKRhENzhkg5\nhh7DXoFIwqE5Q6QcQ49hr0Ak4dCcIVKOocewVyCScGjOECnH0GPYKxBJODRniJRj6DHsFYgk\nHJozRMox9Bj2CkQSDs0ZIuUYegx7BSIJh+YMkXIMPYa9ApGEQ3OGSDmGHsNegUjCoTlDpBxD\nj2GvQCTh0JwhUo6hx7BXIJJwaM4QKcfQY9grEEk4NGeIlGPoMewViCQcmjNEyjH0GPYKRBIO\nzRki5Rh6DHsFIgmH5gyRcgw9hr0CkYRDc4ZIOYYew16BSMKhOUOkHEOPYa9AJOHQnCFSjqHH\nsFcgknBozhApx9Bj2CsQSTg0Z4iUY+gx7BWIJByaM0TKMfQY9gpEEg7NGSLlGHoMewUiCYfm\nDJFyDD2GvQKRhENzhkg5hh7DXoFIwqE5Q6QcQ49hr0Ak4dCcIVKOocewVyCScGjOECnH0GPY\nKxBJODRniLS5EJEQiVAgRCREIhQIEQmRCAVCEZHKN4g0Cc0ZIm0ulBCp7L4gkh6aM0TaXIhI\niEQoECISIhEKhMIi/Xvzn/XFAJsk0BkJYF8gEoAAiAQgACIBCIBIAAIE+mQDwL4I9Fk7gH2B\nSAACIBKAAIgEIAAiAQiASAACBBBpwr/pU4SRqySUDsVFmvKP0B6m1h7CNSEiRQtTaw/hmhCR\nooWptYdwTRhOJIAdgUgAAiASgACIBCAAIgEIEEOk7jeZSlM4/F2nyUvMzxo3rJ4uDamtLkeD\nypaPQ2eV6ptP2/MybrapDrL2UDnzne0pz1Cre+7ZCCI1B/Rl3tHRb9/2YWnp+q5Aazhfp+VQ\n2hvkDE112qCDXFh7qHtuXl5nsfaaTT0RRiTnAa0fPu1586F+OYeCKzR1kMdQsA15Q5100Mve\nQSt6qFXM1F6f9wTTC0pjHOiM5PRePX48Tmy7um6cvF6G2UmocTJ9apLtuoOW95A1sJZsF6k0\nngIrQk3tfPbD8l5jLKE0vQup/bQPBVM/2A5ZjHFCB7lqtfaQa+P2bGfYyFHit0Wy0e3D7EEr\nh6+ZvsDxnmG/rp0v0Yb7WtpVp7382Q0tLaODXo4eem3pjBSQBf2+L3bfQYuvkWzXZc22XxfJ\nPR3YQQgxsJ6tVh6vL6/avRIc1IiUHLa7QctD66WXwLH89n2k5EY1JiVGfUliW4lYFBqfdYZr\nDmW4M5JtsrmL0Dbjts7GdxN2D6Lha1sigfW4JHfujBD21zKWOdjn4fI22beyvScgUlAwyRp2\nkfV+2cfhK8gcYqbyAQGvkWwt2ksINnzuly0Il7fH5zWGF7FqFzKErPA4lsaXIFKsMLX2bClc\nOIdwimSZeCBSrDC19hC6RLLO37mPFC1MrT0bCoNMxu3bxfh9pJ2EWv5pRrgynH3WJ7RsZj/W\nLH/DBvn++hAiwcZYNYdYTOBrJFNrdxO6Jyefb0joDEPQ1fbdqV05eNh3+HFGuC4MQfeGOV8z\nIn0h/DgjXBeumEO4ioz1yYbkBjUi7SdccMQcZZrnkogUMsSkOGE4kYxx6A+t7ji0LRBZF48I\nV4WvjYkEEAtEApBg2RzCXmT3ZRZECoqt95evLBG6wgA4ppOIFBLbLCK5lZENhRFApJAgEiKB\nAIi0LZEs00lECgkibUokW52IFBTbAtHylSVCVxgGRAIQAJEABEAkAAks00lEAhAAkQAEQCQA\nX/iIEMB6bIvtiATgCSIBCIBIABJYTEIkAE/4fSSAwCASwGdwRgIQAJEABEAkAAEQCUAARAIQ\nAJEAFsPftQMIDyIBCIBIAAIgEoAAiAQgACLlRlGYfzI/B4Ghz3MDkZKEPs8NREoS+jw3lCa3\nU1GUl/qnU3F8VMHzXBTn5wuRokCf50alybVQXKqf3koVZeVPWT11eCFSFOjz3Kg0ORR/r9e9\n+q4ojs/XsVLqp/pyKX4RKQr0eW7UmjyuP8dapPe07lGdiA7q+eKESFGgz3NDaXKs53atNPV3\nw+fgq9DnuVFpci4Ov9cHIiUEfZ4bnSrPuald+wr4MvR5btT63F7P5hrpWH33U60zXF6vv/eP\niBQD+jw3Kk0u2jSuXf5+quXv4o5IUaDPc0Npcn6fiW7NJO9UnNQN2Yd68oVIUaDPAQRAJAAB\nEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACR\nAARAJAABEAlAAEQCEACRAARAJAABEAlAAE2k9o9HF8ffeO0ByJI5kdR/sMqFd2uv7Td+L19Z\n4aUsivMHlS+rsdnqbCnhr/obq6e/FZUI0o6cw+W5tpyPno+/5w3zIhV/8Vr0IdUf7G2/8Xv5\nuvrOVfdoIjkrXyHSrSyMJdzL5lCV18WVCNIPnXKVSRsRST0+L/V/UMyD6tj9Nt/4vXx1fY+P\nKl8hUlEYRbpr73r3BIaT1HRmSyKl0zgf1LGr/wXxt0T6rPIVNVpEOrzf7N6noudfWdT/gCK+\nSOrxrz1HryzH9/n4e95gEGlVZ3yV7k2wbn67E907+u+hOPzqf2C+qK5yDs3c9Xp8P39tNngc\nqjF5P5fVVP+h1aGeOt9e3RuvoXL9lW/+DtV/Hm+ivqZJDcfqv7So/3ekXnmr1Ki+b+uqZL28\nzTnfuxZdu0nD/fDzGL2k2xW9OTN9MWzuoFWuDbW96Xpi8M2jak1xarv879htauqVUTkzhZiO\n20yXfp8Zkar+yme14d3qgza7GYtU/4/IU31BcdOeUnvY/HuUn66cUzWMuwlTQ/tPVH7mRNIr\n11/Z/nCsI62mSQ1/dWveD2rp4lJdoo5EKkdtOrerHH1D+pe0uzJozkxfDPJBqzw3/Bk0QD3+\n1oY/2tKuWgHaptNeGZXzMhUyc9xmuvT7zC82rLtg/CpVPxf1vyCeFelZHdvicK8e6jlQeVdX\n6jd1oaH+J3gz+t7fP161GH/qxTW3bpP7a3ikx5UPXtn/UAxrmtTwrGcA1SJctYpRVrPFtv1N\nLeVNjZ52kaN+zbAX+pe0uzJozkxfDHK9Va4N9b3pG9BSXTQe1XJV3+XX+j/RqE3ne6Uvp/t+\nXIjpuM106feZF+mUlUhVP14NIjXTgpv2VPVWflUj7twelksXvKZva+d6bPwOJ3FzlQ9eee5q\nKiY1jWo4qgUM1fPqffj0GotUlfQsBqfCcS/0L2l3ZdCcmb4Y5/r5zrVhtzd9Axouw2aN2tb0\nyp+lnLkjbD1uc136dQzL3/mckqqOvFfziXmRXq/hQ/N9fRI4tEH9lqf2uZqVHwar/827/7Oe\ntExE0iofvLLsaiqGNU1r+KsG7lVNyK7V8L3Otl+vekakwW4/Jw2f6YtBrrfKtaG+N30DFMfu\nMuV5+znqe9Ft+pztlfkdGxcye9zmuvTrzC023I8ZXSSpVr9nRb8fidT90NAHzXRbu684HMcT\nkRyVT2ua1vCshsT7bfZevd8e+6LMIs1M7Qw1+/aF3iqPDbu90RvQXuVV3Xic9Ky9V0Y74ipk\nGEy79OvMiVR9m82yXdFMtkshkV73c/3UbVBD9zgVyV75zJCZ1FBJ9GgketRXQg6RTtPFBkPN\nvn2ht2qpSNXZtJ7LVBPa899jnUi2QkbHbdql38YkUjH34hSpm3ppD4fHGBhMZ0bl1FyrZdfu\nnrRjaqdVPp4rTSYxhhqqIaPU+FOP+gWd3jSt6r/p8vdktz+b2umt8pzaTQ9DlVyax25lZtA2\na6+MundUiOO4jbr028yJ9DjHXP74kLrVT02kqsNvFpGq6+irOuAXNWgH42VYrMKx2KBVPnjl\nRb+s1mua1tC8xzYllXpj9VfqW5T1fZnnb6mtGo5G7njNYNwXwx3TWuXa0Lw3t0K/QT1zMmkW\nG359emVUiPu4xXz/Hx/Mlqt5k7RoOu/SjOWT6ulbaRGpvLUru/Uy6rm/un+pFbS/4Urq1br8\nrVc+eOXMQm9d07SG+hN857r2elS3pwytlkHVN+1Ymc5I19Eq9rgvBrneKteG+t4Me6Lp/+qc\n8auutcci1b1ym/ZKX07DpBDLcZvr0i9jECmbtQZ9HlN90wywwSEciaTd2Pvp97Yt59Hc2Cz7\nu+SD26xzIrWVW27IajXN1KAuB9rz170rudFrTqR6LFYcdL0HI3d8X3XcF4N80CrXhtrejHri\nUbQ30VTbqh+GzTf1Sl9OJ9KoEMtxm+vSLzMr0imb81F/BJuFM/UOdvy1XSPpHzW5vd9BD8OP\nnaqP2pTTjwidZn5hYlz54JXDD8P0Nc3VoC2WH7SST9N5W8/ve0/L87BVQ5H05sz5MGjuoFWu\nDbW9GfZEJcrpVS+4Ha9q5WTU/PrjRpZyOpHGhViO21yXfpd4k0rYMxldhfuBSPBNinqd4DL4\ngN0WQCT4Jj/tzG1jJyREgu9yra6+NvhnQRAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQA\nARAJQABxkf4D2A8BRZIuECBZEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAgNAi\n/c+CdIUA0UAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAk\nAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQ\nAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABIgsEp7BNkAkAAEQCUAARAIQAJEA\nBEAkAAEQCUAARAIQ4DORSsPTb9rH/iWIBPvhI5HKeZHK9sswRiTYD5+IVE7OSGX/FZFgzyyZ\n2vUTuKFII80QCfbDApH6L2OResP+vUEk2A8rRCrLxpzBGYnFBtgjS0TqzzyjayT9EZFgTyw9\nI2k/IxKApEhM7WC3rFtsGD47WgFHJNgP65a/26e1TzaMikYk2AV81g5AAEQCEACRAARAJAAB\nEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACR\nAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlA\nAEQCECBtkfAMMmHHImEpyIFIiAQCIBIigQBGkYrm57JcVTQiwS6YF6ksNFYVjUiwC+ZF+tU8\n+l1VNCLBLnBO7VYWjUiwC1hsQCQQwCzSpeQaCcATo0gXFhsAvDGKVC5fZdCLRiTYBSw2IBII\nYBTpVDwlikYk2AVGkR7l8SFQNCLBLrBM7VhsAPAFkRAJBOCGLCKBAIiESCAAU7sURUq4aTDP\nlkUKWnhQEm4azOOY2j2OP+uKXjeY1w2ooIUHJeGmwTyua6RnsdgkRFpMwk2DeZyLDUztIpBw\n0xzk2/KVuET6K/L9mw1BCw9Kwk1zkG/LV+JebLisKnrdYF53WIIWHpSEm+Yg35avxCVSudgj\nRFrOZlu+XbZ8QzZo4UHZbMu3CyItLDwom235djGL9LwciuJwWf5bSYgUqmkxW75ux7aL5feR\nmoukxb+VhEihmhaz5et2bLsYRToX1S/2PY7FeVXRMUdM0MLXsa5pMVu+bse2i/NvNnBDNgTr\nmhaz5et2bLsg0sLC17GuaTFbvm7HtgtTuzB159u0oDu2XVhsCFN3vk0LumPbheXvMHXn27Sg\nO7ZduCG7vabFJOGmhQWRtte0mCTctLCYRTqpJ4oD10i5NS0mCTctLLb/RqGeZdUuuzgmCTct\nLJb/RnGrHu67vY+UbxyThJsWFm7Ibi+OScJNC4vlv1Gcn9UaeHFcVXTMERO07oTjmCTctLC4\nb8jeVxUdc8QErTvhOCYJNy0szhuy2qJdWTFXiPa8liNSpDgmCTctLJ/cRzL9PaGyD0tEih/H\nJOGmhWWdSGX/tfmCSPHjmCTctLB8IFLnSD+TG4lUMrVLIY5Jwk0LyyciNQLpM7n+61Ckf28Q\nKVIck4SbFpZPz0il9lg2aukBZ6T4cUwSblpYPv3QaidQ+9NreHmESPHjmCTctLAsEWnw06sX\nqRyujyNSpNS6hH4AAAi9SURBVDgmCTctLEundtpTg1MRZ6T48TrWFR60aSnzoUijxYY+iCIS\n8Wy8jnWFB21aynx0H6mZuE0+4BDrkw3Es/E61hUetGkpk/VvyBLPxutYV3jQpqUMIm0vXkfC\nTUsZRNpevI6Em5YyiLS9eB0JNy1lEGl78ToSblrKINL24nUk3LSUQaTtxetIuGkpg0jbi9eR\ncNNSBpG2F68j4aalDCJtL15Hwk1LGUTaXryOhJuWMoi0vdhBwk3LGETaXuwg4aZlDCJtL3aQ\ncNMyBpG2FztIuGkZg0jbix0k3LR1LY8KIm0vdrDZpkUFkbYXO9hs06KCSNuLHWy2aVFBpO3F\nDjbbtKgg0vZiB5ttWlQQaXuxg802LSqItL3YwWabFhVE2l7sYLNNiwoibS92sNmmRQWRthc7\n2GzTooJI24sdbLZpUUGk7cUONtu0qCDS9mIHm21aVBBpe7GDzTYtKoi0vdjBZpsWFUTaXuxg\ns02LCiJtL3aw2aZFBZG2FzvYbNOigkjbix1stmlRQSTiL8YOYha+EkQi3krsYN3WXqNdgUjE\neccO1m3tNdoViEScd+xg3dZeo12BSMR5xw7Wbe012hWIRJx37GDd1l6jXYFIxHnHDtZt7TXa\nFYhEnHfsYN3WXqNdgUjEeccO1m3tNdoViEScd+xg3dZeo12BSMR5xw7Wbe012hWIRJx37GDd\n1l6jXYFIxHnHDtZt7TXaFYhEnHfsYN3WXqNdgUjEe44lRrsCkYj3HEuMdgUiERMbY6/RrkAk\nYmJj7DXaFYhETGyMvUa7ApGIiY2x12hXIBIxsTH2Gu0KRCImNsZeo12BSMTExthrtCsQiZjY\nGHuNdgUiERMbY6/RrkAkYmJj7DXaFYhETGyMvUa7ApGIiY2x12hXIBIxsTH2Gu0KRCImNsZe\no12BSMTExthrtCsQiZjYGHuNdgUiERMbY6/RrkAkYmJj7DXaFYhETGyMvUa7ApGIiY2x12hX\nIBIxsTH2Gu0KRCImNsZeo12BSMTExthrtCsQiZjYGHuNdgUiERMbY6/RrkAkYmJj7DXaFYhE\nTGyMvUa7ApGIiY2x12hXuEUq31iD4QsQiXg7saBIZffFEIxegEjE24nDilQiEvE+4hAi9TM4\nRCLeSRxAJM0Xo0j/3vw3KQFgq6wQqayxn5EAdsESkcpubuee2gHsgqVnJO0pRAJAJAAB1i02\nWANEgv2w5JMNk084WD7ZALALQn/WDmAXIBKAAIgEIAAiAQiASAAChBRpyr+Z54iJtxCHE2mG\nf8TEW48RiZhYIEYkYmKB+BsiAWweRAIQAJEABEAkAAEQCUCAb4jU/4kHYuKNxl8Qqanc1D5i\n4g3EiERMLBAjEjGxQIxIxMQC8ZcWG0z/0YKYeBsxy98AAiASgADfuEbqICbeavyVa6T2i6F9\nxMTZx99ZtWseZptHTLyBGJGIiQViRCImFoi/eR9ptnXExFuIWf4GEACRAAT43tTOcLokJt5C\n/L3FBmLiDceIREwsECMSMbFA/LWPCBETbzn+4odWiYm3G7P8DSAAIgEIwH0kYmKB+Gt//MTc\nOmLi/GP+ihAxsUCMSMTEAvGXbsiW1uYRE+cef+uGrO2vhRETZx+z/A0gACIBCBBcpG7hff6U\nSEy8iTi0SGX7xdA6YuItxN8SyXSJRky8ifhLIpk/UEtMvIX4e1M7YuINx4hETCwQIxIxsUCM\nSMTEAnF4kXqIiTcb88kGAAm+KpLlzEhMnHWMSMTEAjEiERMLxIhETCwQIxIxsUCMSMTEAjEi\nERMLxIhETCwQc0MWQIDYInUfuZh/D1Cx8WMZ7Ytsz1s2rgLzZz7sdTqbZv9AiSt2QK8twd5r\n5cx3s/E8kUVqjtrL1vr+yyR2fDxqUIE5NtVtPV6upnnE5rod0GsBeq171iSxvfTYH1rtHpZ0\nrO2IvjyOuTs2jwqvY24f65a66TVzbN56Va+1kpnbblUp/F8R8krLZUPCmgkMidfLOA0JOiTo\nte/3miNylR9+aufXNOubiKWU0vwGo94ZXcfcfaY3FT78Mr/14iFBr9kLD9Br7gLac95sFH+x\nof3GFtvOqtZ3kbK+NDaGtgtbc6nd1q7LZnfdrlpMW1sbSa8ZtnY1MukzUnAWduvOodcmpH2N\n5He232EMieE4XzkOJyLFiuFz7HeC1sWOCzDX0fzWv3UhhtXUqxT2tYjFseV5n/gbZyT7BeIu\nY9ctD+LZuHsIEL+SFwlmSPhMmW7cX8VYZ1/L4pWtQ6RIpDtc04270HH/bFH8WjcD+co1krV5\nu43hY/zuny2M18GqXawYNgUiERN3+fIpBiIREwvE3EciJhbYOPbvI+001l5BnE5seN4nZvkb\noGWFSYgEULFyivG1aySzzbuNrdMM4gjxCr63aueYmO41NkH8/XgNiBQ5NkH8/XjNFAORIscm\niKPFiw4oIsWKUxgyxHNpqiLZ/vbKXuN1S0TEAeJX+iIBZAEiAUiwfIqBSPFYsUREHCZeASLF\nIuF1kL3Ga0CkWCQ8oPYarwGRYpHwgNprvAZEikXCA2qv8RoQKRorloiIw8QrQCQAARAJQABE\nAhAAkQAEQCQAARAJQABEAhAAkXKkKMw/mZ+DgNDfOYJIyUF/5wgiJQf9nSNKk9upKMpL/dOp\nOD6q4HkuivPzhUhfh/7OkUqTa6G4VD+9lSrKyp+yeurwQqSvQ3/nSKXJofh7ve7Vd0VxfL6O\nlVI/1ZdL8YtIX4f+zpFak8f151iL9J7WPaoT0UE9X5wQ6evQ3zmiNDnWc7tWmvq74XPwNejv\nHKk0OReH3+sDkRKB/s6RTpXn3NSufQV8Efo7R2p9bq9nc410rL77qdYZLq/X3/tHRPo29HeO\nVJpctGlcu/z9VMvfxR2Rvg79nSNKk/P7THRrJnmn4qRuyD7Uky9E+jr0N4AAiAQgACIBCIBI\nAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAA/we2AL00pRRyfwAAAABJRU5E\nrkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "labels$label <- factor(labels$label，levels = c(labels$label))\n",
    "par(mfrow=c(1,2))\n",
    "\n",
    "p = function(data){\n",
    "  library(ggplot2)\n",
    "  ggplot() + \n",
    "    geom_bar(aes(x = data$label, y = data$count),\n",
    "             stat=\"identity\", fill = \"lightblue\")+ \n",
    "    labs(x=\"label\",y = \"count\")+ backgroup\n",
    "}\n",
    "par(mfrow=c(1,2))\n",
    "p1 <- p(labels[c(1:25,50),]) + ggtitle(\"Numbers of Nodes with Chromosome Labels \")\n",
    "p2 <- p(labels[26:49,]) + ggtitle(\"Numbers of Nodes with Chromosome Range Labels\")\n",
    "\n",
    "cowplot::plot_grid(p1, p2, labels = c(\"A\", \"B\"), ncol = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 1 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ALLNODES</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1267023</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 1\n",
       "\\begin{tabular}{r|l}\n",
       " ALLNODES\\\\\n",
       " <int>\\\\\n",
       "\\hline\n",
       "\t 1267023\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 1\n",
       "\n",
       "| ALLNODES &lt;int&gt; |\n",
       "|---|\n",
       "| 1267023 |\n",
       "\n"
      ],
      "text/plain": [
       "  ALLNODES\n",
       "1 1267023 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query = \"MATCH (n) return count(n) as ALLNODES\"\n",
    "cypher(graph, query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in write.csv(labels, file = \"labels.csv\", sep = \",\", quote = F, :\n",
      "\"attempt to set 'col.names' ignored\"Warning message in write.csv(labels, file = \"labels.csv\", sep = \",\", quote = F, :\n",
      "\"attempt to set 'sep' ignored\""
     ]
    }
   ],
   "source": [
    "write.csv(labels, file = \"labels.csv\", sep=\",\",quote = F, col.names = TRUE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get `labels.csv` at [here](labels.csv)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 id='2'>2. Number of relationships of each type (bind, etc)<h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get numbers of relationships of each type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"MATCH ()-[r]->() RETURN type(r) as relationshipType, count(*) as count;\"\n",
    "\n",
    "relationshipType <- RNeo4j::cypher(graph,query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>relationshipType</th><th scope=col>count</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Bind       </td><td>2372086</td></tr>\n",
       "\t<tr><td>Interaction</td><td>1781980</td></tr>\n",
       "\t<tr><td>Inclusion  </td><td>1322494</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       " relationshipType & count\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t Bind        & 2372086\\\\\n",
       "\t Interaction & 1781980\\\\\n",
       "\t Inclusion   & 1322494\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 2\n",
       "\n",
       "| relationshipType &lt;chr&gt; | count &lt;int&gt; |\n",
       "|---|---|\n",
       "| Bind        | 2372086 |\n",
       "| Interaction | 1781980 |\n",
       "| Inclusion   | 1322494 |\n",
       "\n"
      ],
      "text/plain": [
       "  relationshipType count  \n",
       "1 Bind             2372086\n",
       "2 Interaction      1781980\n",
       "3 Inclusion        1322494"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "relationshipType"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get numbers of relationships between two kinds of labels."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "query <- \"MATCH (A)-[r]->(B) \n",
    "return labels(A) AS A, \n",
    "        type(r) AS relationshipType, \n",
    "        labels(B) AS B ,\n",
    "        count(r) AS count\"\n",
    "rel_details <- cypher(graph, query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>A</th><th scope=col>relationshipType</th><th scope=col>B</th><th scope=col>count</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>Inclusion</td><td>chr1_Range</td><td>134855</td></tr>\n",
       "\t<tr><td>chr2</td><td>Inclusion</td><td>chr2_Range</td><td>114769</td></tr>\n",
       "\t<tr><td>chr3</td><td>Inclusion</td><td>chr3_Range</td><td> 81332</td></tr>\n",
       "\t<tr><td>chr4</td><td>Inclusion</td><td>chr4_Range</td><td> 49491</td></tr>\n",
       "\t<tr><td>chr5</td><td>Inclusion</td><td>chr5_Range</td><td> 69435</td></tr>\n",
       "\t<tr><td>chr6</td><td>Inclusion</td><td>chr6_Range</td><td> 80840</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       " A & relationshipType & B & count\\\\\n",
       " <chr> & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t chr1 & Inclusion & chr1\\_Range & 134855\\\\\n",
       "\t chr2 & Inclusion & chr2\\_Range & 114769\\\\\n",
       "\t chr3 & Inclusion & chr3\\_Range &  81332\\\\\n",
       "\t chr4 & Inclusion & chr4\\_Range &  49491\\\\\n",
       "\t chr5 & Inclusion & chr5\\_Range &  69435\\\\\n",
       "\t chr6 & Inclusion & chr6\\_Range &  80840\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| A &lt;chr&gt; | relationshipType &lt;chr&gt; | B &lt;chr&gt; | count &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "| chr1 | Inclusion | chr1_Range | 134855 |\n",
       "| chr2 | Inclusion | chr2_Range | 114769 |\n",
       "| chr3 | Inclusion | chr3_Range |  81332 |\n",
       "| chr4 | Inclusion | chr4_Range |  49491 |\n",
       "| chr5 | Inclusion | chr5_Range |  69435 |\n",
       "| chr6 | Inclusion | chr6_Range |  80840 |\n",
       "\n"
      ],
      "text/plain": [
       "  A    relationshipType B          count \n",
       "1 chr1 Inclusion        chr1_Range 134855\n",
       "2 chr2 Inclusion        chr2_Range 114769\n",
       "3 chr3 Inclusion        chr3_Range  81332\n",
       "4 chr4 Inclusion        chr4_Range  49491\n",
       "5 chr5 Inclusion        chr5_Range  69435\n",
       "6 chr6 Inclusion        chr6_Range  80840"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(rel_details)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To see Bind relatioships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------TF Bind chromosome---------------\n",
    "library(ggplot2)\n",
    "chr_level = c(paste0(\"chr\",1:22),\"chrX\", \"chrY\")\n",
    "BindRel <- rel_details[which(rel_details$relationshipType == \"Bind\"), ] \n",
    "BindRel_TF <- BindRel[which(BindRel$A == \"TF\"),]\n",
    "BindRel_TF$B <- factor(BindRel_TF$B, levels = chr_level)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "p3 <- ggplot() + \n",
    "      geom_bar(aes(x = BindRel_TF$B, y = BindRel_TF$count),\n",
    "               stat=\"identity\", fill = \"lightblue\")+ \n",
    "      labs(x=\"label\",y = \"count\")+\n",
    "      ggtitle(\"TF Bind Chromosome\") + backgroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------LncRNA Bind chromosome---------------\n",
    "BindRel_LncRNA <- BindRel[which(BindRel$A == \"LncRNA\"),]\n",
    "BindRel_LncRNA$B <- factor(BindRel_LncRNA$B, levels = chr_level)\n",
    "\n",
    "p4 <- ggplot() + \n",
    "      geom_bar(aes(x = BindRel_LncRNA$B, y = BindRel_LncRNA$count),\n",
    "               stat=\"identity\", fill = \"lightblue\")+ \n",
    "      labs(x=\"label\",y = \"count\")+\n",
    "      ggtitle(\"LncRNA Bind Chromosome\") + backgroup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAOVBMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6et2OaysrK9vb2+vr7Hx8fQ0NDZ2dnh4eHp6enw8PD///9TaHlrAAAACXBI\nWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO2d7WKquhJA2WjrbU9blfd/2CugFi2QhJkhE1zr\nR20dM5l8rA2C3a0aABBT5S4AYAsgEoACiASgACIBKIBIAAogEoACiASgACIBKIBIAAogEoAC\niGRHNeTh5z8v2R3O1x+ncw1//Hq7PPH2NRaCLLAGdiSIVFX1qYkW6VjfGn3PNoL1YA3sSBKp\nem9iRToOWh0RyQesgTGDbT6y429PfYVsGMZ3l1PBy6Ho/HU5MO0RyQesgTFxIjX3Y1b/5Wt/\nceSni3ztqvowbPt98aj/7rj7OHWvPx8ucr0f+wynXatXc3y/ePb+c0v+uat2n83p/Z53EO+/\n3x1Oz4HRht9tbd96M7QNEMmYOJE+Bqd2ly/7/ryt3biH7rv9oO1lS38/5aiH53kXp95u7arq\noxlkfKuf8vbxn8GJYmzDD4O5KhlEMuZRpKe3SKMXG6r+IsJ7d9522eP1sb+6cGtz+f781EP9\n023w937rn5vToN3x9uRnK9mxfXjIe+zOFY/d6eVzYKThsXtq30sHdxDJmFiR3s7Nr0jtEefc\n/XA9/HzPHNgeXn/9oW33eXm4bP5D9+RPM3z4G79rEdHwp9PpoD1TZYNIxsSKVNXnwXuke8u6\n/+E8L9LvY3U9XF0PW+fu/dT1FcOHh/iuPeT0N6WCDXe37/daM7QNEMmYGQMGT53ebkeAv2I8\ntx05tZt+/fDJ4cND/Poe6XCOa/jnnwNAJHPiRGq/qyNFevt7sWH69VEiNcf33o0fRFoK02FM\nvEjV2B7f3c+07m2//l7+fsrQJJ7atXy3V9B30ad28ASzYkycSOfD773VBzEO/eHn4YZt3d/S\nOX/W3YXuUZGerxkMOqv+XGwYFhNqeOiOW78CQg8iGRMU6c7nmEjHv5e/f2/7XJg6In0/XcUe\n9N89PMT3VfV1u/wdathf/n6+lwWIZEy0SPtmTKSRG7IXk24fWt0dmwmRnu+rNs8Pw/jpmq+7\nlxVq+NGHufr9CCIZEynS/vP+ikeR/n5EqOXzYlb9/v2Q9lGk/pM+b9+DVzw+DOJN9xGjevgR\noZmGP29V+6EheACRABRAJAAFEAlAAUQCUACRABRAJAAFEAlAAUQCUACRABRAJAAFEAlAAUQC\nUACRABSwEuk/gBfAXiSjvACOQCQABRAJQAFEAlAAkQAUQCQABRAJQAFEAlAAkQAUQCQABRAJ\nQAFEAlBgJZH+N4NRvwArgkgACiASgAKIBKAAIgEogEgACiASgAKIBKAAIgEogEgACiASgAKI\nBKAAIgEogEgACiASgAKIBKAAIgEogEgACiASgAKIBKBAikj1hZTHQQ+IBNsmQaT6+iX2cdgD\nIsG2QSQABVLfIyESwAi2Iv27gEjwAiSKVDcckQD+gkgACqSJVA+/IBLAjSSR6t+viAQwIOmG\n7OABkQAGpNxHqq8fWeCTDQBP8Fk7AAV8iIRmUDiIBKAAIgEogEgACiASgAKIBKAAIgEogEgA\nCiASgAKIBKAAIgEogEgACiASgAKIBKAAIgEogEgACiASgAKIBKAAIgEogEgACiASgAKIBKAA\nIgEogEgACiASgAKIBKAAIgEogEgACiASgAKIBKAAIgEogEgACiASgAKIBKAAIgEogEgACiAS\ngAKIBKAAIgEogEgACiASgAKIBKAAIgEogEgACiASgAKIBKAAIgEogEgACpQgUqAxQH4QCUAB\nRAJQIE2kuv/acn1sZh4HPSASbJskka5+XC2pr1+mHoc9IBJsmxSR6gaRAEZZcGpXP/2ASABL\nRLq9Rbo+MyPSvwuIBC/A0iPSjEAckeD1WHLV7vodIgHcQCQABTi1A1BgoUiRFxvuPSASbJul\nn2yIeRz0gEiwbfisHYAC2xcJC2EFXl4kPAMNEAmRQAFEQiRQAJEQCRRAJEQCBRAJkUABREIk\nUACREAkUQCREAgUQybBreB0QybBreB0QybBreB0QybBreB0QybBreB0QybBreB0QybBreB0Q\nybBreB0QybBreB0QybBreB0QybBreB0QybBreB0mRaquPw//i7rlPQQ2pOVuDjS27Bpeh3GR\n6mqARg+BDWm5mwONLbuG12FcpM+BR58aPQQ2pOVuDjS27Bpeh+CpnU4PgQ1puZsDjS27hteB\niw2GXcPrMC3SoS7kPZIoNyKBCpMiHYq52CDKjUigwqRItfgqw7AH0WZf3hiRYCU2cLFBlBuR\nQIVJkd6qs2IPos2+vPGGRXJb2IsyKdKp3p/0ehBt9uWNEQlWYubUjosNiASxIJIkd0bcFvai\nbOCGrCg3IoEKiCTJnRG3hb0onNpJcmfEbWEvCiJJcouQ5TYsDBYQOLU77T9UeghsGtFuFuVG\nJFAh9B7pXElN2rJIgfA8osaI5IzgxQZO7ZaH5xE1RiRnhET6qtz/nw2i3IgEKoQvNhw0eli+\nIUWNEQlWIiRSLfUIkaYQNUYkZ3BD1rDreWS5JT2DPohk2PU8stySnkGfaZHOh11V7Q7i30pC\npAlkuSU9gz4zv490fZMk/a0kRJpAllvSM+gzKdJ71f5i32lfvWv0INoVpjvOsut5ZLklPYM+\nwf+zgRuyy8PzyHJLegZ90kTqb87Wdf8/64ceBz2IdoXpjrPseh5ZbknPoE/Sqd3Vk+uX0OOw\nB9GuMN1xll3PI8st6Rn0SbnYUDeIlBSeR5Zb0jPok3b5G5GSwvPIckt6Bn3SbsimivTvAiJN\nIMst6Rn0sRXp3oNoV5juOMuu55HllvQM+kyL9NY9Ue2GN2QRKSk8jyy3pGfQZ+6vUXTPPtyQ\nRaSk8Dyy3JKeQZ+Zv0bx0z4c/95HQqTY8Dyy3JKeQZ8FN2QRKTY8jyy3pGfQZ+avUbyf22vg\n1X7wJJ9sSArPI8st6Rn0Cd+QPWr0INoVpjvOsut5ZLklPYM+wRuy4r/t8sIi5csNa8NvyGbr\n2jI3rA0iZevaMjesDSJl69oyN6wNImXr2jI3rA0iZevaMndGAnVvFUTK1rVl7owE6t4qiJSt\na8vcGQnUvVUQKVvXlrkzEqh7qyBStq4tc2ckUPdWQaRsXVvmzkig7q2CSNm6tsydkUDdbhHW\njUjZurbMnZFA3W4R1o1I2bq2zJ2RQN1uEdaNSNm6tsydkUDdbhHWjUjZurbMnZFA3W4R1o1I\n2bq2zJ2RQN1uEdaNSNm6tsydkUDdbhHWjUjZurbMnZFA3W4R1o1I2bq2zJ2RQN1uEdaNSNm6\ntsydkUDdbhHWjUjZurbMnZFA3W4R1o1I2bq2zC1DlDtQt1uEdSNStq4tc8sQ5Q7U7RZh3YiU\nrWvL3DICuUV1u0VYNyJl69oyt4xAblHdbhHWjUjZurbMLSOQW1S3W4R1I1K2ri1zB5DlFtXt\nFmHdiJSta8vcAWS5RXW7RVg3ImXr2jJ3AFluUd1uEdaNSNm6tswdQJZbVLdbhHUjUrauLXMH\nkOUW1e0WYd2IlK1ry9wBZLlFdbtFWDciZevaMncAWW5R3W4R1o1I2bq2zB1AlltU9/LKUjde\nIrK6ESlf15a5A8hyi+peXlnqxktEVjci5evaMncAWW5R3csrS914icjqRqR8XVvmDiDLLap7\neWWpGy8RWd2IlK9ry9wBZLlFdS+vLHXjJSKrG5HydW2ZO4Ast6ju5ZWlbrxEZHUjUr6uLXMH\nkOUW1b28stSNl4isbkTK13WpuRFpDETK1nWpuTOKJMu9uGdEQiSD3Ig0BiJl67rU3Ig0BiJl\n67rU3Ig0BiJl67rU3Ig0BiJl67rU3Ig0BiJl67rU3Ig0BiJl67rU3Ig0BiJl67rU3Ig0xhKR\n6pbrYzPzOOghUKVo+kS5N7rZLXMj0hiLRBo81NOPwx4CVYqmT5R7o5vdMjcijYFI2bouNTci\njbFApHr4iEiLuy41NyKNsUSk21ukpgmJ9O8CIm0sNyKNsfSINCMQR6S4rkvNjUhjLL38jUiu\nN7tlbr/DEiHMjUjZui41t99hiRDm5tQuW9el5vY7LBHC3AtFirzYcO/BcvpEuf3uCre5/Q5L\nhDD30k82xDwOerCcPlFuv7vCbW6/wxIhzM1n7bJ1XWpuv8MSIcyNSNm6LjW332GJEOZGpGxd\nl5rb77BECHMjUrauS83td1gihLkRKVvXpeb2OywRwtyIlK3rUnP7HZYIYW5EytZ1qbn9DkuE\nMDciZeu61Nx+hyVCmBuRsnVdam6/wxIhzI1I2bouNbffYQWwzI1I+bouNbffYQWwzI1I+bou\nNbffYQWwzI1I+bouNbffYQWwzI1I+bouNbffYQWwzI1I+bouNbffYQWwzI1I+bouNbffYQWw\nzI1I+bouNbffYQWwzI1I+bouNbffYQWwzI1I+bouNbffYQWwzI1I+bouNfdGhxVoHLXNOxBp\n5a5Lzb3RYQUaR23zDkRauetSc290WIHGUdu8A5FW7rrU3BsdVqBx1DbvQKSVuy4190aHFWgc\ntc07EGnlrkvNvdFhBRpHbfMORFq561Jzb3RYgcZR27wDkVbuutTcGx1WoHHUNu9ApJW7LjX3\nRocVaBy1zTsQaeWuS8290WEFGkdt8w5EWrnrUnNvdFiBxlHbvAORVu661NwbHVagcdQ270Ck\nlbsuNfdGhxVoHLXNOxBp5a5Lzb3RYQUaR23zDkRauetSc290WIHGUdu8A5FW7rrU3BsdVqBx\n1DbvQKSVuy4190aHFWgctc07EGnlrkvNvdFhBRpHbfMORFq561Jzb3RYgcZR27wDkVbuutTc\nGx1WoHHUNu9ApJW7LjX3RocVaBy1zTsQaeWuS8290WEFGkdt8w5EWrnrUnNvdFiBxlHbvAOR\nVu661NwbHVagcdQ270CklbsuNfdGhxVoHLXNOxBp5a5Lzb3RYQUaR23zDkRauetSc290WIHG\nUdu8A5FW7rrU3BsdVqBx1DbvQKSVuy4190aHFWgctc07EGnlrkvNvdFhBRpHbfMORFq561Jz\nb3RYgcZR27wDkVbuutTcGx1WoHHUNu9ApJW7LjX3RocVaBy1zTv0RKovPPVgOURRbr8r5zb3\nRocVaBzCQKT6/uW3B8shinL7XTm3uTc6LFluRMrXdam5NzosWW5Eytd1qbk3OixZbmOR/l34\nb/bFAJtgpSMSwLZBJAAFEAlAAUQCUACRABRY6ZMNANtmpc/aAWwbRAJQAJEAFEAkAAUQCUAB\nRAJQAJEAFFhBpL/8G3kuOixqvNHcGx1WUVNmLtII/yRhUeON5t7osMqcMkQqN/dGh1XmlCFS\nubk3Oqwyp2xFkQC2CyIBKIBIAAogEoACiASggBOR7r8VWE9GH39xcORFc89PN26fryfD812G\nKqtvTDeeCQdgypKZnbJ65LvR8Dg+RLquXDMzxKdfZR+G69npvWcdDf/2OdF4ds3mKwvVPfJd\nPExZMvNTdn9yyuH55CuJFLFy/cOC2Z1b1Sa07sFd0dU+kTpmVzTBhVu+2Rum7DEsmbKbY9OV\nz6q01hFp3ufffwyW7IrZmHxXNM3kmYjlrmDKJnNPhUVTNh8Jpl/t1C6uwslzifkk9eS/Mv2/\njvO7YnrhQsdz013BlM3k1p+ycPvbIW805OM90u/QJg7Zjy8ae8ncHNT92+Op2Nyb2+mk99aB\nd87TdUvfOc/VyJSNtg7UWMIRyZqFc/vKMGWPFPEeKe6Qb9PYb25wROBwFVhKREKkgpi9EyQL\nB95/hVZyxVO7Unc7Jrmhv0gxfy1icVi20isekebfJsbd0d5S7tBdD0l4q7nvDwbhUkSCZzZ6\nlDXM/fs2Zvb0a1k42PtrXLUrkW3udsPc9+DsAWthWHbyseZ7pNkqA2FRY7+5IY3AnSBZWARX\n7bjWAAogUrG5NzqsjFMmOflApGJzb3RYhU4Z95HKzb3RYeWbsgJEKvbGhWHuwSsswuRO73qi\nVUyYy98ANwQmIRJAi+zcZPX3SNNSz4dFjR3nnj0RKfb2mNvcdvf1Vr9qNzl9s2FRY/e5p3D7\nnr3U3Ib39RDJQe4p3G7IUnNHhZedfCCSg9xTuN2QpeaO63rRUiNSztzF7shSc8+HyxBp7r9g\nCYdFjX3mll0mmg+TOz1cjEgAvkEkAA3mzy5mw4iUFcFlomCY3Onh5SBSTjZ6DcVt7uDFn+Ug\nUk5K3ZCl5kakjVLqhiw1NyJtlFI3ZKm5EWmrCC4TBcPkTg8vB5EAFEAkAAUQCUABRAJQAJEA\nFEAkAAUQCUABRCqVqpr+afo5MIK5LhVEcgVzXSqI5ArmulQ6TX7eqqo+9D+9VftTGzi/V9X7\nuUGkVWGuS6XV5LvqOLQ/XZSq6tafun1q1yDSqjDXpdJqsqu+mubYfldV+3Ozb5X6aL8cqk9E\nWhXmulR6TU7fH/tepMtp3ak9EO2656s3RFoV5rpUOk32/bndTZr+u8fnYBWY61JpNXmvdp/f\nJ0RyAHNdKndVzmOndrdXwEow16XS6/PTnK/vkfbtdx/tdYZD03xdfkSkNWGuS6XV5DA4jbtd\n/j53l7+rIyKtCnNdKp0m75cj0c/1JO+teutuyJ66JxtEWhXmGkABRAJQAJEAFEAkAAUQCUAB\nRAJQAJEAFEAkAAUQCUABRAJQAJEAFEAkAAUQCUABRAJQAJEAFEAkAAUQCUABRAJQAJEAFEAk\nAAUQCUABRAJQAJEAFEAkAAUQaSFV1H+/ePsf7XeH8+3n76fm7R+UeJ9rNN3TU+Sr/c9W374S\nygM1mO6FpIl0/WN67c/1Y/Pz9f8cHm90mu3pIXKsb42+o8sDNZjuhaSK1P7X9v3Pnw/Nv7rg\n90Sj99mehpHjoNURkVaH6V5ItEjd49f1SNTt8/Ow+eXM7vB0bjdoNN/HML67nApedDx/1VX/\nlyhY2VVhuhcy3Knt91/7639d3/+p8d3n46uq218Cux6bboFza1j9uOufGl3/2sRDF1+79o+Z\nD2r47v/+8oXj7uPUvf58uMj1fuwznHatXs3x/dLX+88t+eeuLfT+v+4/xPvvd4fTc2C04Xdb\n28Nx9cVApIU8iXT9G5Ttrvrov90PX/V53eftJYSHU6+v1qvLIelrJPXH4NTusYv+z7nsBzW8\n/z09rIfneZde337/DMzHsOi3+ilvH/8ZnCjGNvzQnOGyQKSFPInUvsN/7+w5dn9g/NC/Fxq8\ncbm9Nfqu+j+U3Dfftxvx56rdb7orvxcbHrq4vL4+9lcXbm3q/ozxoaSfboO/91v/3JwG7Y63\nJz9byY7tw0PeY3eueOxEfw6MNDxe/6h6J91rgkgLeRKpPRycq+sf//rpvt81QycO91a79tXX\n5uf+vdOjBr+N3u5vpx66uB5+vgc1VE/viR5ef/2hbdfq/NlVcz2W/D78jd+1iGj40+l0UJrd\n8kCkhTyJ9Pu4G27pmxH70+8rj61j1yZf/d57PLcbHMbai+Y3F367qO8Wzon0+1hdPb362kte\n3d+1DfP+xnftIedaVajh7vb98MD6WiDSQiZFqp5E6q++fQ9+frv8+3591f6uzP5v6tPb7Qgw\n0cX8qd3064dPDh8e4tf3SIdzXMMryfO4FV535EISRGpPierz78/H/kJAc7sb23MeSd2d+EWJ\n9Pb3YsP066NEao7vfWE/iBTB645cyKRIz6d27cNu+B7pdo2rud2N7fkaSf27cZ+7uJ5p3bv6\n+nv5+ylDk3hq1/LdXkHfRZ/avTIvPwFLmRTp+WJDG/i5HXGGB6Lmes3u+oL9n3Tnw++91ecu\n2sPPww3bur+lc/6sH64KPjR+vmYw6Kz6c7FhWEyo4cOYXxNEWsjwbOZh1z5f/u4ib78fEWof\nDn278+2Td4/vcQaHqTbJaBfPl79/b/tcmDoifT9dxR5U1D08xPfdQbK//B1q2I/5+V7WS4FI\nC5kUafSG7KnqLyYPTrKq+zW7luF1u4ES+2ZMpJEbsheTbh9a3Q07emj8fF+1eX4Yxk/XfN29\nrFDD65hf9+o3Ii1lWqTmp/24zNNHhA7dCdf95/6q3f3M7vHc7lej38+3PnXx5yNCXdJLvvr9\n8dc0HkXqP+nz9j14xePDIN50HzGqhx8Rmmk4+FjUa4JIAAogEoACiASgACIBKIBIAAogEoAC\niASgACIBKIBIAAogEoACiASgACIBKIBIAApoifQfwAuiL5JSHoCCQCQABRAJQAFEAlAAkQAU\nQCQABRAJQAFEAlAAkQAUQCQABRAJQAFEAlDASKT/zaDUD4AjEAlAAUQCUACRABRAJAAFEAlA\nAUQCUACRABRAJAAFEAlAAUQCUACRABRAJAAFEAlAAUQCUACRABRAJAAFEAlAAUQCUACRABRA\nJAAFEAlAAUQCUACRABRAJAAFEAlAAUQCUACRABRAJAAFEAlAAUQCUACRABRAJAAFEAlAAUQC\nUACRABRAJAAFEAlAAUQCUACRABRAJAAFEAlAAUQCUACRABRAJAAFEAlAAUQCUACRABRAJAAF\nEAlAAUQCUACRABRAJAAFEAlAAUQCUACRABRAJAAFEAlAAUQCUACRABRAJAAFEAlAAUQCUACR\nABRAJAAFEAlAAUQCUACRABRAJAAFEAlAAUQCUACRABRAJAAF0kSq+68Xxh4HGREJXoskka7e\nXL88Pw4zIhK8Fiki1Q0iAYyy4NQOkQCe0RXp3wVEgheEIxKAAogEoAAiASiASAAKIBKAAnyy\nAUABPmsHoAAiASiASAAKIBKo8cqrjkigxiuvOiKBGq+86ogEarzyqiMSqCFc9aI3DSKBGojU\ngkggBJFaEAmEIFILIoEQRGpBJBCCSC2IBEIQqQWRQAgitSASCEGkllVFKnrOYAJEakEkEIJI\nLYgEQhCpBZFACCK1IBIIQaQWRAIhiNSCSCAEkVoQCYQgUgsigRBEakEkEIJILa8jUtEL5hnh\nxBa9LohU2IJ5RjixRa8LIhW2YJ4RTmzR64JIhS2YZ4QTW/S6IFJhC+YZ4cQWvS6IVNiCeUY4\nsUWvCyI5WzDXxQUQ1l7y0KdFqq4/D/8YX3zG0JxknTPXC+a6uADC2kse+oRIdTVgScbQnGSd\nM9cL5rq4AMLaSx76hEifA48+l2QMzUnWOXO9YK6LCyCsveShh0/tlmUMzUnWOXO9YK6LCyCs\nveShc7HB24K5Li6AsPaShz4j0qHmPVIGXBcXQFh7yUOfFulQ8MWGQHbXC1ZyccLaXQ89xKRI\ndfJVhmHG0JzYzlkgu+sFK7k4Ye2uhx5ikxcbAtldL1jJxQlrdz30EJMivVVnQcbQnNjOWSC7\n6wUruThh7a6HHmJSpFO9Py3PGJoT2zkLZHe9YCUXJ6zd9dBDzJzacbEhByUXJ6zd9dBDIJKz\nBSu5OGHtroceYpM3ZAPZXS9YycUJa3c99BCI5GzBSi5OWLvroYfg1M7ZgpVcnLB210MPgUjO\nFqzk4oS1ux56iMCp3Wn/sShjaE5s5yyQ3fWClVycsHbXQw8Reo90rlJNQiTL2l0XJ6zd9dBD\nBC82cGq3LiUXJ6zd9dBDhET6qgr8PxsC2V0vWMnFCWt3PfQQ4YsNhyUZQ3NiO2eB7K4XrOTi\nhLW7HnqIkEh1qkeIZFq76+KEtbseeghuyDpbsJKLE9bueughEMnZgpVcnLB210MPMS3S+bCr\nqt0h+beSEMmydtfFCWt3PfQQM7+PdH2TlPpbSYhkWbvr4oS1ux56iEmR3qv2F/tO++p9ScbQ\nnNjOWSB71s6Nm9tiO7Guhx4i+H82cENWt3Pj5rbYTqzroYdApJU7d1173uLyjk0Ip3ZJ4Yji\nhEOT1W6MbXF5xyaEiw1JYUQyLC7v2IRw+TspjEiGxeUdmxBuyCaF5SKFssvCxtgWl3dsQhAp\nKYxIhsXlHZuQaZHeuieqHe+R0ooTZpeFjbEtLu/YhMz9NYruWa7aJRYnzC4LG2NbXN6xCZn5\naxQ/7cPx9e4jzYURybC4vGMTwg3ZpDAiGRaXd2xCZv4axfu5vQZe7ZdkDM2J7ZwFsgvCiGRY\nXN6xCQnfkD0uyRiaE9s5C2QXhBHJsLi8YxMSvCGb/LddNi+S5+zG2BaXd2xCuI+UFEYkw+Ly\njk2IT5FsV0QQRiTD4vKOTQgiJYURybC4vGMTgkhJYUQyLC7v2IQgUlIYkQyLyzs2IYiUFEYk\nw+Lyjk0IIiWFEcmwuLxjE4JISWFEMiwu79iEIFJSGJEMi8s7NiGIlBRGJMPi8o5NCCIlhRHJ\nsLi8YxOCSElhRDIsLu/YhCBSUhiRDIvLOzYhS0SqW66PzeBxkDE0J1lXRBBGJMPi8o5NyCKR\nBg/17+MwY2hOsq6IIIxIhsXlHZsQREoKI5JhcXnHJmSBSPXwEZFcZTfGtri8YxOyRKTbW6Sm\neRbp3wVEypjdGNvi8o5NyNIjUs0RyWN2Y2yLyzs2IUsvfyOSy+zG2BaXd2xCECkpjEiGxeUd\nmxBO7ZLCiGRYXN6xCVko0sTFhnvG0JxkXRFBGJEMi8s7NiFLP9kw9jjIGJqTrCsiCCOSYXF5\nxyaEz9olhRHJsDjXQw+BSElhRDIszvXQQyBSUhiRDItzPfQQiJQU3rhIoey2xdlmNwaRksKI\nZFicbXZjECkpjEiGxdlmNwaRksKIZFicbXZjECkpXLpIwuyui8sLIiWFEclvcXlBpKQwIvkt\nLi+IlBRGJL/F5QWRksKI5Le4vCBSUhiR/BaXF0RKCiOS3+LyUqRI+bIjkt/i8oJISWFE8ltc\nXhApKYxIfovLCyIlhRHJb3F5QaSkMCL5LS4viJQURiS/xeUFkZLCiOS3uLwgUlIYkfwWlxdE\nSgojkt/i8oJISWFE8ltcXhApKYxIfovLCyIlhRHJb3F5QaSkMCL5LS4viJQURiS/xeUFkZLC\niOS3uLwgUlIYkfwWlxdESgojkt/i8oJISWFE8ltcXhApKYxIfouTZZeCSElhRPJbnCy7FERK\nCiOS3+Jk2aUgUlIYkfwWJ8suBZGSwojktzhZdimIlBRGJL/FybJLQaSkMCL5LU6WXQoiJYUR\nyW9xsuxSEPeCljEAAAfGSURBVCkpjEh+i5Nll4JISWFE8lucLLsUREoKI5Lf4mTZpSBSUhiR\n/BYnyy4FkZLCiOS3OFl2KYiUFEYkv8XJsktBpKQwIvktTpZdCiIlhRHJb3FZsyNSWhiR3BY3\nF7afWERKCyOS2+Lmwoi0bNB22b2LZJsdkQLbvgORIsLet7ptdtfF2WaP2/YdiBQR9r7ettld\nF2ebPW7bdyBSRNj7ettmd12cbfa4bd+BSBFh7+ttm911cbbZ47Z9ByJFhL2vt21218XZZo/b\n9h2IFBH2vt622V0XZ5s9btt3IFJE2Pt622Z3XZxt9rht34FIEWHv622b3XVxttnjtn0HIkWE\nva+3bXbXxdlmj9v2HYgUEfa+3rbZXRdnmz1u23cgUkTY+3rbZnddnG32uG3fgUgRYe/rbZvd\ndXG22eO2fQciRYS9r7dtdtfF2WaP2/YdiBQR9r7ettldF2ebPW7bdyBSRNj7ettmd12cbfa4\nbd+BSBFh7+ttm911cbbZ47Z9ByJFhL2vt21218XZZo/b9h2IFBH2vt622V0XZ5s9btt3IFJE\n2Pt622Z3XZxt9rht34FIEWHv622b3XVxttnjtn0HIkWEva+3bXbXxdlmj9v2HYgUEfa+3rbZ\nXRdnmz1u23cgUkTY+3rbZnddnG32uG3fgUgRYe/rbZvddXG22eO2fQciRYS9r7dtdtfF2WaP\n2/YdiBQR9r7ettldF2ebPW7bdyBSRNj7ettmd12cbfa4bd+BSBFh7+ttm911cbbZ47Z9ByJF\nhL2vt21218XZZo/b9h2IFBH2vt622V0XZ5s9btt3IFJE2Pt622Z3XZxt9rht34FIEWHv622b\n3XVxttnjtn0HIkWEva+3bXbXxdlmj9v2HYgUEfa+3rbZXRdnmz1u23cgUkTY+3rbZnddnG32\nuG3fgUgRYe/rbZvddXG22eO2fQciRYS9r7dtdtfF2WaP2/YdiBQR9r7ettldF2ebPW7bdywX\nqb7wlDFUlu2g7bJ7X2/b7K6Ls80eREGk+v7lN2OoLNtB22X3vt622V0XZ5s9CCIlhb2vt212\n18XZZg+CSElh7+ttm911cXmzK4v078J/sy8G2CRGRySA1wKRABRAJAAFEAlAAUQCUMDokw0A\nr4XRZ+0AXgtEAlAAkQAUQCQABRAJQAFEAlAAkQAUMBDpL/9GnkuIy8Kus7sujuwpcXWRRvgn\ni8vCrrO7Lo7si+KIlCG76+LIviiOSBmyuy6O7IvihiIBvA6IBKAAIgEogEgACiASgAK5RLr/\nWmA9HX781cGRF80GZlq3gXoyPt9puLj6xkzzuXhE5903i4pr5toGJ044c6HaAjMjm7jAzNUj\n343Hx8kk0nXFmrlRPf0y+zBcz8/pPe94/LfbiXDELpwpbj4cXLF5AhMnnLnAxAlnLmpmAhM7\n3XmIqJmbSR7o1EikmK3ePwg24+wRKfCPTyg8syOi1ju8YMu3ejM9ccKZC8yMcObia1siknDL\n3SSbWfdZlayOSKGD/P1xkUizMQ2Rmmb6FMRWpMDMBSZOOHMKIjXTM2crknDLBULBHsxO7SJL\nmv13dyZNPf0vS/+vYmi9Y3b6RPbHL+PNl4s0339g4oQzF5g44cwZiyTbchEZbke90VC+iw23\nb+bCc0fT+X8+6v5t8WRw7j3tTNp789DFhpnapRcb5suUzlxc7VOJ54m62BBetmUTF5w5p0ck\ne5ZOKDBzY/h8jxR5hmQVdp0dPBI6XgUWFZHWzw7LmL8RFHeHbTIuvJRheWpX8l7HJH/0lynm\nr0UE77BNx4WLbnhECrwzDMRlYcfZA/c7ZOFNZ78/LAkH425Fggk8Hys9Z/99EzN7bjZ7h20m\nHux/s1ftimW7e900+z0YuPcYusNmcx5i+R5pvqxQXBZ2nR0WEXfvceZiw3xcBlft1s8OGwSR\n1s/uurgXzi46D0Gk9bO7Lo7sy+LcR8qQ3XVxr5vdo0i+b0jkzD58iUmY7IuzT7WLiXP5G+CK\nxCREAmjE5yHm75FmJA7EZWHX2V3f5Hrh7BLMr9pND3o+Lgu7zr7h9+tFZxeBSOtn3/JmLDm7\n6DwEkdbPvuXNuIHsy1YdkdbP7mO/kH0i6k2kuf9yJSYuC/vNHrj8IwuTfWn2xq9IAEWBSAAa\nBM4zZuOIlIPA5SFZmOyL4wIQKQOer4S8cHYRiJSBDW/GkrOLQKQMbHgzlpxdBCJlYMObseTs\nIhApB4HLQ7Iw2RfHBSASgAKIBKAAIgEogEgACiASgAKIBKAAIgEogEilUVXTP00/B8Yw56WB\nSC5hzksDkVzCnJdGp8nPW1XVh/6nt2p/agPn96p6PzeIlAXmvDRaTb6rjkP700Wpqm79qdun\ndg0iZYE5L41Wk1311TTH9ruq2p+bfavUR/vlUH0iUhaY89LoNTl9f+x7kS6ndaf2QLTrnq/e\nECkLzHlpdJrs+3O7mzT9d4/Pwaow56XRavJe7T6/T4jkCOa8NO6qnMdO7W6vgJVhzkuj1+en\nOV/fI+3b7z7a6wyHpvm6/IhIOWDOS6PV5DA4jbtd/j53l7+rIyJlgTkvjU6T98uR6Od6kvdW\nvXU3ZE/dkw0iZYE5B1AAkQAUQCQABRAJQAFEAlAAkQAUQCQABRAJQAFEAlAAkQAUQCQABRAJ\nQAFEAlDg//PbmKPyUsorAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#-------grouping plots-------------\n",
    "par(mfrow=c(1,2))\n",
    "cowplot::plot_grid(p3, p4, ncol = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To see the Inclusion relationships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAOVBMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6et2OaysrK9vb2+vr7Hx8fQ0NDZ2dnh4eHp6enw8PD///9TaHlrAAAACXBI\nWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO2diZabOtNo+ehO+j+5GfX+D3vbbQYNJSFMoQH2\nXittGwMCSTsS2K4aDAAcZqh9AABXAJEAFEAkAAUQCUABRAJQAJEAFEAkAAUQCUABRAJQAJEA\nFECkl/n5bRiGbz+/nn8+q3os6fKPvGvR0gk/DuCLt49/dQ9kApFe5M84teT4y9TvVwVEauuE\nF5E+j6cJkxDpNf6sDTn8qd+vzhepsRO2RBo+6h7JE0R6jbfPScXn/8z/fn7+P/1ev18pTd4S\nNHbCywH8/ByS6h7JE0R6iV+f3er57M/bf3+/mvXfx2df+/75n/Xjxd+3R28zf75/drvvv81z\nofnxNrz9MH+/f3bF38+N1/efz98+/vpviBv+ev98+ms9nq9u9fjz831Z5/fnJc3nVva7ywu7\ntOkNr0hrRy2e8Gry9OTv42iGb7/Dw//5Nowf82rBfrRApJf4bGGnNR5TdXva89mo34z5mOYe\n/z3XeP968e254qOR7fd/W/Om3A3/s8u31vxa57/n03cjiWSX9nxDLHJYTGrthJfT+fE0/O+8\nt1/+4T83fXfO8j+jDiK9xGcTO5e4j371+6udvj8b8p/5++gq45+vi/Q/88Ifjz735/Hwbtz3\n377+/gzfEDb887Xo/dkH5/KfB/Hrq88v63we0A9JJLu02SynyHVHbZ6wfY30GHXfh8fdxB/z\nCTn1MO16EPejBSK9xOBdIjz/KzT/5i779b/392cb//i6HJ7+f1wfwveX1s3Y8PdXr/iwyvcO\n4uNrnX9f/1+HIg2+hH6R647aPGFLpA/3sLxjm4bSn7H9aIFILyH0q/VxmP73nv4Xtzqz++C8\n/5jhvz0/o9nc8G1+/u6U7485JvnuUlp4LM6qbZ7wItL733nJv9//vYtnOu1a3o8WiPQSwkxn\nffT7ob3QfnDeny4Zvj5ezNhwwilfPIjM0pIn0OAJP5f/HJZLt9/vyxrxYxP2owUivcS38Np7\nfXypX5k/359N/LuESEFpWyK1dsLT2z/mz2N/fb7+/vMvInXGz/Bu8Ncrp+12zXQe/Hrcwn3L\nnulYhH0mPbWzSwuPRRCptROe336brneety7kegimdmeASK8xPj+m+Pdj/LrvK/Yr/xJ6WWF6\ncN6fCN8QNrRvJJiw3GCd+d1Hj/otD1VSkc6o1tgJz2//nk7r+VIYkaabDT9i+9ECkV5j+RTk\nk9h/0L+8m7rLCtOD8/7nFP/nfDd4a8PnXVz7o52wAwW3v7999dLfz/vAdmnhsUgiNXbCS5nP\n03qMNT/MH+Fmw/P29+8xth8tEOlFfs/f4XxbphTG61f+x4zGf7Df/zvtb/ybseH0Yav3/7rb\n//0PZCcTnn3NLi08Fkmkxk54KfMxCP1Zzu5rhucevvOBbLgfLRDpZX58Ns74/fl/m9yvnl98\n+fbLWsN9sN43X9+4Ge1vzCQ2XL/+Y8Jy53Ue34ZZvyL09f/15+vnC6u08FhEkdo6YUfeb/PZ\n/vr7+IDYO/zn141i+9ECkeAWnPHZkbP/U/cOUJnheX/h45Qv2NnlnLp3gMpMV0VnD0iIBBfn\n1+Pqa7pYPBFEAlAAkQAUQCQABRAJQAFEAlAAkQAUQCQABRAJQAFEAlDgdiL9PwA9ln51P5Fq\nHwBcCEQCUACRABRAJAAFEAlAAUQCUACRABRAJAAFEAlAAUQCUACRABRAJAAFEAlAgcMijcLz\nMXxLWDu+aHdJj2cTuXtDJNBEVaTpRdjL5bX1SkoXGiFHpDV+++0GbNiFrkhfr0b5rdiivB2n\nSzIniTRY/wASvC7SNI8azfoYijS/u679fHt0Fvn7WgrILcm4Ivm7ezxYy4NTjzJJhEewxcsi\nzT4sfsyTrdFexbfHfRosCuxYFqRLsjYTdjdOe1oUNv97sEMkZnawwTGRjOeFcTzwrclbFMzg\nMkoy4c0GWVl7q/xrJGZ3sEU9kdaRJS6SM0RtiOQc1DgKu/Vu6+24RjKIBGnOEWnpyCmRxInf\ndGUzXdFsl+Qdz/o3NiKtIBLoUXFEWl+LI5I9SdsqydgKRad0L4vE1A62OH6zwX40Tm+V+/I8\nq/MWRa6R8kpaX81zQX9qN4a73/c5UsbKcGMUbn+bDJG8e93CIntfu0sy7tKphNF7vf/2N0Am\nt/muXegnIoEedxApMtAhEujRu0hZ31WVV0Ak0KN3kQ6ASKAHIgEogEgACiASgAKIBKAAIgEo\ngEgS/6dEuZOByiCSBCLBThBJApFgJ4gkgUiwE0SSQCTYCSJJIBLsBJEkEAl2gkgSiAQ7QSQJ\nRIKdEERfApFgJwTRlzhVpGdFO2FVhmcs14GQrt1CEH2JM0WaBDJeoC/C5/UNQfQlThRpMLZI\nxiDSNSCIvsSJIhlZJDzqHILoS5wv0hpzkgHpEmiL5I9I7rve1E5cSx6R0iUZb9ujQfTPF8m/\nOEKkzlEWaZl9nR5EP5gFen9jI5J06gEFRcKja3COSMtbaZHm1+KIZO9mqyRjL41O6VoUyTUI\nkbpFN4h+MIQIa82zOm+ROLXLLWl9Ovt1LIj++SK5HyAhUveoBtH3BhO77yoH0Q+GLWfJ4SD6\np4oEV+Q237UL/UQk0OMOIkUGOkQCPXoX6Zwg+ogEO+ldpAMgEuiBSBKIBDtBJAlEgp0gkgQi\nwU4QCUABRAJQAJEAFEAkCa1rJC6mbgMiSSAS7ASRJBAJdoJIEogEO0EkCUSCnSCSBCLBThBJ\nApFgJ4gkgUiwE0SSQCTYyQsiub+Rm3805wVnTG8UW/RqgSbzF34OiAR6HBUpHUkrstFOtgtM\nlx2hcZGcdBVmjuK1BB/KPk0oQoMiyfG4kgWmy47Qtkh+uoo1iYUTEA8aYY9IYlaIrzdM8Hhe\nGgqpQG9RM9koDoj0IBDJGpwQqS12iLSMBaP1uL7jPp6YhkIo0CtbOIY62Sj0RHpmg7HCSSJS\nW+wVyXhCBG/Yb45y/44s8u4p7CnQhDcbZHPtbZoXyb5GGuwgx8M804NmqCDS0TQUEZGcNxrJ\nRnFcJEsde+nAiNQYCiKFVycbIokTP2My01DIHpnR+xsbkaRTD2hCJCdF5uDlU0Kkxjgukny9\nkhZpfi2OSPLAkyxweTUG63crkn+Dzh2fEKkxXrnZYD/GplnCyvOszlskTu12Fri+nKeEbWSj\nOCDS4H9k5FwxcYnUGi/d/jZW/56uO5zrD7vTKqehkAu0trYKGr3X3d3+hn64zXftQk0RCfS4\ng0iR8Q6RQI/eRWo7GwUi3YbeRToAIoEeiCSBSLATRJJAJNgJIgEogEgACiASgAKIJMEVEOwE\nkSQQCXaCSBKIBDtBJAlEgp0gkgQiwU4QSQKRYCeIJIFIsBNEkkAk2AkiSSAS7ASRJBAJdqKW\njcJeLvyKLm/RqwWazF/4OSAS6KEVRN9YAYUPWfNSgcYNfpJHXZGmGFuRlBPZJwGNoCbSuHtE\nytp/XoGmO5HW5BJipHxM6g3FbBSOSCWyUURFai0bRVjAYOIizW9ATyhmo3BF8uzx4zx6i0bj\nepGVjSImUrDdtMN1Vlg4G4VUhCTSGtEbkXpDMYi+NyK5K2Us8u4p7CnQhDcbZHPtjdoQyQ6n\nuoiER91RQSS1bBThiLRs0VY2CqmI+WaDNSKtKScQqTvIRiFRTCRjT+2WlBN41B962SjyR6T1\ntTgiBTvcLnB5MQbrNy1SkF0Cj3pFLRtFSqTVGsVsFKJI82DUVjYKqYj450hWLiToBq1sFMbr\n4UunPSsbhb+BPXy1lo0iWatwCW7zXbtQU0QCPe4gUmS8QyTQo3eRushGcfAcoQN6F+kAiAR6\nIJIEIsFOEEkCkWAniASgACIBKIBIAAogkoTyNZIy5aoIskEkidqqpClXRZANIknUViVNuSqC\nbBBJorYqacpVEWSDSBK1VUlTroogG0SSqK1KmnJVBNkgkkRtVdKUqyLIBpEkaquSplwVQTaI\nJFFblTTlqgiyQSSJ2qqkKVdFkI1aNgovOGN6o9iiVws0mb/wc0Ak0EMriL4cUCuy0U62C0yX\nHeFKIjnRiOYwRH6IIjiRBkWSw3IlC0yXHeFCIonx8YKFcCJa2Sh8kU7PRpESqbVsFMrIxyyK\ntL4H56KVjSIQybPHi9foLxqNY0VGNoqESMF20w7XTQpno1BGPmYnPczzD1liyqEVRD8ckdyV\nMhZ59xT2FGhM09kolJGOeE5RZgUPHwZ7fIJzqSCSTjYKcURatmgrG4Uy8jE/by84IxIilUMr\nG8X0p1w2Cnlqt/6NjUjSqQfUViWNfMx29HD3GR4VQCsbxZ4RaX0tjkjOwJNXoLD0ViIFmf+s\nhXhUBK1sFEmRVmv0slHIIi2zvqayUSgjHLD7kdFgLyS5RRHUslFELvTPy0bhf7PBemwtG4Uy\nmw0F5bnNd+1CTREJ9LiDSJHxDpFAj95F6iIbhTIHqwzOoHeRDoBIoAciSdRWJU25KoJsEEmi\ntippylURZINIAAogEoACiASgACJJ1L4KSlKuhiAfRJKo7UqScjUE+SCSRG1XkpSrIcgHkSRq\nu5KkXA1BPogkUduVJOVqCPJBJInariQpV0OQDyJJ1HYlSbkagnwQSaK2K0nK1RDkg0gStV1J\nUq6GIJ/DQfSXUAqlguhLBZrMHyY5IBLocVyk0YncI60RWZTNdoHpsiNcSCQ/XL79QOiTIhxP\n62K0RZKjCSULTJcd4Toi+eHyJ3Xm6Kv3m7VXQCGIvijSmUH0N0W6dhB9+ZARqS4KQfRlkTx7\nvDBz/qLROFakg+hviRRsN+1w3aTrIPryITvxIecg+gaRSqEZsniUOnP+In8Gt6NAE95skM21\nS7mSSE4MfV8kLpIKoCXSMn2y35StORxEPyzQeLu4dBB9+ZDXGPrr2DTP8xDpfNREej4rF0Tf\ne2ZG729sRJJOPaC2K0nkQ15v1E0xiq1QxYh0PsrZKJYlaZHm1+KIZM/OMgtcno7B+ncQyb9r\n52emQKTzORxEP0ek1RqFIPpJkebB6MpB9IXj5XOk6hwNoj+/LhZEXyzQuG9eO4h+/LChHrf5\nrl2oKSKBHncQKTLeIRLo0btINwyif7DG4BR6F+kAiAR6IJJEbVeSlKshyAeRJGq7kqRcDUE+\niCRR25Uk5WoI8kEkAAUQCUABRAJQAJEkuOSBnSCSBCLBThBJApFgJ4gkgUiwE0SSQCTYCSJJ\nIBLsBJEkEAl2gkgSiAQ7QSQJRIKdkI1CApFgJ2SjkKgt0jPyzxyZzo0RRFSgJiEbhURlkdwA\ndW7UOuLUtQnZKCQqi/TAifToBKpDpBYhG4VEKyINzitEahjNIPr2zQZhpa1F/gxuR4EmvNkg\nm2uX0rRIbuhhRyCukVqkRuxvlWwU4oi0rNFRNgr5CNyZnTzRg4ZQFqlcNgp5arf+jY1I0qkH\nVBfJ80ie6EFDkI1CorZI7lTO+sddu1YhG4VEZZGWD5CWl/ZCrpFahGwUEpVFgv64zXftQk0R\nCfS4g0iR8Q6RQI/eReo/G8XBCoA26F2kAyAS6IFIEogEO0EkCUSCnSASgAKIBKAAIgEogEgS\nJa+RuqBcq/QKIknU7rfNUa5VegWRJGr32+Yo1yq9gkgStfttc5RrlV5BJIna/bY5yrVKryCS\nRO1+2xzlWqVXEEmidr9tjnKt0iuIJFG73zZHuVbpFUSSqN1vm6Ncq/QKIknU7rfNUa5VekUr\nG4Xz8zrhV3R5i14t0GT+ws8BkfLJr9W7opWNwnnniDUvFZguOwIi5SPUkZgj474ZM7SyUThv\nHBNJDsuVLNAg0rmEVSTmyLhxxgytbBT+k9OzUThPjCtSV9kouiCsIiuG5eB0HkTaWnUJLzda\nj8YIY4G9klkehXiP/i7XPTjx7DYLdBcG2007XC1sKRtFF4RVNJglhuWwvDIGkY5mo/BFClfa\nWuTP4HYUaMKbDbK59kaIlE9YRcMszBT+dY5PzjVSMZF0slF4G7i76CkbRReEVRS9KmJE2lp1\no1/PVyPCPM74tihkoxBGMOdvbESSTj2gdr9tjrCKEMlFWaTlaVqk+bU4Itmzs8wCl1djsD4i\naRBWEXftXLSyUSRFWq3Ry0YhizQPRh1lo+gCoY74HMlBLRtFRKTzslHIItkFjd5rbn+/TLyq\n4MltvmsXaopI+ZzYMBfhDiJFxjtEyue0trkMvYvUfzaKLjjYSjegd5EOgEj5lGuVXkEkidr9\ntjnKtUqvIJJE7X7bHOVapVcQCUABRAJQAJEAFEAkidqXJEcpV4kwgUgStUU4SrlKhAlEkqgt\nwlHKVSJMIJJEbRGOUq4SYQKRJGqLcJRylQgTiCRRW4SjlKtEmEAkidoiHKVcJcIEIknUFuEo\n5SoRJhBJorYIRylXiTChFUTf+c2P8OOfvEWvFmgyf5jkgEigh1YQfTkOUHSjXWwXmC47wt1E\nIlbJmWgF0VcUSY4mlCwwXXaEm4nkRhi+e/QsdbSC6Dv9d1xXOi2Ivjzfk/d6uyhC0jnZIpnY\nA7yKVhB9RwV7JbM8CmHq/F2uZTlhuDYLtLYW9jpOO1wFvHoQfeGUBoNIZ6IVadUP1i2stLXI\nn8HtKNCENxtkc+1N7ijSejn0fBicB3gdNZG8NRLW6ATR9zdxD+DmQfTDM7IuhYZ5gf2ISQdR\nFqlcEH1RJOtvbESSTj2gtghHCc9oGIbBvq+wtDYi6VBjRFpfiyOSPTvLLHB5OgbrI9LMYKRR\nibt2OmgF0U+KtFqjF0RfFmkejG4eRF88KesaaRqf+BxJD7Ug+pEL/fOC6PvfbLAebx9EP35m\ncBK3+a5dqCkigR53ECky3iES6NG7SATRlzhYqbCf3kU6ACKBHogkUVuEo5SrRJhAJInaIhyl\nXCXCBCJJ1BbhKOUqESYQCUABRAJQAJEAFEAkidrXOGqUq827g0gStfu/GuVq8+4gkkTt/q9G\nudq8O4gkUbv/q1GuNu8OIknU7v9qlKvNu4NIErX7vxrlavPuIJJE7f6vRrnavDuIJFG7/6tR\nrjbvDiJJ1O7/apSrzbujmo0iiAMU30he9GqBJvMXfg6IBHpoZaMwUiCS9KJstgs0xiQkjnBT\nkdwAQlMwLn8h7EUrG4WiSHJYrmSBBpFkhHNzQtqt0Yq9yOCwE7JRSNTu/2qEp+Z5E4gEr6GW\njcKPa+fZI0SDHMNdrmU58ew2C7S2FvY6TjtcBbx6NoqF8NTmSdzyzwoFPhBN/2UqhCyOLfJn\ncDsKNOHNBtlce5ObimS8iPq2SAOj0qtoieSvkbBGJxuF98wz+ebZKBbk03OS9TkjEiK9irJI\n5bJReM/M6P2NjUjSqQfU7v9qyKdniWTN9BDpCGSjkKjd/9UITy3IQGFleUGk19HMRrFxfTLP\n6rxF4tRuZ4Hr83nxvbNRLAjn5megsNMl8TnSy5CNQqJ2/1cjfoqgy22+axdqikigxx1Eiox3\niAR69C4S2SiSHKxdyKZ3kQ6ASKAHIknU7v9qlKvNu4NIErX7vxrlavPuIJJE7f6vRrnavDuI\nBKAAIgEogEgACiCSRO1Lm2zK1RakQSSJ2n5kU662IA0iSdT2I5tytQVpEEmith/ZlKstSINI\nErX9yKZcbUEaRJKo7Uc25WoL0iCSRG0/silXW5AGkSRq+5FNudqCNIgkUduPbMrVFqRRy0Yh\nPolvJC96tUCT+Qs/B0QCPeIi/ft4G4b3/8JNvL66JnPZKVI22wUa40VGyeGyIklRgrzQQaBN\nVKS/41fwwGH8628yeq9G78lRkeSwXMkCDSKtOLG9nVfErTuPqEjvw/unQn/fh+/LolRyCCcA\n47iudFo2Cj/ioxeOy9nrdcNxSQe/jDqBSMYg0llERZqa4986GVjCy43Wo4mJ5Nnjx3n0FgXh\nHp14dpsFWlsLex2nHa4WXiYbhXDsg7FFGvzER4h0DlGRvg3/novf5yXO5MmP1+iY4FuTt8if\nwe0o0IQ3G2Rz7W2uLtJgvXKvmECf+M2G7+9/HlO79+UaKdWv3T6atkYlG0W4hXOId8lGER76\n4Iw9TgIXYxDpNBJTO5vHkmS/njppsWwUwc3u0fsbG5GkUw+o7Uc24aEvzTVYr9ZbD/FzhkMo\niRQ8SYs0vxZHpGB/2wUuz8dg/buJZHxlBuOPSnACWtko/CfCyvOszlskTu12Frg+n6d9N8lG\nIR79lzLzf3/2NdK6ELRRy0bhPVk67WnZKLwtvHmffwx2acGpB9T2I5v4KUBZZJGGwbjTuv4J\nNUUk0OMOIkXGO0QCPXr/9vfNs1EcrD1Qo3eRDoBIoEdcpI/xKlO7CIgEekRF+rjONVIERAI9\noiKNg/BTpEtByGLQY+vb3xcGkUCPxNTuX+FDKQ0igR7xmw3f3oPfxl6Lhq6Ryp00nETsA9nh\nQh/IRkAk0AORJBAJdsIHshKIBDtBJAlEgp3ERfoxGvN7GK/7aRIigR5RkX58Xhx9xba7rEmI\nBHpERXobfn/++/FnSH+xumMQCfRIfbPh1/B25W84IBLokfiu3d/vw5/HVVLhIyoGIoEeUZH+\ne8T9fgxIH/4mvllT/B43OOPWRvIime0CTeYv/BwQCfRI/B5pGH999s/AI79fW/G4g+gk0Y12\nsV1guuwIbYvk5Y8YpIXQDofzI41GWyQ5LFeywHTZEZoWyQ+TOggLoSFeCsclJYewn43rSudl\no3Cf+eG4nL02H45LPorBUscO+ohIDbLjp+ZLeLnRejQxkTx7vHiN/qLR2Y+9IKtAr2zhGNrO\nRiEfRRBreAiStEAr7PipuTN5CuM1JjJB5C3yZ3A7CjThzQbZXLuU1kUKr5GmMN7DlT+U6JXE\n7e8f3qob/drkWqOSjcJ/5l2kdZaNQjwIZ/AZnCUDI1Jr7PipeU6/LpaNwn9mRu9vbESSTj2g\nBZEcdbxXTO2aYzPR2EqyX4/eGmmR5tfiiGTPzjILXJ6Owfq9irTO6AKRuEZqkEQyZv+n5rF+\nuS3Sao1eNgpZpHkw6iwbRXgE89UpnyN1QkZ+pGVRMjlE5EL/vGwU/jcbrMfuslHEjwQ6YY9I\nXRNqikigxx1+IRsZ7xAJ9OhdpGtkozhYCVCfuEj/Pt6G4e3jumEiEQn0SNy1e14hjZcNE4lI\noEdUpO/D4/b33/fhe+EjKgYigR6b32y4yl27EGJ/gx6IBKAAUzsABbjZIFH6GqlRyrVF/3D7\nW6J2D26Ecm3RP71/IHsARNqiXFv0DyJJ1O7BjVCuLfonlh/pcl9aDUGkLcq1Rf8gkkTtHtwI\n5dqif5jaSdTuwY1Qri36B5EkavfgRijXFv2z+c2GPdG0+wKRtijXFv0jizQOQ/wayTdL/Kn5\n1kbyIpntAk3mD5McEGmL/LoEWaQflkd+eDu/XxNE/7JIVePGX3EeLn1raosdce1mRu+Vtkhy\nNKFkgemyIyDSFkLNuBHBgvhgt/WIIPoitXtwIwg1E8andJYgkpQfiSD6N0eomSWO/9IvbJHu\n65FiEH33ibDS1iJ/BrejQBPebJDNtbdBpC2Empnj+A/uGDTcfUBKBdH/8z78/fc+/F6W2I8H\nRNILou9s4RxiZ0H0G0WomehVESPSTHiz4b/hl/k3vM9Ltvr1ehWyIZI48TNmdxB9Zwgbvb+x\nEUk69YDaPbgRhJrxRRq8N25LSqRfj1vf+fmR1m03RZpfiyNSMOBsF7i8GoP1Eel1hJrxRHJD\n/SPSgyAbxc+/w5v5LdxssB+N04XXJcLK86zOWyRO7XYWuL6cp4R9BdFvFKlqxE+O5oXxGr08\nUZEeBr0/6mmN2ZCKae9cf9id9qwg+v73GOw3uwui3yjxCgKf+O3vX2+PCCjDR9njOY1QU0Ta\n4sTmuBx3+PZ3ZLxDpC1Oa5EL0rtI1wii3ygH2+ZWxH4hm/j291VApC3KtUX/IJJE7R7cCOXa\non96n9odAJG2KNcW/YNIErV7cCOUa4v+SYj049vjo6Q/RQ+nJMT+Bj2iIv17+7o+GtYvrV4N\nRAI9EtkoPh7fs/u5fmn1aiAS6JH8qfn875pwjbSfcq3TG4gkUbu/Nku51umNrandxy0TjdXu\nr81SrnV6I36zgURjEFCudXojcfv7PxKNgUe51ukNPpCVqN1fm6Vc6/TGtkj/lTmQ8iDSfsq1\nTm9ERPr9Nrx9fRL75+2yIxUi7adc6/SGLNLvr/sMfz6Ho8/LpPIHVQZE2k+51ukNWaRvw8fX\nje/3YQhndv5v5MLoWMKv6PIWyWwXaDJ/4eeASPvJr927Efs90j/zbxjehzfhO6t++J7RfQzX\niCzKZrtAE7ErCSLtR6grN5KQH1foup/ne0RzyH5Vghj5xI8nNzqP4RqxRTHksFzJAg0ilUGu\nrmiUuxsFu0uK9MtddyM5hCNSiWwUUZHIRnEWcnWFIk2PN5HoQVIkd9UlvNxoPZqYSJ49QjTI\nMdzlugcnnt1mge7CYLtph6uFZKN4EbG2luCQQRDj4cJf1fTYK5LxhPCXG/vNUe7fkUX+DG5H\ngSa82SCba2+ESPsRa0sWae5CNzGpgkhq2SjCEWl5h2wU5yDWVkIkc3uRhChCOf2abBTXRqos\nOZC+a9f1URZpeZYWaX4tjkj27CyzwOXFGKyPSJpIlSWK5C68ATu+tBrrl9sirdYoZqMQRZoH\nI7JRnINUWbM6dkYK7+OkG/BSMmaTL9Jp2Sj8DeyFZKM4i3iV3Z3b/Iwi1BSR9nNiA3XOHUSK\njHeItJ/T2qh7eheJbBRFOdhaF6Z3kQ6ASPsp1zq9gUgStftrs5Rrnd5AJIna/bVZyrVObyAS\ngAKIBKAAIgEogEgStS9FqlOuFa4CIknU7sfVKdcKVwGRJGr34+qUa4WrgEgStftxdcq1wlVA\nJIna/bg65VrhKiCSRO1+XJ1yrXAVEEmidj+uTrlWuAqIJFG7H1enXCtcBUSSqN2Pq1OuFa4C\nIknU7sfVKdcKV+EFkfzfyE3xe9zgjFsbyYteLdBk/sLPAZHi5NciPDkukhWPW4xKEl2UzXaB\n6bIjIFIcoQHbDtcAABUaSURBVE6WgEBzmolbJp2Iclik0WiLJIflShaYLjsCIsUJq2SNpGpH\n3TJ3SzoR5aVwXEJyCF+k07NRpEQiG8VRwipZY+N72twnvHeSVwJEjtajiYnk2ePHefQWjcax\nIiMbRUKkYLtph+smZKPYIKySYRmK5nxiyxt3SjoRZa9IxhPCSP3atyZvkT+D21GgCW82yOba\nmyBSnLBKhmUK5wQnXtIt3N2kCiLpZKMQR6RlDbJRHCOskvVyyAmTzzXShLJI5bJRyFO79W9s\nRJJOPaB2P65OWCWzSIN79w6RJmqMSOtrcUSyZ2eZBS5Px2B9RHqFsEpWWaypHXftFrSyUSRF\nWq3Ry0YhizQPRmSjOIZQJ8HnSOHDnVHLRhG50D8vG4X/zQbrkWwUR4lXDcjc5rt2oaaIFOfE\nhrgodxApMt4hUpzT2uKy9C4S2ShO4WCr3JDeRToAIsUp1wpXAZEkavfj6pRrhauASBK1+3F1\nyrXCVUAkAAUQCUABRAJQAJEkal+iFKJcXV8fRJKo3cMLUa6urw8iSdTu4YUoV9fXB5Ekavfw\nQpSr6+uDSBK1e3ghytX19UEkido9vBDl6vr6IJJE7R5eiHJ1fX0QSaJ2Dy9Eubq+PogkUbuH\nF6JcXV8fRJKo3cMLUa6ur4+CSO5v5oplpSAbxWHy6wq20BZJDqwVWVmtoHSZERBJOHMxywSx\ngrZpWCQ5PJdYULrMCIgknLmUZYLodRkcEUlKFhHGtzspK0VKJLJR5BGeuBW6TjQIkWIcEGkJ\nNze6j65Inj1CVMgx3NVaxphVkFemUDbZKATCE59ndgaRdnJUJBOKsaRfiURJ3Vrk3VPIKciE\nNxtkY+29I1J44otFnk/zRVK8yu6OtkjLNMpeKFtzMCuFXZDxNiUbRRbyyYdD0TAwIm2hLtLz\nRZGsFMLUbv0bG5FWEEk++dScDpFinDG1yx2R1tfiiGTP0jYKWp6O0nqIFCE88dTtOu7apdC4\n2eA9Jq9+xnVW5y2KXCNlFbQ+naeCZKPIQDjz1AdIfI6UQOX2t3EvXeyhZOm86lkpyEZxmIw2\nhkxu8127UE9EOrG6b8cdRIqMc4h0Wo3fkN5FIhvFAQ7WPVj0LtIBEKlcXV8fRJKo3cMLUa6u\nrw8iSdTu4YUoV9fXB5EAFEAkAAUQCUABRJKoffFSlHI1fmUQSaJ23y5KuRq/MogkUbtvF6Vc\njV8ZRJKo3beLUq7GrwwiSdTu20UpV+NXBpEkavftopSr8SuDSBK1+3ZRytX4lUEkidp9uyjl\navzKIJJE7b5dlHI1fmUIoi9Ru28XJb/GIE7Dsb93FJQuMwIiPRHOPwilHwbWB5eGRZKjCYkF\npcuMgEhPhPN3QumLgfXBgyD6ErX7dlHC0w9D6a9vIJIMQfQlavftooSn74XSXydziBTljNjf\nUlw7qZ9HFnn3FHIKMuHNBtlYe++I9CQ8fTd4/qIO10gJaoYsPhREXxyRljUJop+PXAVi8HxG\npCgniXR+EH15arf+jY1I0qkH1O7bRZGrQAyej0hRCKIvUbtvFyU8fULp74cg+hK1+3ZRhPMn\nlP5uCKIvUbtvFyWjpWGT23zXLtQTkZ6cWOk34g4iRcY5RHpyWr3fit5FIoj+YQ62AHzRu0gH\nQKQn5Wr8yiCSRO2+XZRyNX5lEEmidt8uSrkavzKIJFG7bxelXI1fGUQCUACRABRAJAAFEEmi\n9mVLDcpV/CVBJInanboG5Sr+kiCSRO1OXYNyFX9JEEmidqeuQbmKvySIJFG7U9egXMVfEkSS\nqN2pa1Cu4i8JIknU7tQ1KFfxlwSRJGp36hqUq/hLgkgStTt1DcpV/CVRj/1t/8xO+DVd3qK9\nBZnMX/g5IJJDfsWBgHpaF3vJEWt2FZQuMwIiOUj1MEcMskIWzw/DQDghmzNECgLSpVZ+Zb/S\nAkQ6iFANSwDwWaDBfrjfxUAK7WwUX8uXFdY31bNROAUZVySyUbyAUA2LMIi0iXY2CmN3b/tN\nszxGYkbaa697cOLZRQuyXgh7CwNEko0iRKiGwThTu0mowbjzPXigno3CFyl8c2uROznzdywW\nZMKbDbKx9kaI5CBUw5yLApG2qSjSsWwUxhfJOSSyUexGqIY1ocv0x0k9hkg2+iJZV0QbIokT\nv+nSZrqk2SzIO5r1b2xEWkEkB6EaPJGe9+kGKw0ZLJwk0vI0LdL8WhyR7FnaRkHGVig6pUOk\nFEI1uCKtj84y+EI7G0VSpNWa49koZJHmwYhsFLuR6iHxORIiuahno4iIpJ+NQhbJLmD0XnP7\nO0WyqWGL23zXLtQTkRxOrPs7cAeRIuMcIjmcVv33oHeRyEahxcGGuDu9i3QARHIoV/GXBJEk\nanfqGpSr+EuCSBK1O3UNylX8JUEkAAUQCUABRAJQAJEkal+vdE+5VmwFRJKo3Q+7p1wrtgIi\nSdTuh91TrhVbAZEkavfD7inXiq2ASBK1+2H3lGvFVkAkidr9sHvKtWIrIJJE7X7YPeVasRUQ\nSaJ2P+yecq3YCogkUbsfdk+5VmwFRJKo3Q+7p1wrtoJ67G8vSGN65diivQWZzF/4OSDSeeS3\nwlVQT+tiLzliza6C0mVGQKTzkCp1CknkPqwx87LbrU0aFkkOyyUWlC4zAiKdh1CnVpTW9WGK\nyH+BgJPa2Sj8zn5aNgpfGC8cl7M3wnGVRqhTRIqxhJkbnUcvkL1nTzxmpL32WsaYVZC1lbC3\nMEAk2ShORqhTWaTBIJIzmVr66WjkTp2/KAj8uF2QCW82yMbamyDSeUiVOjiDz3yphEiR/u29\nk7DmWDYK8RppWZNsFHUR6lQckZyxqWtOEun8bBSiSNbf2IgknXpA7X7YPUKdyiJNGS4MIh0Y\nkdbX4ohkz9I2ClqejtJ6iFQcoU4jNxsYkRKXIDGRVmuOZ6OQRZoHI7JR1EWqVD5HiiEmiYhc\n8Otno/C/2WA9ko2iNhm952Lc5rt2oZ6IdB4nNmSj3EGkyDiHSOdxWls2S+8ikY2iSQ62aof0\nLtIBEOk8yrViKyCSRO1+2D3lWrEVEEmidj/snnKt2AqIBKAAIgEogEgACiCSRO1LjBYp1zBd\ngkgStTtti5RrmC5BJInanbZFyjVMlyCSRO1O2yLlGqZLEEmidqdtkXIN0yWIJFG707ZIuYbp\nEkSSqN1pW6Rcw3QJIknU7rQtUq5hugSRJGp32hYp1zBdckoQ/SB8SWzl2KK9BZnMHyY5INIu\n8iv2luiLFH0ntkihIOPGbMgDkXYh1VMqkMnQf0CTPTQskhxNKPY2Ip2MUE3JON53ssgQRF+m\ndqdtEaGarHFoFWlacjOP9IPo++G4PHvioe7stdcyxqyCrK2EvYVx7Qiivx+hmgZ/DLJFutfM\nrmak1diiYAa3XZAJbzbIxtqbINIuhGoalujdgy3SVTK17EJbJP+dhDXHgugbb1XXYILoqyNU\n05pZwgqIP79hECmTVP8+P4i+8866w+VvbERaQaRdCNVkjTprQHxjEGkfR6d21mtxRLJnaRsF\nLU9HaT1E0kCoplRAfKZ22UQvQTauU+ZZnbdInNplFrQ+nxcTRF8bqZ42PkeKV/D1IIi+RO1O\n2yIZHeLO3Oa7dqGeiLSLE9vmCtxBpMg4h0i7OK15rkHvIhFEvxQHG+rq9C7SARBpF+UapksQ\nSaJ2p22Rcg3TJYgkUbvTtki5hukSRJKo3WlbpFzDdAkiASiASAAKIBKAAogkUft6RIVyNQmI\nJFPbARXK1SQgkkxtB1QoV5OASDK1HVChXE0CIsnUdkCFcjUJiCRT2wEVytUkIJJMbQdUKFeT\ngEgytR1QoVxNAiLJ1HZAhXI1CSfE/rYXCL+my1u0tyCT+Qs/B0QCPU4QadwpkkJBxg1+ksct\nRYpH/bnflEQVfZHGvSNS5n5TBRlECpFOzI0m7EbuRqQjaGejcAIxjuub6tko/IiPXjguZ2/3\nDMclnVhcpIER6RDa2SgCkTx7hDiPY7irtYwxqyBrK2FvYYDIW2SjEM5rcLIXua8Q6RBnhCyW\nAkRK/TyyyJ2cjTkFmfBmg2ysvc2dRRqG4BUiHUJZJLevpq05lI0iXNM5JLJRSCIN65Ru8F8h\n0jG0RZo66+nZKIKb3aP3NzYiSaceUNsBFcLTcvNF+NkjEOkIZ6R1yR2R1tfiiBTsL16QsRWK\nTuluL5Ix7lUR10h6aGejWN6Q3lytOZ6Nwl9zXB+EAu6YjUI8Mz5HOgn1bBTGiCLpZ6Mwokh2\nAaP3mtvfcBq3+a5dqCcigR53ECkyziES6NG7SGSjiHKwZmEXvYt0AEQCPRBJorYDKpSrSUAk\nmdoOqFCuJgGRZGo7oEK5mgREAlABkQAUQCQABRBJovblzQUo145tgEgStXvhBSjXjm2ASBK1\ne+EFKNeObYBIErV74QUo145tgEgStXvhBSjXjm2ASBK1e+EFKNeObYBIErV74QUo145tgEgS\ntXvhBSjXjm2ASBK1e+EFKNeObaAe+9sL0pheObZob0Em8xd+Doh0JvntcA20RZIDakVWViso\nXWYERDoTqVrdeEXTK/ehWxoWSQ7LJRaULjMCIp2JUKtu3osleL/90C/q2Si+3jDL43nZKOyC\n3DLJRtEAQq2GeS8G6aFP1LNRLO94b5rlUYj36O9qLcOJZxctyCtTKPuG2SjqItRqGNMVkR44\nk6k1bmMiI0TeIu+eQk5BJrzZIBtr7x2RzkSqVi+m6/zqCh6dEvs705pD2Sjcgoy3KdkoqiPU\n6saIhEjTH7t/n56Nwj2EdYfL39iItIJIZyLUalqkvj06JdFY7oi0vhZHpOBKKF7Q8nSU1kOk\nCgi16ork3a7r3CP1bBRJkVZrjmejkEWaByOyUdRGqtbE50hWsqY+Uc9GEbng189G4X+zwXok\nG0V9MvrPpbjNd+1CPRHpTE5syia5g0iRcQ6RzuS01myU3kUiG0WjHGzX7uhdpAMg0pmUa8c2\nQCSJ2r3wApRrxzZAJInavfAClGvHNkAkAAUQCUABRAJQAJEkal9ggBI1ehMirdRuf1CiRm9C\npJXa7Q9K1OhNiLRSu/1BiRq9CZFWarc/KFGjNyHSSu32ByVq9CZEWqnd/qBEjd6ESCu12x+U\nqNGbEGmldvuDEjV6EyKt1G5/UKJGbyIbxUrt9gcl8nvDURoOor+joHSZERDp+sQa2Avuascx\nejE7RsMiyWG5xILSZUZApOsTaV8/WLLl1avZMchGIVG7/UEJuXmHOURlRCRTVqQlzNxoPa7v\neG+a5VGI9+jvai3DiWcXLcgrUyibbBS3JNbAXtBxK27l8NSquEjGE8N+I7Amb5F3TyGnIBPe\nbJCNtbdBpOsTa+CESMM8tdt3kVRRpOPZKJw1nUMiGwU8iDWwF73fGZGWtIK1RVqvRjZEEid+\n06XNtJPNgryjWf/GRqQVRLo+sQZ2RJpv1y2LXknYdEKiMWeNtEjza3FECgaeeEHGVig6pUOk\n2xFrYOH2tytSlZsN9qM/3RJWmmd13qLINVJWQevLeSpINgr4v1dEqvE5kpQkwrkOsTuvdjYK\n/3sM3sTSL9suJTj1gNrtD0pk9mQFbvNdu1BPRLo+J3aoaG+6rkiRcQ6Rrs9pfSrRm/oUiWwU\nkOBg79pB7yIdAJGuT43ehEgrtdsflKjRmxBppXb7gxI1ehMiAbwMIgEogEgACiCSRO2pPTRK\nTm9CpJXa7QWNktObEGmldntBo+T0JkRaqd1e0Cg5vQmRVmq3FzRKTm9CpJXa7QWNktObEGml\ndntBo+T0JkRaqd1e0Cg5vQmRVmq3FzRKTm9SCqJvvxZ+BJS3aG9BJvOHSQ6IBHvJ6U1aIo17\nRVIoyLgxG/JAJNhLrMMMJ6R12T0i5e03WZBBJChBpL8Mx0SKxLZ3RDoziH5UJILowznI3WU4\nNiIt0bFG69H4Inn2xEPd2Wuve3DCcEULsl4Iewvj2hFEH14i1mGOi2Q8MYzxRArf3Frkz+Ay\nCjLhzQbZWHsjRIK9xDpMVZEOB9EPR6TlHYLowxnEOsxpIhFEH65IrMPUHZHW1+KIZM/SNgoy\ntkLRKR0iwUFiHUbnZoPXT+MirdYoBNEXRZoHI4LowxnEOozS7W8T6d9259UOou+v6E0s/bLt\nUiYQCfYS7zK3+a5dqCciwV7iXeYOIkXGOUSCvcS7TO8iEUQfChLvMr2LdABEgr3k9CZEWqnd\nXtAoOb0JkVZqtxc0Sk5vQqSV2u0FjZLTmxAJ4GUQCUABRAJQAJEAFLizSHH+l3ivIziNgiz9\n6nYiJfhf7QPQgdOoASKtdNZ0MTiNGiDSSmdNF4PTqAEirXTWdDE4jRogEoACiASgACIBKIBI\nAAog0sSY9avbTuj9JJ5hArpqDER6MgrPOmQcL/Efwho9qpvTQKQnFxHJCT3bL4tI/ZwJIj25\nikix4El9gUjdchmR1uiyHTPH9uzoTBBp4hLXFk/GfrpfnGc+hn5OBJEuSEf97zIgEoACiDRx\nkakdp1EJRHpykZsNnEYtEOlJh00nwWnUApGedNh0EpxGLRDpSYdNJ8Fp1AKRJvq7vBXhNCqB\nSAAKIBKAAohk0c9EIgmnUQFEsuir6aJwGhVAJIu+mi4Kp1EBRJrp7TZRBE6jDohk0VG7peA0\nKoBIFn01XRROowKIZNFX00XhNCqASAAKIBKAAoi0Ml4hbginUQdEWhjtMFD9wmlUAZEWemu6\nCJxGFRBpobemi8BpVAGRFnprugicRhUQaaW3b6VE4DRqgEgACiASgAKINNFflACJKalQ52fR\nI4j0pMO4NRLL9TkqFQaRnlxNpM5Po7/Eg4j0BJGaor+jR6QnVxGpvwxdMt0dPiJNdDeXiNFb\nhq6LgEgACiASgAKIBKAAIgEogEgACiASgAKIBKAAIgEogEiQxw8+402BSJDHQFdJQe1AHoiU\nhNqBv9+G8ePryfdh+P7XzNI8/g7D9O4wYFIKKuf2/Bsfknybn4z/XJG+Fn4g0gZUzu35GL6b\n3w9LPoZ3Y96HD1ek93/mxzAytduA2rk9b8O/+cnntO7v8OZN7ZZnFY+xfaid27MYYusTewYR\nqJ3bg0gaUDu3Jza1+4tIO6B2bs/H8GH+uDcbxuGn+feOSDugdm7P368b3G/27e+Px5P/fJH4\njlACRII/79PnsMsHsuZjHP7zrpF+IFIKRAJQAJEAFEAkAAUQCUABRAJQAJEAFEAkAAUQCUAB\nRAJQAJEAFEAkAAUQCUCB/w/Kuc2YGticmwAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#--------chr inclusion chr_Range----------------------------\n",
    "InclusionRel <- rel_details[rel_details$relationshipType ==\"Inclusion\",]\n",
    "\n",
    "InclusionRel$rel <- apply(InclusionRel[1:3], 1， FUN = function(X){ y = paste0(X[1],\"-\",X[2],\"->\",X[3]) })\n",
    "chrRange_level <- c(paste0(\"chr\",1:22,\"-Inclusion->chr\",1:22,\"_Range\"),\n",
    "                    \"chrX-Inclusion->chrX_Range\",\n",
    "                    \"chrY-Inclusion->chrY_Range\")\n",
    "InclusionRel$rel <- factor(InclusionRel$rel, levels = chrRange_level)\n",
    "\n",
    "ggplot() + \n",
    "  geom_bar(aes(x = InclusionRel$rel, y = InclusionRel$count),\n",
    "           stat=\"identity\", fill = \"lightblue\")+ \n",
    "  labs(x=\"Relationship\",y = \"count\")+\n",
    "  ggtitle(\"Chromosome Inclusion Chromosome Range\") + backgroup +  #设置背景为白色\n",
    "  coord_flip()+ #将条形图坐标轴转换过来变成横向条形图\n",
    "geom_text(aes(x = InclusionRel$rel, y = InclusionRel$count, label=InclusionRel$count), size = 2.5,hjust=-0.5 )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To see the Interaction relationships"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I don't how to create a picture for the Interaction relationships, so I show you its data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 227 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>A</th><th scope=col>relationshipType</th><th scope=col>B</th><th scope=col>count</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>49</th><td>TF        </td><td>Interaction</td><td>TF         </td><td>  2488</td></tr>\n",
       "\t<tr><th scope=row>50</th><td>chr1_Range</td><td>Interaction</td><td>chr1_Range </td><td>166397</td></tr>\n",
       "\t<tr><th scope=row>51</th><td>chr1_Range</td><td>Interaction</td><td>chr2_Range </td><td>    12</td></tr>\n",
       "\t<tr><th scope=row>52</th><td>chr1_Range</td><td>Interaction</td><td>chr9_Range </td><td>     3</td></tr>\n",
       "\t<tr><th scope=row>53</th><td>chr1_Range</td><td>Interaction</td><td>chr5_Range </td><td>    14</td></tr>\n",
       "\t<tr><th scope=row>54</th><td>chr1_Range</td><td>Interaction</td><td>chr4_Range </td><td>     6</td></tr>\n",
       "\t<tr><th scope=row>55</th><td>chr1_Range</td><td>Interaction</td><td>chr3_Range </td><td>    12</td></tr>\n",
       "\t<tr><th scope=row>56</th><td>chr1_Range</td><td>Interaction</td><td>chr17_Range</td><td>    12</td></tr>\n",
       "\t<tr><th scope=row>57</th><td>chr1_Range</td><td>Interaction</td><td>chr7_Range </td><td>     8</td></tr>\n",
       "\t<tr><th scope=row>58</th><td>chr1_Range</td><td>Interaction</td><td>chr15_Range</td><td>     6</td></tr>\n",
       "\t<tr><th scope=row>59</th><td>chr1_Range</td><td>Interaction</td><td>chr16_Range</td><td>     3</td></tr>\n",
       "\t<tr><th scope=row>60</th><td>chr1_Range</td><td>Interaction</td><td>chr10_Range</td><td>     5</td></tr>\n",
       "\t<tr><th scope=row>61</th><td>chr1_Range</td><td>Interaction</td><td>chr13_Range</td><td>     2</td></tr>\n",
       "\t<tr><th scope=row>62</th><td>chr1_Range</td><td>Interaction</td><td>chr8_Range </td><td>    11</td></tr>\n",
       "\t<tr><th scope=row>63</th><td>chr1_Range</td><td>Interaction</td><td>chr12_Range</td><td>     9</td></tr>\n",
       "\t<tr><th scope=row>64</th><td>chr1_Range</td><td>Interaction</td><td>chr11_Range</td><td>     2</td></tr>\n",
       "\t<tr><th scope=row>65</th><td>chr1_Range</td><td>Interaction</td><td>chr14_Range</td><td>     3</td></tr>\n",
       "\t<tr><th scope=row>66</th><td>chr1_Range</td><td>Interaction</td><td>chr6_Range </td><td>     4</td></tr>\n",
       "\t<tr><th scope=row>67</th><td>chr1_Range</td><td>Interaction</td><td>chr19_Range</td><td>     2</td></tr>\n",
       "\t<tr><th scope=row>68</th><td>chr1_Range</td><td>Interaction</td><td>chr20_Range</td><td>    11</td></tr>\n",
       "\t<tr><th scope=row>69</th><td>chr1_Range</td><td>Interaction</td><td>chrX_Range </td><td>     4</td></tr>\n",
       "\t<tr><th scope=row>70</th><td>chr1_Range</td><td>Interaction</td><td>chr21_Range</td><td>     1</td></tr>\n",
       "\t<tr><th scope=row>71</th><td>chr2_Range</td><td>Interaction</td><td>chr2_Range </td><td>136934</td></tr>\n",
       "\t<tr><th scope=row>72</th><td>chr2_Range</td><td>Interaction</td><td>chr21_Range</td><td>     4</td></tr>\n",
       "\t<tr><th scope=row>73</th><td>chr2_Range</td><td>Interaction</td><td>chr11_Range</td><td>     1</td></tr>\n",
       "\t<tr><th scope=row>74</th><td>chr2_Range</td><td>Interaction</td><td>chr8_Range </td><td>    13</td></tr>\n",
       "\t<tr><th scope=row>75</th><td>chr2_Range</td><td>Interaction</td><td>chr6_Range </td><td>     5</td></tr>\n",
       "\t<tr><th scope=row>76</th><td>chr2_Range</td><td>Interaction</td><td>chr20_Range</td><td>    10</td></tr>\n",
       "\t<tr><th scope=row>77</th><td>chr2_Range</td><td>Interaction</td><td>chr7_Range </td><td>    11</td></tr>\n",
       "\t<tr><th scope=row>78</th><td>chr2_Range</td><td>Interaction</td><td>chr9_Range </td><td>     5</td></tr>\n",
       "\t<tr><th scope=row>...</th><td>...</td><td>...</td><td>...</td><td>...</td></tr>\n",
       "\t<tr><th scope=row>246</th><td>chr18_Range</td><td>Interaction</td><td>chr8_Range </td><td>    1</td></tr>\n",
       "\t<tr><th scope=row>247</th><td>chr18_Range</td><td>Interaction</td><td>chr6_Range </td><td>    1</td></tr>\n",
       "\t<tr><th scope=row>248</th><td>chr19_Range</td><td>Interaction</td><td>chr19_Range</td><td>69619</td></tr>\n",
       "\t<tr><th scope=row>249</th><td>chr19_Range</td><td>Interaction</td><td>chr3_Range </td><td>    1</td></tr>\n",
       "\t<tr><th scope=row>250</th><td>chr19_Range</td><td>Interaction</td><td>chr20_Range</td><td>    2</td></tr>\n",
       "\t<tr><th scope=row>251</th><td>chr19_Range</td><td>Interaction</td><td>chr2_Range </td><td>    3</td></tr>\n",
       "\t<tr><th scope=row>252</th><td>chr19_Range</td><td>Interaction</td><td>chr6_Range </td><td>    1</td></tr>\n",
       "\t<tr><th scope=row>253</th><td>chr19_Range</td><td>Interaction</td><td>chr7_Range </td><td>    1</td></tr>\n",
       "\t<tr><th scope=row>254</th><td>chr19_Range</td><td>Interaction</td><td>chr8_Range </td><td>    1</td></tr>\n",
       "\t<tr><th scope=row>255</th><td>chr20_Range</td><td>Interaction</td><td>chr3_Range </td><td>   26</td></tr>\n",
       "\t<tr><th scope=row>256</th><td>chr20_Range</td><td>Interaction</td><td>chr9_Range </td><td>   10</td></tr>\n",
       "\t<tr><th scope=row>257</th><td>chr20_Range</td><td>Interaction</td><td>chr8_Range </td><td>    6</td></tr>\n",
       "\t<tr><th scope=row>258</th><td>chr20_Range</td><td>Interaction</td><td>chr7_Range </td><td>    8</td></tr>\n",
       "\t<tr><th scope=row>259</th><td>chr20_Range</td><td>Interaction</td><td>chr6_Range </td><td>    7</td></tr>\n",
       "\t<tr><th scope=row>260</th><td>chr20_Range</td><td>Interaction</td><td>chr5_Range </td><td>   11</td></tr>\n",
       "\t<tr><th scope=row>261</th><td>chr20_Range</td><td>Interaction</td><td>chr22_Range</td><td>    2</td></tr>\n",
       "\t<tr><th scope=row>262</th><td>chr20_Range</td><td>Interaction</td><td>chr4_Range </td><td>    5</td></tr>\n",
       "\t<tr><th scope=row>263</th><td>chr20_Range</td><td>Interaction</td><td>chr21_Range</td><td>    4</td></tr>\n",
       "\t<tr><th scope=row>264</th><td>chr20_Range</td><td>Interaction</td><td>chrX_Range </td><td>    6</td></tr>\n",
       "\t<tr><th scope=row>265</th><td>chr20_Range</td><td>Interaction</td><td>chr20_Range</td><td>65699</td></tr>\n",
       "\t<tr><th scope=row>266</th><td>chr21_Range</td><td>Interaction</td><td>chr8_Range </td><td>    4</td></tr>\n",
       "\t<tr><th scope=row>267</th><td>chr21_Range</td><td>Interaction</td><td>chr5_Range </td><td>    2</td></tr>\n",
       "\t<tr><th scope=row>268</th><td>chr21_Range</td><td>Interaction</td><td>chr21_Range</td><td>33968</td></tr>\n",
       "\t<tr><th scope=row>269</th><td>chr21_Range</td><td>Interaction</td><td>chr4_Range </td><td>    1</td></tr>\n",
       "\t<tr><th scope=row>270</th><td>chr21_Range</td><td>Interaction</td><td>chr3_Range </td><td>    3</td></tr>\n",
       "\t<tr><th scope=row>271</th><td>chr22_Range</td><td>Interaction</td><td>chr22_Range</td><td>55182</td></tr>\n",
       "\t<tr><th scope=row>272</th><td>chr22_Range</td><td>Interaction</td><td>chr7_Range </td><td>    1</td></tr>\n",
       "\t<tr><th scope=row>273</th><td>chr22_Range</td><td>Interaction</td><td>chr4_Range </td><td>    1</td></tr>\n",
       "\t<tr><th scope=row>274</th><td>chrX_Range </td><td>Interaction</td><td>chrX_Range </td><td>20308</td></tr>\n",
       "\t<tr><th scope=row>275</th><td>chrY_Range </td><td>Interaction</td><td>chrY_Range </td><td>  321</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 227 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & A & relationshipType & B & count\\\\\n",
       "  & <chr> & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t49 & TF         & Interaction & TF          &   2488\\\\\n",
       "\t50 & chr1\\_Range & Interaction & chr1\\_Range  & 166397\\\\\n",
       "\t51 & chr1\\_Range & Interaction & chr2\\_Range  &     12\\\\\n",
       "\t52 & chr1\\_Range & Interaction & chr9\\_Range  &      3\\\\\n",
       "\t53 & chr1\\_Range & Interaction & chr5\\_Range  &     14\\\\\n",
       "\t54 & chr1\\_Range & Interaction & chr4\\_Range  &      6\\\\\n",
       "\t55 & chr1\\_Range & Interaction & chr3\\_Range  &     12\\\\\n",
       "\t56 & chr1\\_Range & Interaction & chr17\\_Range &     12\\\\\n",
       "\t57 & chr1\\_Range & Interaction & chr7\\_Range  &      8\\\\\n",
       "\t58 & chr1\\_Range & Interaction & chr15\\_Range &      6\\\\\n",
       "\t59 & chr1\\_Range & Interaction & chr16\\_Range &      3\\\\\n",
       "\t60 & chr1\\_Range & Interaction & chr10\\_Range &      5\\\\\n",
       "\t61 & chr1\\_Range & Interaction & chr13\\_Range &      2\\\\\n",
       "\t62 & chr1\\_Range & Interaction & chr8\\_Range  &     11\\\\\n",
       "\t63 & chr1\\_Range & Interaction & chr12\\_Range &      9\\\\\n",
       "\t64 & chr1\\_Range & Interaction & chr11\\_Range &      2\\\\\n",
       "\t65 & chr1\\_Range & Interaction & chr14\\_Range &      3\\\\\n",
       "\t66 & chr1\\_Range & Interaction & chr6\\_Range  &      4\\\\\n",
       "\t67 & chr1\\_Range & Interaction & chr19\\_Range &      2\\\\\n",
       "\t68 & chr1\\_Range & Interaction & chr20\\_Range &     11\\\\\n",
       "\t69 & chr1\\_Range & Interaction & chrX\\_Range  &      4\\\\\n",
       "\t70 & chr1\\_Range & Interaction & chr21\\_Range &      1\\\\\n",
       "\t71 & chr2\\_Range & Interaction & chr2\\_Range  & 136934\\\\\n",
       "\t72 & chr2\\_Range & Interaction & chr21\\_Range &      4\\\\\n",
       "\t73 & chr2\\_Range & Interaction & chr11\\_Range &      1\\\\\n",
       "\t74 & chr2\\_Range & Interaction & chr8\\_Range  &     13\\\\\n",
       "\t75 & chr2\\_Range & Interaction & chr6\\_Range  &      5\\\\\n",
       "\t76 & chr2\\_Range & Interaction & chr20\\_Range &     10\\\\\n",
       "\t77 & chr2\\_Range & Interaction & chr7\\_Range  &     11\\\\\n",
       "\t78 & chr2\\_Range & Interaction & chr9\\_Range  &      5\\\\\n",
       "\t... & ... & ... & ... & ...\\\\\n",
       "\t246 & chr18\\_Range & Interaction & chr8\\_Range  &     1\\\\\n",
       "\t247 & chr18\\_Range & Interaction & chr6\\_Range  &     1\\\\\n",
       "\t248 & chr19\\_Range & Interaction & chr19\\_Range & 69619\\\\\n",
       "\t249 & chr19\\_Range & Interaction & chr3\\_Range  &     1\\\\\n",
       "\t250 & chr19\\_Range & Interaction & chr20\\_Range &     2\\\\\n",
       "\t251 & chr19\\_Range & Interaction & chr2\\_Range  &     3\\\\\n",
       "\t252 & chr19\\_Range & Interaction & chr6\\_Range  &     1\\\\\n",
       "\t253 & chr19\\_Range & Interaction & chr7\\_Range  &     1\\\\\n",
       "\t254 & chr19\\_Range & Interaction & chr8\\_Range  &     1\\\\\n",
       "\t255 & chr20\\_Range & Interaction & chr3\\_Range  &    26\\\\\n",
       "\t256 & chr20\\_Range & Interaction & chr9\\_Range  &    10\\\\\n",
       "\t257 & chr20\\_Range & Interaction & chr8\\_Range  &     6\\\\\n",
       "\t258 & chr20\\_Range & Interaction & chr7\\_Range  &     8\\\\\n",
       "\t259 & chr20\\_Range & Interaction & chr6\\_Range  &     7\\\\\n",
       "\t260 & chr20\\_Range & Interaction & chr5\\_Range  &    11\\\\\n",
       "\t261 & chr20\\_Range & Interaction & chr22\\_Range &     2\\\\\n",
       "\t262 & chr20\\_Range & Interaction & chr4\\_Range  &     5\\\\\n",
       "\t263 & chr20\\_Range & Interaction & chr21\\_Range &     4\\\\\n",
       "\t264 & chr20\\_Range & Interaction & chrX\\_Range  &     6\\\\\n",
       "\t265 & chr20\\_Range & Interaction & chr20\\_Range & 65699\\\\\n",
       "\t266 & chr21\\_Range & Interaction & chr8\\_Range  &     4\\\\\n",
       "\t267 & chr21\\_Range & Interaction & chr5\\_Range  &     2\\\\\n",
       "\t268 & chr21\\_Range & Interaction & chr21\\_Range & 33968\\\\\n",
       "\t269 & chr21\\_Range & Interaction & chr4\\_Range  &     1\\\\\n",
       "\t270 & chr21\\_Range & Interaction & chr3\\_Range  &     3\\\\\n",
       "\t271 & chr22\\_Range & Interaction & chr22\\_Range & 55182\\\\\n",
       "\t272 & chr22\\_Range & Interaction & chr7\\_Range  &     1\\\\\n",
       "\t273 & chr22\\_Range & Interaction & chr4\\_Range  &     1\\\\\n",
       "\t274 & chrX\\_Range  & Interaction & chrX\\_Range  & 20308\\\\\n",
       "\t275 & chrY\\_Range  & Interaction & chrY\\_Range  &   321\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 227 × 4\n",
       "\n",
       "| <!--/--> | A &lt;chr&gt; | relationshipType &lt;chr&gt; | B &lt;chr&gt; | count &lt;int&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 49 | TF         | Interaction | TF          |   2488 |\n",
       "| 50 | chr1_Range | Interaction | chr1_Range  | 166397 |\n",
       "| 51 | chr1_Range | Interaction | chr2_Range  |     12 |\n",
       "| 52 | chr1_Range | Interaction | chr9_Range  |      3 |\n",
       "| 53 | chr1_Range | Interaction | chr5_Range  |     14 |\n",
       "| 54 | chr1_Range | Interaction | chr4_Range  |      6 |\n",
       "| 55 | chr1_Range | Interaction | chr3_Range  |     12 |\n",
       "| 56 | chr1_Range | Interaction | chr17_Range |     12 |\n",
       "| 57 | chr1_Range | Interaction | chr7_Range  |      8 |\n",
       "| 58 | chr1_Range | Interaction | chr15_Range |      6 |\n",
       "| 59 | chr1_Range | Interaction | chr16_Range |      3 |\n",
       "| 60 | chr1_Range | Interaction | chr10_Range |      5 |\n",
       "| 61 | chr1_Range | Interaction | chr13_Range |      2 |\n",
       "| 62 | chr1_Range | Interaction | chr8_Range  |     11 |\n",
       "| 63 | chr1_Range | Interaction | chr12_Range |      9 |\n",
       "| 64 | chr1_Range | Interaction | chr11_Range |      2 |\n",
       "| 65 | chr1_Range | Interaction | chr14_Range |      3 |\n",
       "| 66 | chr1_Range | Interaction | chr6_Range  |      4 |\n",
       "| 67 | chr1_Range | Interaction | chr19_Range |      2 |\n",
       "| 68 | chr1_Range | Interaction | chr20_Range |     11 |\n",
       "| 69 | chr1_Range | Interaction | chrX_Range  |      4 |\n",
       "| 70 | chr1_Range | Interaction | chr21_Range |      1 |\n",
       "| 71 | chr2_Range | Interaction | chr2_Range  | 136934 |\n",
       "| 72 | chr2_Range | Interaction | chr21_Range |      4 |\n",
       "| 73 | chr2_Range | Interaction | chr11_Range |      1 |\n",
       "| 74 | chr2_Range | Interaction | chr8_Range  |     13 |\n",
       "| 75 | chr2_Range | Interaction | chr6_Range  |      5 |\n",
       "| 76 | chr2_Range | Interaction | chr20_Range |     10 |\n",
       "| 77 | chr2_Range | Interaction | chr7_Range  |     11 |\n",
       "| 78 | chr2_Range | Interaction | chr9_Range  |      5 |\n",
       "| ... | ... | ... | ... | ... |\n",
       "| 246 | chr18_Range | Interaction | chr8_Range  |     1 |\n",
       "| 247 | chr18_Range | Interaction | chr6_Range  |     1 |\n",
       "| 248 | chr19_Range | Interaction | chr19_Range | 69619 |\n",
       "| 249 | chr19_Range | Interaction | chr3_Range  |     1 |\n",
       "| 250 | chr19_Range | Interaction | chr20_Range |     2 |\n",
       "| 251 | chr19_Range | Interaction | chr2_Range  |     3 |\n",
       "| 252 | chr19_Range | Interaction | chr6_Range  |     1 |\n",
       "| 253 | chr19_Range | Interaction | chr7_Range  |     1 |\n",
       "| 254 | chr19_Range | Interaction | chr8_Range  |     1 |\n",
       "| 255 | chr20_Range | Interaction | chr3_Range  |    26 |\n",
       "| 256 | chr20_Range | Interaction | chr9_Range  |    10 |\n",
       "| 257 | chr20_Range | Interaction | chr8_Range  |     6 |\n",
       "| 258 | chr20_Range | Interaction | chr7_Range  |     8 |\n",
       "| 259 | chr20_Range | Interaction | chr6_Range  |     7 |\n",
       "| 260 | chr20_Range | Interaction | chr5_Range  |    11 |\n",
       "| 261 | chr20_Range | Interaction | chr22_Range |     2 |\n",
       "| 262 | chr20_Range | Interaction | chr4_Range  |     5 |\n",
       "| 263 | chr20_Range | Interaction | chr21_Range |     4 |\n",
       "| 264 | chr20_Range | Interaction | chrX_Range  |     6 |\n",
       "| 265 | chr20_Range | Interaction | chr20_Range | 65699 |\n",
       "| 266 | chr21_Range | Interaction | chr8_Range  |     4 |\n",
       "| 267 | chr21_Range | Interaction | chr5_Range  |     2 |\n",
       "| 268 | chr21_Range | Interaction | chr21_Range | 33968 |\n",
       "| 269 | chr21_Range | Interaction | chr4_Range  |     1 |\n",
       "| 270 | chr21_Range | Interaction | chr3_Range  |     3 |\n",
       "| 271 | chr22_Range | Interaction | chr22_Range | 55182 |\n",
       "| 272 | chr22_Range | Interaction | chr7_Range  |     1 |\n",
       "| 273 | chr22_Range | Interaction | chr4_Range  |     1 |\n",
       "| 274 | chrX_Range  | Interaction | chrX_Range  | 20308 |\n",
       "| 275 | chrY_Range  | Interaction | chrY_Range  |   321 |\n",
       "\n"
      ],
      "text/plain": [
       "    A           relationshipType B           count \n",
       "49  TF          Interaction      TF            2488\n",
       "50  chr1_Range  Interaction      chr1_Range  166397\n",
       "51  chr1_Range  Interaction      chr2_Range      12\n",
       "52  chr1_Range  Interaction      chr9_Range       3\n",
       "53  chr1_Range  Interaction      chr5_Range      14\n",
       "54  chr1_Range  Interaction      chr4_Range       6\n",
       "55  chr1_Range  Interaction      chr3_Range      12\n",
       "56  chr1_Range  Interaction      chr17_Range     12\n",
       "57  chr1_Range  Interaction      chr7_Range       8\n",
       "58  chr1_Range  Interaction      chr15_Range      6\n",
       "59  chr1_Range  Interaction      chr16_Range      3\n",
       "60  chr1_Range  Interaction      chr10_Range      5\n",
       "61  chr1_Range  Interaction      chr13_Range      2\n",
       "62  chr1_Range  Interaction      chr8_Range      11\n",
       "63  chr1_Range  Interaction      chr12_Range      9\n",
       "64  chr1_Range  Interaction      chr11_Range      2\n",
       "65  chr1_Range  Interaction      chr14_Range      3\n",
       "66  chr1_Range  Interaction      chr6_Range       4\n",
       "67  chr1_Range  Interaction      chr19_Range      2\n",
       "68  chr1_Range  Interaction      chr20_Range     11\n",
       "69  chr1_Range  Interaction      chrX_Range       4\n",
       "70  chr1_Range  Interaction      chr21_Range      1\n",
       "71  chr2_Range  Interaction      chr2_Range  136934\n",
       "72  chr2_Range  Interaction      chr21_Range      4\n",
       "73  chr2_Range  Interaction      chr11_Range      1\n",
       "74  chr2_Range  Interaction      chr8_Range      13\n",
       "75  chr2_Range  Interaction      chr6_Range       5\n",
       "76  chr2_Range  Interaction      chr20_Range     10\n",
       "77  chr2_Range  Interaction      chr7_Range      11\n",
       "78  chr2_Range  Interaction      chr9_Range       5\n",
       "... ...         ...              ...         ...   \n",
       "246 chr18_Range Interaction      chr8_Range      1 \n",
       "247 chr18_Range Interaction      chr6_Range      1 \n",
       "248 chr19_Range Interaction      chr19_Range 69619 \n",
       "249 chr19_Range Interaction      chr3_Range      1 \n",
       "250 chr19_Range Interaction      chr20_Range     2 \n",
       "251 chr19_Range Interaction      chr2_Range      3 \n",
       "252 chr19_Range Interaction      chr6_Range      1 \n",
       "253 chr19_Range Interaction      chr7_Range      1 \n",
       "254 chr19_Range Interaction      chr8_Range      1 \n",
       "255 chr20_Range Interaction      chr3_Range     26 \n",
       "256 chr20_Range Interaction      chr9_Range     10 \n",
       "257 chr20_Range Interaction      chr8_Range      6 \n",
       "258 chr20_Range Interaction      chr7_Range      8 \n",
       "259 chr20_Range Interaction      chr6_Range      7 \n",
       "260 chr20_Range Interaction      chr5_Range     11 \n",
       "261 chr20_Range Interaction      chr22_Range     2 \n",
       "262 chr20_Range Interaction      chr4_Range      5 \n",
       "263 chr20_Range Interaction      chr21_Range     4 \n",
       "264 chr20_Range Interaction      chrX_Range      6 \n",
       "265 chr20_Range Interaction      chr20_Range 65699 \n",
       "266 chr21_Range Interaction      chr8_Range      4 \n",
       "267 chr21_Range Interaction      chr5_Range      2 \n",
       "268 chr21_Range Interaction      chr21_Range 33968 \n",
       "269 chr21_Range Interaction      chr4_Range      1 \n",
       "270 chr21_Range Interaction      chr3_Range      3 \n",
       "271 chr22_Range Interaction      chr22_Range 55182 \n",
       "272 chr22_Range Interaction      chr7_Range      1 \n",
       "273 chr22_Range Interaction      chr4_Range      1 \n",
       "274 chrX_Range  Interaction      chrX_Range  20308 \n",
       "275 chrY_Range  Interaction      chrY_Range    321 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "InteractionRel <- rel_details[rel_details$relationshipType ==\"Interaction\",]\n",
    "InteractionRel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 id='3'>3. List of hubs for chr12 (a hub is a node with a high degree)<h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "chr12 nodes have Bind and Interaction relationships. \n",
    "\n",
    "At the first, we use cypher command `'MATCH (n:chr12)--(m) with id(n) as chr12,id(m) AS m   MATCH (al) where id(al) = m or  id(al) = chr12 return id(al) as id'` for the first argument of `algo.degree.stream`; then we will difine a subgraph with all relationships during all nodes including chr12 and nodes directerly related to chr12.\n",
    "\n",
    "After using degree algorithm, we need to filter all chr12 nodes by `'MATCH (n:chr12)-[r]-(m) RETURN id(n) AS source, id(m) AS target'` and to remove those 0 degree score of chr12. \n",
    "\n",
    "At the end, we return back chr12 nodes names and its degree score. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"\n",
    "CALL algo.degree.stream(\n",
    "                        'MATCH (n:chr12)--(m) with id(n) as chr12,id(m) AS m   MATCH (al) where id(al) = m or  id(al) = chr12 return id(al) as id',\n",
    "                        'MATCH (n:chr12)-[r]-(m) RETURN id(n) AS source, id(m) AS target',\n",
    "                        {graph:'cypher',direction: 'both'})\n",
    "YIELD nodeId, score\n",
    "with nodeId, score\n",
    "MATCH (chr12:chr12)\n",
    "where id(chr12) = nodeId and score > 0\n",
    "RETURN chr12.Name AS Name, score as degree\n",
    "ORDER By score DESC\n",
    "\"\n",
    "chr12_hubs <- cypher(graph, query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Name</th><th scope=col>degree</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Bin281</td><td>775</td></tr>\n",
       "\t<tr><td>Bin247</td><td>774</td></tr>\n",
       "\t<tr><td>Bin272</td><td>729</td></tr>\n",
       "\t<tr><td>Bin267</td><td>717</td></tr>\n",
       "\t<tr><td>Bin607</td><td>717</td></tr>\n",
       "\t<tr><td>Bin35 </td><td>695</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       " Name & degree\\\\\n",
       " <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Bin281 & 775\\\\\n",
       "\t Bin247 & 774\\\\\n",
       "\t Bin272 & 729\\\\\n",
       "\t Bin267 & 717\\\\\n",
       "\t Bin607 & 717\\\\\n",
       "\t Bin35  & 695\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| Name &lt;chr&gt; | degree &lt;dbl&gt; |\n",
       "|---|---|\n",
       "| Bin281 | 775 |\n",
       "| Bin247 | 774 |\n",
       "| Bin272 | 729 |\n",
       "| Bin267 | 717 |\n",
       "| Bin607 | 717 |\n",
       "| Bin35  | 695 |\n",
       "\n"
      ],
      "text/plain": [
       "  Name   degree\n",
       "1 Bin281 775   \n",
       "2 Bin247 774   \n",
       "3 Bin272 729   \n",
       "4 Bin267 717   \n",
       "5 Bin607 717   \n",
       "6 Bin35  695   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(chr12_hubs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>667</li>\n",
       "\t<li>2</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 667\n",
       "\\item 2\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 667\n",
       "2. 2\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 667   2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(chr12_hubs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(chr12_hubs, file = \"chr12_hubs.csv\",sep=\",\",quote = F, col.names = TRUE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get chr12_hubs.csv at [here](chr12_hubs.csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAOVBMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6et2OaysrK9vb2+vr7Hx8fQ0NDZ2dnh4eHp6enw8PD///9TaHlrAAAACXBI\nWXMAABJ0AAASdAHeZh94AAAZZUlEQVR4nO3diVbiagKFUSqgXm3HvP/DNiEBwyD84AkEs89a\nXVq6hYh8Mmj1ndVm9uvNbn0AZn9hQjILTEhmgQnJLDAhmQUmJLPAhGQWmJDMAhOSWWBCMgts\nAiG9Psxms4fX1evL1256LMfP/zfv7W1Mn3BzAKvNn75ueyAD78+H9FF1X8nqrb799eoKIY3r\nE96EtDyeP13SXw/p4/sLOfu4/fVq+JBG9gn3Qpo93fZIht1fD2m+vFOx/M789br8Pr24/fUq\ndOftyEb2CW8O4HV5k3TbIxl2fzykt+XVqn3tY/78ufqyfj0tr2uPy2/WzV8+5821rf54XF7t\nHt/r9o31y3w2f6k/H5dXxff2g7/f374+f/rcfcfBD3xbLF99+z6e1dWq+eN1sTHvy4c0y4/q\nv3fzl/65de/YOcveCY3xE/4uuXvlszma2cP7/uG/zmfV05rtnc7Y98dDWn6Ft74azV31/t2e\n5Rf1oa6fuvsez61YrP7y0MLmi9x//3vvflPpBz73z78nV+a5fXVRHwqpf27tOw6e5WxT0tg+\n4c2n89IW/rk+tbfdw28/dLH1WT7Xd7M/HtLyS7z1ELe5Xr2vvk6P7Rfyq/5srirVx+pB+sf6\njS/Nde6jebGot98/X/35uv+OAx/4sXrTor0Ors+/PYi31XV+Y5YH9HIopP65rcvaOsvvExrn\nJ9x/jNTc6i5mzbOJL+tPaOty6E56dvB0xr4/HtJs5yFC+62w/lpfZVffvR/br/HL6uFw9/3x\n+8X++zdf3YIPfF9dK556579zEE8r87X6fr0f0mw3wt2z/D6hcX7CvZCetg9r59i6m9LXn05n\n7JteSN8vZ9137+67eO/KvP1i6/3NPfx5+zOakx84X7++2Dr/3duc+uh7N+e2fyxbdJyf8Cak\nxef6LV/vz4uDn2l30odPZ+z74yEduKfz/XL3eth/Y//F1vu7hwyrHy8WfGC3rfM/eBCF53b0\nExjhJ9y+/XW2eej2vtiIn4/twOmMffdzpBftYf+x9/fLi65X9cdj+yV+v0ZIe+d2KqSxfcLd\nu1/WP499W/798fVTSHe21/1ng1d/2/ranXVPp9lb8xTuvPieTm/715njd+3657Z/LAdCGtsn\nvH73vHu80z51cfhy2Ltrd0+7uwM+c1X7Y4qvl2r1vO/B69XuQ+gN6F5svb/b/jsOfGD/iYR6\n/3z3zPq9zTXq/fBN1aGz3LpVG9knvH73e/dptX89cIvUPdnw8tPpjH1/PaTNT0GW++kb9NvO\nk7ob0L3Yev/yLv7r+tngUx/YPovb/9HO/hVo7+nvh9W19L19Hrh/bvvHciikkX3Cm/NsP63m\ntual/jjwZEP79Pd79dPpjH1/PaT2S9NsvrlLUe9cr3Z/zFjvvui//7M7veqz4AO7H7bufF/f\nvv7v/kC2K6G9rvXPbf9YDoU0sk94c57NjdDH5rNb3cPbPvytH8jun87Y9+dDWt5bWH5xqsf2\ne9vh61X7iy8Pbz2x/aL3/nr1GzdV/zdmjnzg96//1PvnuzbNb8N8/4rQ6vv18u/tX3rntn8s\nB0Ma1ye8Fe/D+rN9+2x+QLxz+O2vG/10OmPfBEKye9o9/eyoPyHZKDZrn194uqtfsOtNSDaK\ndY+K7vUGSUg2kr01j766B4t3OCGZBSYks8CEZBaYkMwCE5JZYEIyC0xIZoEJySwwIZkF9tuQ\n/mc24eVC+uXHm93xhGQWmJDMAhOSWWBCMgtMSGaBnRNStdyhl7unZDa5nRFS1f2x+3LvlMwm\nNyGZBSYks8AiIf1rJiSb8C54ssEtktnu3LUzC0xIZoEJySwwIZkF5jcbzALzu3ZmgQnJLDAh\nmQUmJLPAhGQWmJDMAhOSWWBCMgtMSGaBCcksMCGZBSYks8CEZBaYkMwCC4f03+n98mzMxrib\nhCQ3+2sTkllgQjILTEhmgQnJLDAhmQUmJLPAhGQWmJDMAhOSWWBCMgtMSGaBCcksMCGZBSYk\ns8CEZBaYkMwCE5JZYEIyC0xIZoEJySwwIZkFJiSzwIRkFpiQzAITkllgQjILTEhmgQnJLDAh\nmQUmJLPAhGQWmJDMAhOSWWBCMgtMSGaBCcksMCGZBTbikORm9zMhmQUmJLPAhGQWmJDMAhOS\nWWBCMgtMSGaBCcksMCGZBXb3ISWV2aUTUk+ZXToh9ZTZpRNST5ldOiH1lNmlE1JPmV06IfWU\n2aUTUk+ZXToh9ZTZpRNST5ldOiH1lNmlE1JPmV06IfWU2aUTUk+ZXToh9ZTZpRNST5ldOiH1\nlNmlE1JPmV06IZ2tzPYnpLOV2f6EdLYy25+QzlZm+xPS2cpsf0I6W5ntT0hnK7P9CelsZbY/\nIZ2tzPYnpIGUTWtCGkjZtCakgZRNa0IaSNm0JqSBlE1rQhpI2bQmpIGUTWtCGkjZtCakgZRN\na0IaSNm0JqSBlE1rQhpI2bQmpIGUTWtCGkjZtCakgZRNa0K6qbK/MiHdgbLxT0h3oGz8E9Id\nKBv/hHQHysY/Id2BsvFPSHegbPwT0p9RdssJaWLKhpmQJqZsmAlpYsqGmZAmpmyYCWliyoaZ\nkCambJgJaWLKhpmQJqZsmAlpYsqGmZAmpmyYCWliyoaZkCambJgJiTqg7NwJiTqg7NwJiTqg\n7NwJiTqg7NwJibpY2feERF2s7HtCoi5W9j0hURcr+56QqIHVNCYkamA1jQmJGlhNY0KiBlbT\nmJCogdU0JiRqYDWNCYm6IzXeCYm6IzXeCYm6IzXeCYn6c+oWExL159QtJiSK+rUSEkUFlJAo\nKqCERFEBJSSKCighUVRACYmiAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKqPNCqpY79FJI\n1MTVWSFV3R+7L4VETV0JiaIC6vyQDgQlJGrq6syQ2sdE+yH9ayYkarLqvJCq2l07ijqgPEai\nqIASEkUFlJAoKqCERFEB5TcbKCqg/K4dRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBC\noqiAEhJFBZSQKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSK\nCighUVRACYmiAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBCoqiA\nEhJFBZSQKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSKCigh\nUVRACYmiAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBCoqiAEhJF\nBZSQKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSKCighUVRA\nCYmiAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBCoqiAEhJFBZSQ\nKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSKCighUVRACYmi\nAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBCoqiAEhJFBZSQKCqg\nhERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSKCighUVRACYmiAkpI\nFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBCoqiAEhJFBZSQKCqghERR\nASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSKCighUVRACYmiAkpIFBVQ\nQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBHQpp1f68qIVHUcfVTSNWsNyFR\n1HH1U0gvvY5ehERRx1XBXbviCYmarPJkA0UF1LGQniqPkSiqSB0J6engkw2rp/Cqqtp6KSRq\n4upISNWhZxlW4VRdUeuXQqKmrs58sqGqhURR++pISA+zr/2OhERRB9SRkD6rxWdpSP+aCYma\nrDp61273yYaqdotEUYfUOSFtuhESRW2rc34gW7UTEkXtqrN/s8EtEkXtq7MeIwmJog6ri0Ly\nmw0Uta1O3rX7XDwfevOhCYmarDr9GOlrVlqSkKjJqoInG/z2N0WdUqdDep35/2ygqBOq5MmG\nJyFR1HF1OqSqtCMhUdNV/qk5RQWUkCgqoI6F9PU0n83mT3v/KklIFLWjjv57pO5B0u6/ShIS\nRe2oIyE9zpp/2Pe5mD0KiaKOq4L/zwY/kKWoU0pIFBVQ7tpRVEB5soGiAsrT3xQVUH4gS1EB\nJSSKCqhjIT2s3jCbe4xEUSfU8f8axeqtnrWjqFPq6H+N4r158eHnSBR1SvmBLEUF1NH/GsXj\nV/Mc+GwhJIo6rkp+IPshJIo6rgp+IFv6pJ2QqOkqP0eiqIASEkUFlJAoKqCERFEBJSSKCigh\nUVRACYmiAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBCoqiAEhJF\nBZSQKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSKCighUVRA\nCYmiAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBCoqiAEhJFBZSQ\nKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSKCighUVRACYmi\nAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBCoqiAEhJFBZSQKCqg\nhERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSKCighUVRACYmiAkpI\nFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBCoqiAEhJFBZSQKCqghERR\nASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSKCighUVRACYmiAkpIFBVQ\nQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBCoqiAEhJFBZSQKCqghERRASUk\nigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSKCighUVRACYmiAkpIFBVQQqKo\ngBISRQWUkCgqoIREUQF1XkjVcodeComauDorpKr7Y/elkKipKyFRVECd/xhJSBS1p0Ih/Wsm\nJGqy6uyQqtotEkXtKiFRVECdG1LV/0NIFPXfJSFVWzUJiaL+uyCkqvdCSBS1Uef9HKnqfpXB\nbzZQ1Jbyu3YUFVBCoqiAEhJFBZSQKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUF\nlJAoKqCERFEBJSSKCighUVRACYmiAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJ\niaICSkgUFVBCoqiAEhJFBZSQKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAo\nKqCERFEBJSSKCighUVRACYmiAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaIC\nSkgUFVBCoqiAEhJFBZSQKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCE\nRFEBJSSKCighUVRACYmiAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgU\nFVBCoqiAEhJFBZSQKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEB\nJSSKCighUVRACYmiAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBC\noqiAEhJFBZSQKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSK\nCighUVRACYmiAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBCoqiA\nEhJFBZSQKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEBJSSKCigh\nUVRACYmiAkpIFBVQQqKogBISRQWUkCgqoIREUQElJIoKKCFRVEAJiaICSkgUFVBCoqiAEhJF\nBZSQKCqghERRASUkigooIVFUQAmJogJKSBQVUEKiqIASEkUFlJAoKqCERFEB9buQquWERFG/\nC6na/CEkatpKSBQVUEKiqIAKhfSv2f+Oa7O/vPAtktk0JySzwIRkFpiQzAITkllg4d9sMJvm\nwr9rZzbNCcksMCGZBSYks8CEZBaYkMwCE5JZYEIyC0xIZoEJySwwIZkFJiSzwIRkFpiQzAIT\nkllgQjILTEhmgQnJLDAhmQUmJLPAhGQWmJDMAhOSWWBCMgtMSGaBBUM6vH8/vJ2i/pSKhfTD\n/lHUpJSQKCqghERRASUkigqogUIym9aEZBaYkMwCE5JZYDcLqf/fzzzKOniMl51Wodo616Mn\nVp1kRUffqgmc1unLqz59QsXHVX5aJ4+r5Dpxu5CKrtjdp1CfCKnwtErP8ftcT6oTx1UXHf3W\nH3/3tHZeO6Cqkut14XElT6uuT14nBgip/FM4/U2l2rzInFah6p/3L1VddPSFV9j7Pq2d135k\np+4MFB3XGafV+2r+rE6d3BC3SOV3oU7eYFablyVfzKLTKgl867x/p9qXsSvsPZ/WzmvHz/O3\nx3XGaY02pKKSimr7VkW5Rc6xqk5eaL0vUsE3xbr46IvUdU+rOKTEcXWnUnTlT57WyZAKrhM3\nfIxUxr6/4/32tArPsb3cTj2eKnvsXHT0a3XyHK9/WmVPNpSeVtmTDcdv2nrnGDqtkq92wXXC\n0982shU+n3v10zo+IZkFNsiTDbnn+Yt/2rS5kT5xXIU/+wn9lCK55NGXq+uf43VVZ0+9//T1\na4inv7f/+EkVPkl+1s9+As8cJY8+uejRl6vrn+N1VfmPar5P9eBuFlLBNb97d/JnP7uv/aBS\nR1/2ZSr+Yu6+9oMSUrE641p4At4upBKyfnfgO0Y4pKKjTz6dKKQhVPm1cMQhnfczg6JzLPjm\nc+pUy1RddPTJkkYd0t0eV11+LTxx/brtkw0nn+cvuyIW/1Sn7MFz/8Xxgys8vMjG+2RD/8V1\njqvsHEvUipy+Z3H6grjx09/Xe55/iN330dt6ia+jnyOZBTbYXbvVK2NTdcFPm+rNLfm1zjGo\nqq1jO3JaY728SlTZOV75uAZ7suHUY//bqKIvU8ENffIcB1EFv446zsvrtCo7x2sf11AhrV4U\nPKtyZRW9YoTO8TaqHunldVqVneO1j2u4kE48GXIbFb5iRM7xZiGN8vI6rcrOsUzljmvAkAr/\nhcp1VcH9hpKvQPIco6rgX86M/PI6qcrOsUzljmuYJxt2D2AsqughapFLnmNS1af/5Uw95sur\nRCUvsdhxefrbLDAhmQU2UEix5/mvrqriuzTUSlz98io7x2sf1zAhxZ5VuYkq21iPfgqXV3Kh\n4xLSpRvr0U/h8kpOSAOpav2bMVc7x/tW17+8ys7xysc11GOku1WNaC98JY3z8io7x2sf10C3\nSLHn+a+uqu0/RnNcY1XXv7zKzvHax+Xp750VfwFstetfXuGQQhPSzoR03oTUbpjftSu7gR6r\nKvsyjfXop3B5xULKHZdbpJ25RTpvbpHaCWln1fdufSh3setfXmXneO3jGuifmtep5/mvr8o2\n1qOfwuWVXOy4Bvr3SKnn+a+vyjbWo5/C5ZVc7rhu9//9PVJVtrEe/RQur+RyxyWkizbWo5/C\n5ZWckAZTZRvr0U/h8kpOSIOpso316KdweSUnpMFU2cZ69FO4vJIbd0jB5/mvr8o21qOfwuWV\nXO64/EDWLDAhmQUmJLPAhGQWmJDMAhOSWWBCMgtMSGaBCWlUm3V7eN95c/9vL9Xemw6e1OGP\ntkHmMh7VZpu9b795729CGtdcxqNad53/eprND7z54N9OnJRdZy7tUW1z7V+98vU4mz1+rf/2\n/jCbVU/trVb3ps8GfK7A58PqnXX9XM3mL+3HPLVv6vjDbPF5i89pGhPSqLZ9i1Q1zcy7N7+1\nd/meeiF9rUD11fytat+5jKfZS/Omh+5NLX/sqA0yIY1qW4+RntswXtoU5rPXuv7Y3Bg1fzzN\nFnW9aFtZfNUvs2p121S/t68s3/Tcvtb9bUVtkAlpVFtn9PhRN+ms3vSwvqH6fHtebIU0X0ZT\nfza3WU0+7duq2eNbd1LrN7X/++ioDTIhjWqrRt6aG5r6u6ruzYutv21e2XntbXkfb75OqB/S\n5uRtiLlkR7X2mv7UlrQd0uNs/vL2eTKk5d2/+ax6F9KV55Id1bpr+nz2XK/v2q3f3D6P9+Nd\nu95HLx8s7VU2a+niup/OhCakUa1L4WP1kOapeW7gtbnytym811/bj5H6TzasP7paso/1Uwz9\nkBbNRz/f7lP74xPSqLa+TXlubmbaZ7ebpNpqvu/oVftPf68/umXPB0KarT7OhpmQRrXNg5jV\nnbvm562L9/WbV39pXnvZ3OD0fiC7+einalY91/shfS5aaoNMSNOYpxkGnst3GhPSwHP5TmNC\nGngu32lMSAPP5WsWmJDMAhOSWWBCMgtMSGaBCcksMCGZBSYks8D+D6/9w9u7lFm4AAAAAElF\nTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "chr12_hubs_20 <- chr12_hubs[1:20,]\n",
    "chr12_hubs_20$Name <- factor(chr12_hubs_20$Name, levels = chr12_hubs_20$Name)\n",
    "ggplot() + \n",
    "    geom_bar(aes(x = chr12_hubs_20$Name, y = chr12_hubs_20$degree),\n",
    "           stat=\"identity\", fill = \"lightblue\")+ \n",
    "  labs(x=\"Relationship\",y = \"count\")+\n",
    "  ggtitle(\"Chromosome Inclusion Chromosome Range\") + backgroup   #设置背景为白色"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 id='4'>4. Degree distribution for chr12 <h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "distribution <- as.data.frame(table(chr12_hubs$degree))\n",
    "library(ggplot2)\n",
    "#ggplot(distribution, aes(x=Freq))+ geom_histogram(binwidth=.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAPFBMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6et2OaysrK9vb2+vr7Hx8fQ0NDZ2dnh4eHp6enw8PD/AAD///8SiDE6AAAA\nCXBIWXMAABJ0AAASdAHeZh94AAAduklEQVR4nO3dYUPayhZGYQ5atb3Hqsf//1+vgJIBQSDZ\nQ/Ywz/pQkBeGN8wsCSFtF+8AJrOYuwBwCxAJCIBIQABEAgIgEhAAkYAAiAQEQCQgACIBARAJ\nCOAmRVpsmeXpH5eLxcO2yZE7HU0OBBEbcs4Yc71gJb+2HX7fLxZ3f2YtcwHzv3IVmFekh9UT\nE2kcz18d3pabGbyfudC5zP7K1WBekT6e9rW4fvxeZwf9iPS8nbSPX0ePH2/ui0Uj70lzv3JV\nmHdBlM8+QqS7xfLs+wYzt0ivv4bffr/uFptKd7NWOpsORHr+2Ne+f/68/fVuvbPw/LFcHzf3\n+7rz5+XOvd//Xf30d53//fW5y/41t3fFs7w8rD4X/X3fvhsOTd4eP+748PK+90yrPz4+TN39\nOzz+7nH1TvZr8Wvo/u/Qs6y2ExRbVdyhvL7zpMUmve9u2Le2n8MOG7ce4M/d6u6vD9thirzc\nkNMPPNDxfk/mud0+m0ZqXsbOq/+4Wdm/N7d/LJNfX7fdHxBp9973m59W0/57u8v+sNl1e91+\nENo+avWwbyJ97uu/fBdpM/rHDsz7369d0ZfVEz3udb//Vm1/A8qtKu+wuX7gSReDScOGfWu7\nM+zua/JruR2mzMsNOfeBZcdfe+b89Y40J1+zudq4l48ZfHu//1obH9dfV7ctX95flt9F2rv3\n8nntzf1XsN5lf97M/ccC/Ppl+nc7XvnG8zno8u96yTx8F+nzMX/Xb24fj/x3/6P137JnWW1/\nA762qrzD9vr77pNuN2lDsWHf2q6H3dm4zY1/VpK9rC7u9za+3JBTD/zecfXutCvSR/pv0KKo\nzM2L9LBeqS/r3/yLz7X/sLl8/i7SgXu/rYPHdfC2/gW5XP+WLD7LPGw+Ev/ZvL3siTSMsS/S\nZ4fNot2upoKdnmW1z+DfcqT9O2yvv3970reiYrlh+22/hh027vO9ZLj4nr8U7U89cK/jtmdR\nbidMzM2L9PlJZv078OOWt9UPy83lgeW9d+9vwZrH1b7dazHHw3h3799EGi4P/fDxmOV69K9P\nSyXL7X32qt3tbcB6q/busL1eNNlpsOHuh7Z7L9bd9h7lxU5ebsipB37vuFdt9UI3smN3uyKV\n17dW7a+jg8v78L3LIf+u9u1+F58zdse9SKTN5edHi8e3w9txXrW9Owy/S4486cHX6vh9yxvL\ni6MbcsYD9zru9VntCL6+NwKRLhdpvW93V/yynCzS+8vD5lnLo2ktilRuyFSRVnu0jXxAeu9A\npEO7Lp97RgcWwtEdnTJY7XI8l0ebRu7aFY/54Hl14Hl3T2ZnD+5ud6O+75veHdnqoskBkX7a\ntfu+cQd8WB7bkDN37Q71XPGx7zwcFs3PzYu092H667adz+qr+f67/uHQvb8F6zku9zp+PNgw\nXO4802LzmOfy4/TeL+fdnmWDz4MNf3ZWfHmHnbo7Dfaf6OgWf/2wf8ygePz6Ynfjiyc79cDv\nHXeq3bdzetCKmxdpc5B1s/K+bt8cPX7eHD3+tZ7ov+Ux5t17F8HXGSvrz9TD8z3/dPh7uNx5\npsX6UPPnYzZHeb8d/n75fvh7U20T/N09fr93h+31oskBkfYOf3+/7/PeUezi8euLnbzckFMP\n/N6xrPa8aOgD0nsHIn194bjzXrG5bfN95ufn489vPQ/ceye43z6+3H0vv3o8LtLOMy3KL2Rf\nP78HXe4tnZ0vZItqB79R3rlDeX1ocECkvS9kv993/3vV9/2LMt/ZkFMP/NaxrPb5WWtnKjPT\nSM3L2H31d06B+bxtdYbN76/Px6udiD9f4fd7fwWbu61ZfTjZOcK2Phnm1/O3Z98do3ym1R/D\nKULrM3OWj99+Be+cIjRUe/883WZ3xe/coby+bXBIpGLDDotUbtwhH3Y2fmdDTj1wv2PZYUmk\ndhg9S29J9t+T1ACRxj3wYeYDs4vN5/TH8tAhZoVIox428zfuv792e7whZYFIl7NcLB/eTt+t\nJs+rTx/bj2yYn65FAqIgEhAAkYAAiAQEQCQgACIBARAJCIBIQABEAgIYK9L/APxvukgjHwfc\nEEQCAiASEACRqvPff//NXQHVIVJ1iNQDRKoOkXqASNUhUg8QqTpE6gEiVYdIPUCk6hCpB4hU\nHSL1AJGqQ6QeIFJ1iNQDRKoOkXqASNUhUg8QCQiASEAARAICIBIQAJGAAIgEBEAkIAAiAQEQ\nCQiASEAARAICIBIQAJGq46TVHrhEpOUH5eX+CDgIkXrgApGWn38stz/sjYCDEKkHiFQdIvXA\npSK9E+lSiNQDF4m0+Ww0iPTPikQiLX5gvlZE6oFLRPq0KO870uLpKERCVW7qMxKRMBdEqg6R\neoBI1SFSDxCpOkTqgZs6s4FImIubOtcup0joASIBARAJCIBIQABEAgIgEhAAkYAAiAQEQCQg\nACIBARAJCIBIQABEqo6TVnuASNUhUg8QqTpE6gEiVYdIPUCk6hCpB4hUHSL1AJGqQ6QeIFJ1\niNQDRKoOkXqASNUhUg8QqTpE6gEiVYdIPUAkIAAiAQEQCQiASEAARAICIBIQAJGAAIgEBEAk\nIAAiAQEQCQiASEAARKqOk1Z7gEjVIVIPEKk6ROoBIlWHSD1ApOoQqQeIVB0i9QCRqkOkHiBS\ndYjUA0SqDpF6gEjVIVIPEKk6ROoBIlWHSD1AJCAAIgEBEAkIgEhAAEQCAiASEACRgACIBARA\nJCAAIgEBEAkIgEhAAESqjpNWe4BI1SFSDxCpOkTqASJVh0g9QKTqEKkHiFQdIvUAkapDpB4g\nUnWI1ANEqg6ReoBI1SFSDxCpOkTqASJVh0g9QCQgACIBARAJCIBIQABEAgIgEhAAkYAAiAQE\nQCQgACIBARAJCIBIQABEqo6TVnuASNUhUg8QqTpE6gEiVYdIPUCk6hCpB4hUHSL1AJGqQ6Qe\nIFJ1iNQDRKoOkXqASNUhUg8QqTpE6gEiVYdIPUAkIAAiAQEQCQiASEAARAICIBIQAJGAAIgE\nBEAkIIAGRVoch0iYiRZFOm4LkTATRAICIFJ1nLTaA0SqDpF6gEjVIVIPEKk6ROoBIlWHSD1A\npOoQqQeIVB0i9QCRqkOkHiBSdYjUA0SqDpF6gEjVIVIPEKk6ROoBIgEBEAkIgEhAAEQCAiAS\nEACRgACIBARAJCAAIgEBEAkIgEhAAEQCAiBSdZy02gNEqg6ReoBI1SFSDxCpOkTqASJVh0g9\nQKTqEKkHiFQdIvUAkapDpB4gUnWI1ANEqg6ReoBI1SFSDxCpOkTqASIBARAJCIBIQABEAgIg\nEhAAkYAAiAQEQCQgACIBARAJCIBIQABEAgIgUnWctNoDRKoOkXqASNUhUg8QqTpE6gEiVYdI\nPZBWpMVx2hTphw2arxuiyCvSCF1yi5SxG6IgUnWI1ANEqg6ReoBI1SFSDxCpOkTqASJVh0g9\nQKRrQaSbhkjXgkg3DZGuBZFuGiJdCyLdNEQKrj3mzKYrdUNFiJSg9pW6oSJESlD7St1QESIl\nqH2lbqgIkRLUvlI3VIRICWpfqRsqQqQEta/UDRUhUoLaV+qGihCpeu3//vuPSDcPkarXJlIP\nEKl6bSL1AJGq1yZSDxCpem0i9QCRqtcmUg8QqXptIvUAkarXJlIPEKl6bSL1AJGq1yZSDxCp\nem0i9QCRqtcmUg8QKUHtK3VDRYiUoPaVuqEiRIr9D8CI1ClEil3gROoUIhEJARCJSAiASERC\nAEQiEgIgEpEQAJGIhACIRCQEQCQiIQAiVRfJSas9QCQiIQAiEQkBEIlICIBIREIARCISAiAS\nkRAAkYiEAC4Uabn+44NDI4RCJLTEZSKtBdrIdGCEUIiElrhIpOU7kS6vTaQeuESk5TuRRtQm\nUg9MEumfFUSaXHv8q4QsXCDS8t07Up3a418lZOF8kbb+ECm69vhXCVm4QKQNRIqvPf5VQhZG\nfI9EpOja418lZIFIREIAzmwgEgJwrh2REACRiIQAiEQkBEAkIiEAIhEJARApwUmrof9nIGaB\nSAlECm2AWSASkRAAkYiEAIhEJARAJCIhACIRCQEQiUgIgEhEQgBEIhICIBKREACRiIQAiJTg\npNXQBpgFIhEJARCJSAiASERCAEQiEgIgEpEQAJGIhACIRCQEQCQiIQAiEQkBdCPScUKXMZE6\npRuRrrSMidQpRHLSKgIgEpEQAJGIhACIRCQEQCQiIQAiEQkBEIlICIBIREIARCISAiDSzYnk\n/y2bAyLdnkixw+EsiEQkBECkmztplUhzQCQiIQAiEQkBEIlICIBIREIARCISAiASkRAAkYiE\nAIg0Zt2N+jeJQhuMe+VGzwZOQaRRIsU+z4gG41650bOBUxCJSAiASERCAESqLpKTVnuASERC\nAEQiEgIgEpEQAJGIhACIRCQEQCQiIQAiEQkBEIlICIBIREIARCISAiASkRAAkaqLVKXBuFdu\n9GzgFEQiEgIgEpEQAJGIhACIRCQEQCQiIQAiEQkBEIlICIBIREIARCISAiASkRAAkYiEAIhU\nXSQnrfYAkYiEAIhEJARAJCIhACIRCQEQiUgIgEiNijTqPw2sME/YcFSkxefPy+XZI4RCpB8b\njHt9KswTNhwWaVn+gjt7hFCI9GMDIiXjsEh/Co/+nD1CKET6sQGRknFy1+6CEUIh0o8NiJQM\nBxuIhACOi/S49BnpOuWIdAMcFenRwQYi4WyOirQ8eZTh2wihEOnHBkRKhoMNREIAR0X6tXi7\ndIRQiPRjAyIl46hIr8v71wtHCIVIq2TUiUDHkzpThfcfd+0cbLhSuQQbhKkQiUgIwBeyREIA\nRCISArBrRyQEQCQiIYATu3av97/PHyGU1OvuotGmnLR6NCFSMk59RnpbnDKJSCdGI1IPnDzY\nYNduajki9cApkf5d+DcbJpYjUg+cPtjwePYIoaRed0TCHqdEWp7yiEinRiNSD/hClkgIgEhE\nQgDHRXp7vFss7h5P/q0kIp0YjUg98MPfR/r8kHTqbyUR6cRoROqBoyI9LFZ/se/1fvFw9gih\npF53RMIeJ//NBl/ITi1HpB4gUnWRqiRESoZdOyIhAAcbiIQAHP4mEgLwhSyREACRiIQAjov0\na33D4s5npNrlEmwQpvLT/0axvtVRu+rlEmwQpvLD/0bxd3Xx4nuk6uUSbBCm4gtZIiGAH/43\nioe31THwxf3ZI4SSet0RCXuc/kL25ewRQkm97oiEPU5+IXvy/3Yh0onRnLTaA75HIhICIBKR\nEACRiIQAiEQkBEAkIiEAIhEJARCJSAiASERCAEQiEgIgEpEQAJGIhACIVF2kKgmRkkGkH5Lj\nxD7PiIRIySBSmwmRkkGkNhMiJYNIbSZESgaR2kyIlAwitZkQKRlEajMhUjKI1GZCpGQQqc2E\nSMkgUpsJkZJBpDYTIiWDSNUTJ632AJF6Euk4dSaxI4jUk0jHkzqT2BFEItITkaZDJCI9EWk6\nRCLSE5GmQyQiPRFpOkQi0hORpkMkIj0RaTpEItITkaZDJCI9EWk6RCLSE5GmQ6Q2EyIlg0ht\nJkRKBpHaTIiUDCK1mRApGURqMyFSMojUZkKkZBCpzYRIySBSmwmRkkGkNhMiJYNIbSZESgaR\n2kyIlAwitZkQKRlEqp44abUHiESkJyJNh0hEeiLSdIhEpCciTYdIRHoi0nSIRKQnIk2HSER6\nItJ0iESkJyJNh0hEeiLSdIhEpCciTYdIRHoi0nSIRKQnIk2HSG0mREoGkdpMiJQMIrWZECkZ\nRGozIVIyiNRmQqRkEKnNhEjJIFKbCZGSQaQ2EyIlg0htJkRKBpHaTK4m0uI4dSa+UYjUZnI9\nkUY8pkeI1GZCpGQQqXrS9kmrRDoPIhHpiUjTIRKRnog0HSIR6YlI0yESkZ6INB0iEemJSNMh\nEpGeiDQdIhHpiUjTIRKRVslxjj+mzsQ3CpGINDapM/GNQiQijU3qTHyjEIlIY5M6E98oRGoz\nSVCBSCVEajNJUIFIJURqM0lQgUglRGozSVCBSCWXiLT8oLzcHyGU21t3oUmCCkQquUCk5ecf\ny+0PeyOEcnvrLjRJUIFIJURqM0lQgUgll35GIlKOJEEFIpVMEumfFUSaI0lQgUglF4q0fPeO\nlCJJUIFIJURqM0lQgUgll4m03P1jb4RQbm/dhSYJKhCp5CKRlsOfRDo7cdJqD1z0hWxxQaSz\nEyL1wCXfIy0/T2lwZgORnoi0i3PtiDQ2qTPxjUIkIo1N6kx8oxCJSGOTOhPfKEQi0tikzsQ3\nCpGINDapM/GNQiQijU3qTHyjEIlIY5M6E98oRCLS2KTOxDcKkYg0Nqkz8Y1CpDaTBBWIVEKk\nNpMEFYhUQqQ2kwQViFRCpDaTBBWIVEKkNpMEFYhUQqQ2kwQViFRCpDaTBBWIVEKkNpMEFYhU\nQqQ2kwQViFRCpDaTBBWIVEKkNpMEFYhUQqQ2kwQViFRCpOqJk1Z7gEhEGpvUmfhGIRKRxiZ1\nJr5RiESksUmdiW8UIhFpbFJn4huFSEQam9SZ+EYhEpHGJnUmvlGIRKSxSZ2JbxQiEWlsUmfi\nG4VIRBqb1Jn4RiESkcYmdSa+UYhEpLFJnYlvFCK1mSSoQKQSIrWZJKhApBIitZkkqECkEiK1\nmSSoQKQSIrWZJKhApBIitZkkqECkEiK1mSSoQKQSIrWZJKhApBIitZkkqECkEiK1mSSoQKQS\nIrWZJKhApBIitZkkqECkEiJVT5y02gNEItLYpM7ENwqRiDQ2qTPxjUIkIo1N6kx8oxCJSGOT\nOhPfKEQi0tikzsQ3CpGINDapM/GNQiQijU3qTHyjEIlIY5M6E98oRCLS2KTOxDcKkYg0Nqkz\n8Y1CJCKNTepMfKMQqc0kQQUilRCpzSRBBSKVEKnNJEEFIpUQqc0kQQUilRCpzSRBBSKVEKnN\nJEEFIpUQqc0kQQUilRCpzSRBBSKVEKnNJEEFIpUQqc0kQQUilRCpzSRBBSKVEKnNJEEFIpUQ\nqXripNUeIBKRxiZ1Jr5RiESksUmdiW8UIhFpbFJn4huFSEQam9SZ+EYhEpHGJnUmvlGIRKSx\nSZ2JbxQiEWlsUmfiG4VIRBqb1Jn4RiESkcYmdSa+UYhEpLFJnYlvFCIRaWxSZ+IbhUhtJgkq\nEKmESG0mCSosxlBnsSSASG0mCSp4FyshUptJggpEKiFSm0mCCkQqIVKbSYIKRCohUptJggpE\nKiFSm0mCCkQqIVKbSYIKRCohUptJggpEKiFSm0mCCkQqIVKbSYIKRCq5ikhjThm5nXV3syet\njklu9uSh64h0/IUd85gRCZHSJyOXXxaIRKQcycjllwUiESlHMnL5ZYFIRMqRjFx+WSASkXIk\nI5dfFohEpBzJyOWXBSIRKUcycvllgUhEypGMXH5ZIBKRciQjl18WiESkHMnI5ZcFIhEpRzJy\n+WWBSG0mCSoQqYRIbSYJKhCphEhtJgkqEKmESG0mCSoQqYRIbSYJKhCphEhtJgkqEKmESG0m\nCSoQqYRIbSYJKhCphEhtJgkqEKmESG0mCSoQqWRukY4TOk0ZVkpokqBCcNL4PzA0t0hXSjKs\nlNAkQYXrberIlXlViFQ9cdLqtIRIw9AJJmO+hEjTEiINQyeYjPkSIk1LiDQMnWAy5kuINC0h\n0jB0gsmYLyHStIRIw9AJJmO+hEjTEiINQyeYjPkSIk1LiDQMnWAy5kuINC0h0jB0gsmYLyHS\ntIRIw9AJJmO+hEjTEiINQyeYjPkSIk1LiDQMnWAy5kuINC0h0jB0gsm4sSRBBSKVEKnNJEEF\nIpUQqc0kQQUilRCpzSRBBSKVEKnNJEEFIpUQqc0kQQUilRCpzSRBBSKVEKnNJEEFIpUQqc0k\nQQUilRCpzSRBBSKVEKnNJEEFIpUQqc0kQQUilRCpeuKk1WkJkYahE0zGfAmRpiVEGoZOMBnz\nJUSalhBpGDrBZMyXEGlaQqRh6ASTMV9CpGkJkYahE0zGfAmRpiVEGoZOMBnzJUSalhBpGDrB\nZMyXEGlaQqRh6ASTMV9CpGkJkYahE0zGfAmRpiVEGoZOMBnzJUSalhBpGDrBZMyXEGlaQqRh\n6ASTcWNJggpEKiFSm0mCCkQqIVKbSYIKRCohUptJggpEKiFSm0mCCkQqIVKbSYIKRCohUptJ\nggpEKiFSm0mCCkQqIVKbSYIKRCohUptJggpEKiFSm0mCCkQqIVKbSYIKRCohUvXESavTEiIN\nQyeYjPkSIk1LiDQMnWAy5kuINC0h0jB0gsmYLyHStIRIw9AJJmO+hEjTEiINQyeYjPkSIk1L\niDQMnWAy5kuINC15Whxn5JqtAJGIlDz56UEj12wFiESk5AmRhqGv9JLnXHdEmpYQaRj6Si95\nznVHpGkJkYahr/SS51x3RJqWEGkY+koveUfrLkGFDJs6cs1WgEhtJgkqZNjUkWu2AkRqM0lQ\nIcOmjlyzFSBSm0mCChk2deSarQCR2kwSVMiwqSPXbAWI1GaSoEKGTT2+5K59WhGR2kwSVMiw\nqWOW3Mh1fgIitZkkqJBhU8csuZHr/AREajNJUCHDpo5ZciPX+QmI1GaSoEKGTR2z5Eau8xMQ\nqc0kQYUMmzpmyY1c5ycgUptJggoZNnXMkhu5zk9ApOqJk1anJUQahg59YVtbd0SalhBpGDr0\nhW1t3RFpWkKkYejQF7a1dUekaQmRhqFDX9jW1h2RpiU/Peg4xx9zfJlOOa2ISERKnkQ/0Zhl\neokGRKqUEGlaQqRzGl4pIVK7CZHOaXilhEjtJkQ6p+GVEiK1mxDpnIZXSojUbkKkcxpeKSFS\nuwmRzml4pSTBerBBIxMindPwSkmC9WCDRiZEOqfhlZIE68EGjUyIdE7DKyUJ1oMNGpkQ6ZyG\nV0oSrAcbNDIh0jkNr5QkWA82aGRCpHMaXilJsB5s0MiESOc0vFKSYD3YoJEJkc5peKUkwXqw\nQSMTIp3T8EpJgvVgg0YmRDqn4ZWSBOvBBo1Mblek5QeHRhjT8EpJgvVgg0YmNyvScvvH3ghj\nGl4pmXM9OGl1WkKkcxpeKSFSuwmRzml4pYRI7SYdiPTPinNE6poPkeaugIlUFmlvBBzkQ6S5\nK6A6RKoOkXqASNUhUg8QqTpE6gEiVYdIPXCVMxv6hkg9cJVz7fqGSD1ApOoQqQeIBARAJCAA\nIgEBEAkIgEhAAEQCAiASEACRgACIBARAJCAAIgEBEAkIgEjVcdJqDxCpOkTqASJVh0g9QKTq\nEKkHiFQdIvUAkapDpB4gUnWI1ANEqg6ReoBI1SFSDxCpOkTqASJVh0g9QKTqEKkHAkS6kH8u\nfcCsaFuP22o7WaRL+edKzxODtvW40bZEOoS29bjRtkQ6hLb1uNG2RDqEtvW40bbXEgm4aYgE\nBEAkIAAiAQFcRaTlF9d4ssloW4/bbXsNkZYHruVF23rccFsi7aNtPW64LZH20bYeN9yWSPto\nW48bbutgwze0rcfttnX4GwiASEAARPrG7e5+zM/ttiXSPjf8gXh2brgtkfa54cmenRtuS6R9\nbniyZ+eG2xJpnxue7Nm54bZE+sbtfiCen9ttSyQgACIBARDpG9v38jb2P9a0UnX1yra1a/du\n124k61dt/dq1MNvLZUsfO4ZXtZW2xR8/Q6R9ltuLFia7oXX5TqSuWG4vW5js82c6A0TqiO2L\n1sTO0vvn5465S5zHdlU20ZdI01gO70mN0Mqb5/CRbu4e5+FgQ2+0sjJvGCIBARAJCIBIQABE\nAgIgEhAAkYAAiAQEQKQmWJin5JigJiBSdkxQExApOyYoO6/3i18bkd4eFouHt8/b7p5Xty0W\nL8v7Mhmu4aoQKTlvy8Vi8Wst0ura4u7rtsVGpPvFQ5EU13BViJScx8X9+9v9Sprfi8fVj39W\n175uW6xu20m+ruG6ECk5d4vXj125lTR367n62M8rblusru0kX9dwXYiUnM1hho00ww7dcNvm\n2pB8XcN18ZInh0ht4CVPzv6u3e5tG2XKZI6OIFJ6fi/u397XBxYeVwcS/l3cbw5AvN8PIpXJ\n1zVcFyIlZzj8/XnQ+2X38Pf2PmXyMnfr/iBSdl5/fX0h+/qwWNz/Xd92v7j7dxCpTLbXcFWI\n1CwL/+JJIojUIIvF39XHoYe5e2CASA3yuPmI9Dp3DwwQqUX+3C0WDzzKBJGAAIgEBEAkIAAi\nAQEQCQiASEAARAICIBIQwP8BMO0ScOM58r4AAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ggplot(chr12_hubs, aes(x=degree))+ \n",
    "  geom_histogram(binwidth=30,colour=\"black\", fill = \"lightblue\")+ #distribution plot\n",
    "  geom_vline(aes(xintercept=mean(degree, na.rm=T)), color=\"red\", linetype=\"dashed\", size=1)+   #mean\n",
    "  ggtitle(\"Ferquency of hubs' degree in chromosome 12\") + backgroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAPFBMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6et2OaysrK9vb2+vr7Hx8fQ0NDZ2dnh4eHp6enw8PD/AAD///8SiDE6AAAA\nCXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO2diXajuhJFuc7U6deZ8P//6wt4EoNAQFFS\nFXuvdW+cwJHKSLsNGOPqDACbqXIXAOABRAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAA\nRAIQwK9I1YXT68ft1/4ar73VR1fqrj21Qoy3U1W9Bp3Eqk1esKaINW1I9LOVl3sN789V9fQ3\nazFT5N9Ue1Hdebv+2l3+eap6q09NnevaK+bWa1MCIq3j41bDz+kyls+ZC4qSfVPtxkOk6v08\nMit6f5gR6bpkxdz6jXxH+oxXM7ngOCI1Hl1qeG3/PXyrqlJfk3Jvqv24DsFX84rwE18++7e5\nJYl1rG3/qTrJ1bKM3CJ9v1R3kV6erv+QPWUtKY57kdrd7Pfbr1+vv7sIT2/ftxesywvG0+8e\nw/0V6feQ5ulf2EDzM1j7fGvm9fO29N/v7vvzZ9D3Y3lVPSZD+9vP29Pvoq9e+92egzJ/q395\ntPvvV6u3W/Cj6fVjsOD2hLorhI87nQ5q/3y5HYv0q70223vyf5+a1b9f780Ey8MnMh8cqfF5\nbL+hSEqtazv3bf5ZNXOx/fXztrP3Farx1Kxw++X5flgVF+mtu8t4iVSP2RgsH4h0uhUwEOnR\nc1Dm+f1yiBe0+9wp4r2/4PaERla4PB7pNKj9/fKH55FqO812n/zL6d5MuDx8IqnBsMaXnjmf\nvCLp05m+1/89tUP6r50n1f2o5/nn/H1f6fR1/jq1I9uf6PfHn/e1LhPs9NHuw9+PgzvL+7t2\np892yrwO23/0HJYZ8Gj3N/DV1v3crvkVLrg9od4K98fdbdKv/bLm5VikX23bbP/J//7xbyPZ\nV/PjuffkwycyFxzW2Lw6dUX6Xfpv47TYi2OJVPWnUvvjI1yp+eVjdKLfH79eDnn/ti8f18hP\nMOKd5X2RHiv32+/0fC8z4PW+zqWIz3b2vd0X/Atb6q9wf3wedBrW/tau+dP+y9+v9tZs58lf\n918/YxvnK6h+LtirsTuKl+I6C0viWCI9Nf8E/ussr66nIsK5/Tt1ThMinS6R+2zrddddPnqy\nYVSkR89hmQGn+zqXp3IJP7ePf87BjG+fUG+F++P+NumW+DRR7c/gyYVbZWzjhE9kLjissVda\n41GpO3ZHE+m6z/72c+6PZXxuD0WKr9Xt+fJzkUjDMkee0S1wJd5SuML9cX+bjNc+Xe3YVpl7\nIgnBXo29epodwe9zqRxApK/HyYbLyfAqPAQqUaSwzJFnZEekke29WqRmj7bUA6TzIUR6DU5/\n//LRnNEd7F109l9GdtrCtXfatQsyQZkBnT24cB+sv2t3+2O4wtjWGal9atdu+ORGfDjFnkji\nrt1YnQ3f1f36kCJxL9LP+2WODuZzRKTmgPjjdkDcjPznYO3+YXPYXcPkyYZOZ532Hz331r7x\nFp5TCM8KXE82/O3M+HCF8PFwE4Qd9U42DNcde/LRjRN2Nhcc1tgp7bngy4MaXIt0436J0OX0\n6eV07GlcpNPn7RTtSzvkn6fbP7T3tT56J3LPYTvn3vK4SJ32Oz2HZQZ8DU9/XxS6LHi0dFs7\nXOH+OKhkpPbe6e/humNPvoptnPCJzAWHNYalNTt25R4gnY8h0uOi1e/rG4yn78vFpCMnuQdv\niz7f/j19rN17a/GRvxIuj4vUab/Tc1hmSOcN2ffgyfXekL2sHa4QPu483UHt3Tdkh+uOPfnO\nj3B554nMBQc1hqW93kczOuB5KbWu7Vy3e/djFO0lL6fLJSsvwUmIYLYEF+o0uxN/r2t01m4v\ndnl5vP907k7GcHlcpE773Z7DMkM6lwg9ruU5Xy+36c74zgrh487THdT+eamqX+1jneGT7/4I\nn3znicwF+zWGNZwQCdQo/DDCNYjkgOpynP4WXqgGuiCSA95vuz28IGUDkTzw0Rx9XI9sIAuI\nBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCCApEj/AzgY+4gk2BaAARAJQABEAhAA\nkXSp6zp3CbAHiKQLIjkFkXRBJKcgki6I5BRE0gWRnIJIuiCSUxBJF0RyCiLpgkhOQSRdEMkp\niKQLIjkFkXRBJKcgki6I5BREAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAE\nQCQAARAJQABE0oWLVp2yRKTTL/3H/Z/9VqELIjllgUin+/8ej/s/B61CF0RyCiLpgkhOERLp\nHP5EpDiI5BRxkf5rKEWkKkquihDJKWIiFXmyofoTAZFAFt+7dogESiCSLojkFN9n7RAJlEAk\nXRDJKWuubDiFj4u+sgGRQAnf19qVJxI4BZEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQ\nAJEABEAkAAEQCUAARNKFi1adgki6IJJTEEkXRHIKIumCSE5BJF0QySmIpAsiOQWRdEEkpyCS\nLojkFETSBZGcgki6IJJTEEkXRHIKIumCSE5BJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACR\nAARAJAABEAlAAEQCEACRAARAJF24aNUpiKQLIjkFkXRBJKcgki6I5BRE0gWRnIJIuiCSUxBJ\nF0RyCiLpgkhOQSRdEMkpiKQLIjkFkXRBJKcgki6I5BREAhAAkQAEQCQAARAJQABEAhAAkQAE\nQCQAARAJQABEAhAAkQAEQCQAARAJQABE0oWLVp2CSLogklMQSRdEcgoi6YJITkEkXRDJKYik\nCyI5BZF0QSSnIJIuiOQURNIFkZyCSLogklMQSRdEcgoi6YJITkEkAAEQCUAARAIQAJEABEAk\nAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkXbho1SmIpAsiOQWRdEEkpyCSLojkFETS\nBZGcgki6IJJTEEkXRHIKIumCSE5BJF0QySmIpAsiOQWRdEEkpyCSLojkFEQCEACRAARAJAAB\nEAlAAEQCEACRAARAJAABEAlAAEQCEMCDSFUcRAIdXIgU0+UPIoESiAQgACLpwkWrTkEkXRDJ\nKYikCyI5BZF0QSSnIJIuiOQURNIFkZyCSLogklMQSRdEcgoi6YJITkEkXRDJKYikCyI5BZF0\nQSSnIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAAcQqb6ASLAj/kW6GdQxCZFA\nFvciPfwJX5UQCWRxLlJvj+7+Gxetgiy+RapjvyMSyOJapHqw5GYSIoEsBxPpZhIigSyeRarH\n3keqEQl2wLFI9fgbsjUigTzHE6k1CZFAFr8itb4gEuhwQJHalyrVAgMQySluRbocCo2K1JyF\nUC0wAJGc4lWk68m5cZH+1IgEsiCSLojklGOK9IfZDLI4Fel2AQMigQ4HFanCJBAFkQAE8CnS\n/SLvqEhnTAJJEAlAAJciPT6/FxcJk0CS44qESSCIR5GCe5wgEuhwYJEwCeRwKFJwxxNEAiUQ\nCUCAI4uUwyQuWnWKP5HCe9nN3URff1IjklMQSRdEcsqxRdI3CZGc4k6k7pdOIBLoYEikKsp6\nkdRNuooUfy7K9YAQlkSKWhHzqGCRZgsDWyCSLojklKOLpG0SIjnFmUh1dAkiwZ4g0h6lxkEk\npxxeJGWTEMkpiJTlfR1E8oYvkfrfGZs0X3OYhEjeQCREAgEQKYtJiOQNRNpRpPiFQIjkDUQ6\n72dStPvUwsAMrkTqe4RIoAUinREJtoNIDTuZhEjHAZEaEAk24kmkgUeIBFogUss+JiHScXAk\n0tCjAkWq6xqRPIJIF3YxCZGOAyJdQCTYBCJd2cMkRDoOfkQa8QiRQAtEurGDSYh0HBDpBiLB\nBhDpBiLBBhDpjrxJiHQc3Ig05hEigRaI9EDcJEQ6Doj0AJFgNYgUIG0SF60eBy8ijXqESKAF\nIoXUsl8BhkjHwYlI4x6tEklwiiPScUCkEESClSBShxqRYBWI1AGRYB0+RIp4hEigBSJ1qREJ\n1oBIXRAJVoFIPWJNIRJMgUg9EAnWgEj9XqJtiVTMRatOcSFS/FUEkUAHRBr0EntPSqRiRHIK\nIg16QSRYDiINekEkWA4iDXuJfCRDpGJEcgoiDXtRE+n6CyI5wINIUY9WihT52LpIxTeR6vry\n6PYTkYyzRKTTL/3H/Z/9VgXxI1L/lSj4FZGMskCk0/1/j8f9n4NWBdETafxmkyIVNw7Vw8Lu\nKiGSURBprJddRapHC6sRyTRCIp3Dn/ZFGv2uJZGK42ftrgs2bCDIiLhI/zWoihT/5EN5It33\n4EafSI1IZhEUKdfJBkMi3ZuNFFYjklUQabyXoUnbRQpPziGSM+RECjxSFamO61KWSJ23iyJ1\n1YhkFDGRQo8ciDQ0aatInQbjJ052+Xp12B0pkToeIdJIxWkiVbt8KzTszporG07h48vP06lz\naYMDkQYmbRSp29xUYZhkEfvX2tkQqdcaInnDvEgj19vIiDSY+1sqHlg5VRgmGQSRor0IijQ8\n4JosjDMO9kCkeC91YmS2rZFz6dOFIZI5ECnei5RIY5dJzBSGSdZApIle6rTITFudVmY/an6l\nFvjCQFAEkSZ6ERGp/xm+GZFG+17XPShiXaTxD/cMpuVgQVovGz8EPvJp22SRBM51gCKINNlL\nnRKZaqv/wpIu0vZDNFAEkSZ7QSRIA5Gme9l0W5KRm00uEGnzIRoogkjTvWwTaXjiG5GcYlyk\n6VmZW6SRN5CWiLT1XAcogkgzvdTzkRhjH91dJBI3vLMDIs30sl6k0W9IXybSthdEUASR5npZ\ne+vGWkKkLS+IoIltkeppXUREWnnrxjpSMSL5BJGSe1lUbB1pa7aw8c4RqXQQKbmXRcWKicS9\njG2ASPO9/Fl+M+E62tZcYdEnuW3jwc4g0nwvy0Wq423NFRZ9kts2HuyMaZFuez0r5uuCXq4d\nLdg+tahIK14QQR1Emu3l2lP69rl/vFVIpKUeQwYQabaXa1fJ2+fxMXFEOg6INNvLtauMIi17\nQYQcINJsL9e+UrdPcN8SMZH4moriQaTZXm6dpVUZriYnEl9TUTqWRZr8+ruZ+RonFkkTqbMW\nIh2Ho4q0IpJiUncdycK4013ZIFJyJOFOwr0VBC5afQT32KQgBiIlR2bvfzoQTVKkpBdEyAYi\nJUfOMyYNF4qKxG2MiwaR0mfyedKkkUWyImFSySDSgpl8npjMYwuERcKkgjEs0uPDoyusWBG5\nVDF+xmH8r4h0HBBpyUxuGUoTO50nLRImlQsiLZrJFzri1PGz4oh0HBBp0Uy+UdetP/WERSMV\n/9kqEiYVCyItm8khs2/Qyou06OtlJ66DSm8E0kCkhTN5S8V/tou05DVp4kkufy4wjV2RgpuQ\nxifM4gXFi7TAJERSBJGSI5sr3lQYIpUNIiVHNle8qbBH98kmIZIiiJQc2VzxpsLCs+/bu1/+\nXGAasyKFX0opOF8tiJRqEiIpgkjJka0Vbyus+zbW1u6XPxeYBpGSI1sr3lYYIpUNIiVHtla8\nrbBu90kmIZIiiJQc2VrxtsJ63aeYhEiKWBWp8z3FgvN1jUjL70i0XaQUkxBJEURKjqQVtrWX\n5O7nTUIkRRApOZJW2NZe0rufNQmRFEGk5EhaYVt7Se8ekUoCkZIjaYXNtLX9otUHKz7FMftc\nYCWIlBxJK2ymLUmR5kxCJEUQKTmSVthMW6IizZiESIoYFanuTIv4hFm8wJZI0yYhkiKIlBxJ\nKmyuLWGRJk1CJEUQKTmSVNhcW9IiTZmESIogUnIkqbC5tsRFmjAJkRRBpORIUmFzbcmLFDcJ\nkRSxKVLdnRbxCbN4gUGRoiYhkiKIlBxJKWy2rT1EipmESIogUnIkpbDZthDJKYiUHEkpbLat\nXUSKmIRIipgUqe5Ni/iEWbzA0EWrIaN3MkYkRRApOZJQmEAvy7u/kPaFgYmNwWIQKTmSUJhA\nL8u7v5L0FbapjcFSECk5klCYQC/Lu7+R8qXqyY3BQhApOZJQmEAvy7u/g0gZQaTkSEJhAr0s\n7/5BzyREUsSiSHV/WgjOV9Mi9UxCJEUQKTkyX5hEL8u7D+mYhEiKIFJyZL4wiV6Wd98hPOOA\nSIogUnJkvjCJXpZ33+NhEiIpgkjJkfnCJHpZ3n2fu0mIpAgiJUfmC5PoZXn3fRApB1GRnt4/\nJVoVxIVI+1y02uVmEiIpEhWpqqrT68fmVgW5zYu+R4jU53rGAZEUiYr08++l+T6F53/fm1oV\nBJHSqSfrQiR5Jo+RPt5Ovy49LX9dQqRYW0oitSYhkiLTJxu+39qv+Xle36ogiLSE3907RFJk\nSqSvl/bl6PO5elndqiCItAhEUiUu0sfzfa+uWrrhESnW1g4ixb4usEYkReKnv6vq5eu26LS6\nVUEQaWH3gw2V0BisJH76++3rvBpEirWlKVIVNQmRxImf/pZpVZDrhBnODkSKtVVHVEIkcSbe\nkL38PC3dreu2KggiLe8ekbQYF+kUHrRualUQRFrR/bhJiCTOuEh/A4/+bmpVEBcirS5sbfej\nJiGSOLO7dhtbFQSRVnU/dpyESOLY+xgFIi3sHpE0GBfp9+Wo1GOksWkhOF8dijT2T88e43Ns\nECk5MlOYUC97dD98522P8Tk25nbtDitSnPm2Bh+F3GN8jg0iJUdmChPqZZ+2+rfU3GN8jk1c\npL+n8/mzOr1va1WQ6NlcRJpf0PtGqT3G59hERfr7e3D03bwxu8YkRBIvbGtb3a+v3mN8js3E\n1d+fv//9/Vp85Xe3VUEQaUsEkfZl6g3Zj+pp5RuziCRe2Oa2wi2HSOJERTpV36/VV3OUtKlV\nQRBpWyS4xAGRxImK9N7cj6t5QXrb1KogzbwYv3Js8xxLikwWltrW+otWYwvSI4i0I/Gzdm/V\n6eP3hWmNR4gUbSunSI+th0jiGHsfCZG2RBBpPxApOTJZWGpbWUW6bz5EEmdi1+5U4LV2iLQt\nct1+iCROVKS3Ii9aRaSNkcsGRCRxJk5/r/lo7KBVQRBJInLpfo/xOTbGPiGLSFsjiLQPUZFe\nqg035NpLpMitPITm2ExkqrDktrKL1G5DRBInKtL36XnNF7r0WxUEkUQiiLQLU180Vt7JBkQS\niNSItAOIlByZKiy5rQJE+q0BkcSx9YasfZFWFybZFiLJg0jJkanC5HrRaOtPHX0usJIJkf6+\n/O7WPa/6TgpEEi8Mkcom/m0UT+3xUVV9bmpVEESSi2CSNFGRXqu35k3Zf8u/QPa8m0ix71YQ\nnWPRBROFCfai0RYiyTN5ZcPtvw2tCoJIcpEzJgmDSMmRicIEe9Foq3kumCTL3K7dW/W6qVVB\nEEkucsYkYeInG64fRzqtuVAIkcQLQ6SymTj9/f5UVU9vqy5d3Uek2HcLI9LiSFM2Jkli6Q1Z\nRBKLNGUjkiSIlByJFybZi0Zb1+eCSYLEvh+pxC9j9iBSCRetItIOGBIp5hEiLY9ct+gew3RQ\n4rt2L80H+76fX7a1KggiyUVum3SPcTomsx81r9aYhEixtsoSCZPEmLv5yQ+7dv3JNwCRIC7S\nc3XZteMVqT/5BhgWCZOkmLj5SWlXNiCSXOSxUfcYqSMSP9nw89Zc2fBezJUNtdIcm598fSyL\nhElC2HlDFpEEI+F23WGsDggiJUeildkWCZNEQKTkSLQyRAJESo9EKzMuEiZJgEjJkWhlKoUh\nUtkgUnIkjkZhO4qESQIgkpEIIpWNGZFqnT2oYiN7ioRJ20EkI5FdRcKkzSCSkQgilQ0iGYns\nKxImbQWRjEQQqWysiFQrvV1TbGRnkTBpI4hkJIJIZYNIRiJ7i4RJ20AkI5HdRcKkTRgRqRnk\nUqf4okiJF62GGxlWgki6kUJEGqVedxNDaEAk3UghIo3/uZ64xB2mQSTdCCI5BZF0I4jkFETS\njRQt0p8akdZiQ6T2fFJ5VqyIIJJTEEk3UrZIfzgDvhZE0o0ULlKFSStBJN1I6SLxruxKEEk3\ngkhOQSTdSPEiYdI6EMlIBJHKBpGMRBCpbEyIdBnb3HMsb0RNJExaBSIZieiJhElrQCQjEUWR\nMGkFiGQkgkhlg0hGIpoiYdJyEMlIBJHKBpGMRBCpbCyIdB3W3HMsb0RVJExaDCIZiSBS2SCS\nkYiuSJi0FETSjZR/0Wp3o0MiiKQbMSMSJi0DkXQjiOQUAyLdRjT3HBOJ2BEJkxaBSLoRRHIK\nIulGEMkpiKQbMSQSJi2hfJHuw5l7jolELImESQtAJN2IKZEwKR1E0o0gklMQSTdiSyRMSgaR\ndCOI5JTiRXoMZe45ljeCSGWDSEYieUTCpFQQyUgEkcoGkYxEMomESYkgkpFILpEwKQ1EMhJB\npLJBJCORbCJhUhKIZCSCSGVTukjBKOaeY3kj+UTCpBQQyUgkt0hVFLGxNg0iGYlkFKkdhIWR\nw4FIRiKIVDaIpBsxdtHqYxQQaZrCRQqPc3PPMZEIIjkFkXQjNkX6HQdEmgaRdCNGRTrXiDQN\nIulGEMkpZYvUeSsw9xwTiVgV6VwvjxwKRNKNIJJTEEk3YlakKmYSIrUgkm6kdJHiINIkiKQb\nKV2kiUjEJERqKVqk7mXHBc+x9AWGRYqYhEgtiKQbQSSnIJKRSAkVj5qESC2IZCRSQsWIFAeR\njESKqHjMJERqWSLS6Zf+48ffTsGaiCQeKaJiRIqyQKTT/X+Px4+/nRBp10gRFSNSFCmRTju8\nIvVuulH2HNs7UkbFIyYhUovYKxIi7RsppOKhSYjUIi7Sfw2IJB4ppeKBSYjUwiuSkUgpFSPS\nOIhkJFJMxX2TEKkFkYxEiqkYkUZBJCORcirumYRILQWL1L/ltIE5Nr/A8kWrVxBpjDVXNpzC\nx/td2YBI8t2LRLomIVJLwdfaIZJ89zKRjkmI1IJIuhFEcgoi6UZciNQxCZFayhVp8PVWNubY\nzAJEcgoi6UZ8iBSahEgtiKQbQSSnIJJuBJGcUqxIw28ANjLHphc4ESkwCZFaEEk34kWkh0mI\n1IJIuhE3It1NQqQWRDISKa5iROqASEYi5VV8NQmRWkoVaegRIhUWQaQQRDISKbDii0mI1IJI\nRiIlVnw5b7J5rF2ASEYiJVaMSA8QyUikyIobkxCpBZGMRIqsGJHuFCrSiEeIVGCkRqQriGQk\nUmbFiHQDkYxECq24RqQLiGQkUmjFiHQFkXQjfi5avVIjUkuZIo15hEhlRkbH6nggkm7En0gV\nJjUgkm4EkZyCSLoRhyKNj9bRKFKk8ZHJPWFEIojkFETSjXgUCZPOiKQdQSSnIJJuxKVImFSm\nSJFhyT1hRCI+RcIkRFKOIJJTEEk34lQkTEIkI5FyK54as+OASEYi5VY8OWiHAZGMRMqteHLQ\nDkOBIsWGJPeEyRspt+KZYTsIiGQkUm7Fc+N2DBDJSKTcimcH7hAgkpFIuRXPDtwhQCQjkXIr\nnh+5I4BIRiLlVjw/ckegPJGio5F7wuSNlFtxytj5B5GMRMqtOGXs/INIRiLlVpw0eO5BJN2I\n14tWZ0fPO4ikG/Eu0mFNKk6k+EDknjAiEURyCiLpRtyLdFSTEEk3gkhOKU2kiWHIPWFEIv5F\nOqhJiKQbOYBI5/qIKiGSbuQIIh3yNQmRdCOHEOmIJhUm0tQI5J4wIhFEcgoi6UaOIdIBTUIk\n3QgiOQWRjETKrXjxQLoEkYxEyq14+Uh6BJGMRMqteMVQOqQskSY3fu4JkzdSbsWrBtMdiGQk\nUm7FqwbTHYhkJFJuxetG0xuIZCRSbsUrh9MZiGQkUm7Fa8fTF0WJNL3dc0+YvJFyK149oK5A\nJCORciteP6KeKEmkma2ee8LkjZRb8ZYx9QMiGYmUW/GWMfUDIulGDnLRavKgugGRdCOHE+ko\nHzwvSKS5DZ57wohEjifSQV6TEEk3ckCRjmESIulGEMkpiKQbOaJIhzAJkXQjhxTpCCccEEk3\nckyRDvCaVI5Is9s694QRiRxUJP8mIZJu5KgiuTcJkXQjiOQURNKNHFYk7yYhkpFIuRUnD7Br\nlYoRaX4r554weSPlVpxMfX+0YkqVDiIZiXiouL4tWTGlSgeRjERcVHw1CZGWt5oMIhXdvVDF\nF5MQaXmrqSQciJY6YRBpwYLLOcsVU6p0EMlIxEvFjUmItLzVRFLOjJY6YRBp0YIakda0mggi\nFd49Is2ASEYifiquEWlFq4kgUuHdS1bs8hIHRDIS8VQxIi1vNUb06pH49SOlTphFEYcXra5o\nq64iCM5BZbKJ1N20dbCVEyNJI1ZcBJHaJXVkgeAcVAaRdCOIdFkybhIiRVuNgUhZuteMTLc1\nahIiRVuNgUhZuteMzLQ1ZhIiRVuNgUhZuteMzLU1YhIiRVuNgUhZuteMIJJkqzEQKUv3mpHZ\ntoYmIVK01RiIlKV7zch8W3VfJUSKthoDkbJ0rxlJaatnEiJFW42BSFm614wgkmSrMRApS/ea\nkaS2uiYhUrTVGEcVaXXEa8UdkxAp2moMRLLV/X4V1yljXz6IZCTit2JESmk1BiLZ6n7Pih+D\nj0jRVmMgkq3ud634PvqIFG01BiLZ6n7fim/HSYgUbTUGItnqfueKEWmu1RiIZKv7vSs2fy9j\nRDIS8V6x9XsZI5KRiPuKjd/LuECRogiNmM2I/4rrqbEvXrECRdp9xExGDlBxPRURnJ27gEi6\nES5anVhSI1K01RiIlKV7zQgiSbYaA5GydK8ZWdFW5L6RfxApCiJl6V4zsqatwYfP70sEZ+cu\nIJJuBJFmIlZvZoxIuhFEmokg0nirMRApS/eakZVtGb0rOCLpRhBpNmLzruCIpBtBpPnI2BkH\nRIqASFm614ysbwuRhq3GQKQs3WtGNrRl8GbGiKQbQaSkyMAkRIpwVJFWRw5Wsbm7giOSkcjR\nKrZ2M2NEMhI5XMXGbmaMSEYix6vY1s2MEclI5HgVIxIi7RA5YMWhSYgUAZFsdZ+l4uASB0SK\ngEi2us9TMSLNroFItrrPVLGdu4IjkpHIQSu+TQxEioBItrrPVrGVu4IjkpHIYSs2cldwRNKN\ncNHq4oiNu4IjklGhSq8AAAb8SURBVG4EkZZHTNwVHJF0I4i0IlIjUhREytK9ZkSwrRqRYiBS\nlu41I5Jt1YgUAZGydK8ZEe2+Fpydu4BIuhFEWhkp3SRE0o0g0soIIo2DSFm614zIdn8u3CRE\n0o0g0srI+VwXrRIi6UYQaWWkmTQlm4RIuhFEWhlpZ03BJiGSkcjRK75Mm3JNQiQjkaNXfJ03\nxZqESEYiR6/4NnFKNQmRjESOXvF95hRqEiIZiRy94sfUKdMkRDISOXrFwdwp0iREMhI5esXh\n5CnRJEQyEjl6xZ3ZU6BJiGQkcvSKu9OnPJMQyUjk6BX35k9xF94hkpHI0SsezKDCTEIkI5Gj\nVzycQmWZhEi6ES5aXRkZmUNFmYRIuhFEWhkZm0QlHSghkm4EkVZGxqdROSYhkm4EkVZGIvOo\nGJMQSTeCSCsjsYlUyu4dIulGEGllJD6VyjAJkXQjiLQyEqUQkxBJN4JI0pFzGSYhkm4EkaQj\nzWwq4EAJkXQjiCQducyn7CYhkm4EkaQj1wmV2yRE0o0gknTkNqMym4RIRiJUHFlwn1J5D5QQ\nyUiEiiMLgkmVUyVEMhKh4siCzrTKpxIiGYlQcWRBb2LlMgmRjESoOLKgP7MyvSghkpEIFUcW\nDOfWVaWpq4rkQSQjESqOLBibXa1KyyJbQSQjESqOLBifX78qIdLibRlb4ClCxZEFsRlW14sj\nW0AkIxEqjiyIT7GYSYi0dIGnCBVHFkxMsXr8VQmRBDe/uQgVRxZMT7ExlxBJcPPninDRqnRk\ndooNVEIkwc2fK4JI0pGEKdZ7WUIkwc2fK4JI0pG0KRaqhEiCmz9XBJGkI8lT7O4SIglu/lwR\nRJKOxBmse1Vpwr31VxUhkm4EkbJG6mb7J7+IBW0tmPKIpBFBpOyROn59OCKZiSBSAZFzHZEJ\nkcxEEKmASDsDx1RCJDMRRCogcp2Dw9clRDITQaQCIsE87MqESGYiiFRApDcXHzIhkv8IFctF\nRubjRSZE8h+hYrlIZE7WdeRzF4jkKELFcpGpWTnuEiK5iVCxXGRmVo64hEhuIlQsF5mdlYjk\nN0LFcpHZWYlIfiNULBeZnZWI5DdCxXKR2VmJSH4jVCwXmZ2ViOQ3QsVykdlZiUh+I1QsF5md\nlTuLdPql/7j/s9/qXMnDyj2NmGCEiuUis7NyX5FO9/89Hvd/DlqdK3lYuacRG/6Ji1YLiMzO\nSkQqPoJIBURmZyUiFR9BpAIis7OyBJH+a0gQqUsNUBC73o5L9BXpqNQ5v8Ie9gORdEEkpyCS\nLojkFETSBZGcgki6IJJT1lzZcAofr7yy4aggklNyXWt3VBDJKYikCyI5BZEABEAkAAEQCUAA\nRAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJF04aJVpyCSLojkFETSBZGcgki6\nIJJTEEkXRHIKIumCSE5BJF0QySmIpAsiOQWRdEEkpyCSLojkFETSBZGcgki6IJJTdhJpGf8t\nXD8bZgqlUnlmCt1FpIX8l6/rZZgplErlSS0UkeYxUyiVyoNIcpgplErlQSQ5zBRKpfIYEAnA\nD4gEIAAiAQiASAAC5BLpdCNT/8mYKZRK5VlQaCaRTiOPisRMoVQqz5JCEWkaM4VSqTyIJIeZ\nQqlUHkSSw0yhVCqPAZFcHm9mhkrFKf9kA4AvEAlAAESawcxuCJXKw66dGGYOjKlUHgsnG6xg\nZtCpVB5EksPMoFOpPIgkh5lBp1J5EEkQMwfGVCoPJxsAdEEkAAEQaYb7C3vx+yEXLJTZbFEz\nu3Zndu0kaDdhuyFLH/XTycqhx2NrFl7pqfu/SRBpmtP9R+GDbmRuNiDSATndfxY+6OfkIc8O\nIh2Q+xYsfofpfD32yF3EPPepWXqtiCTI6fGaZAALL5yPw7ncdczCyYbDYmB2+gSRAARAJAAB\nEAlAAEQCEACRAARAJAABEAlAAEQyRsWIFQnDYgxEKhOGxRiIVCYMix2+n6uXi0g/r1X1+nP9\n29NH87eq+jo9h0sej0ABRDLDz6mqqpdWpOZR9XT7W3UR6bl6DZYEj0ABRDLDW/V8/nlupHmv\n3ppf/zaPbn+rmr91ltwegQaIZIan6vt3V66R5qkdtd/9vOBvVfOos+T2CDRAJDNcTjNcpHns\n0D3+dnn0WHJ7BBqwoc2ASCXDhjZDf9eu+7eLMuGSHDUeFza3Gd6r559ze2LhrTmR8K96vpyA\nOD8/RAqX3B6BBohkhsfp7+tJ76/u6e/7OuGSr9xVHwVEssP3y+0N2e/Xqnr+bP/2XD39e4gU\nLrk/AgUQyQEVdzzJDiKZpqo+m8Oh19x1ACKZ5u1yiPSduw5AJNv8faqqVzzKDyIBCIBIAAIg\nEoAAiAQgACIBCIBIAAIgEoAAiAQgwP8Bc/AoO5g4AgkAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ggplot(chr12_hubs, aes(x=degree)) + \n",
    "  geom_histogram(aes(y=..density..),binwidth=30,colour=\"black\", fill = \"lightblue\")+ #distribution plot\n",
    "  geom_density(alpha=.2, color=\"red\")+  #density line,Smoothed density estimates\n",
    "  geom_vline(aes(xintercept=mean(degree, na.rm=T)), color=\"red\", linetype=\"dashed\", size=1)+  #mean\n",
    "  ggtitle(\"Distribution of hubs' degree in chromosome 12\")+ backgroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Var1</th><th scope=col>Freq</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>3  </td><td>5</td></tr>\n",
       "\t<tr><th scope=row>63</th><td>110</td><td>5</td></tr>\n",
       "\t<tr><th scope=row>122</th><td>189</td><td>5</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & Var1 & Freq\\\\\n",
       "  & <fct> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & 3   & 5\\\\\n",
       "\t63 & 110 & 5\\\\\n",
       "\t122 & 189 & 5\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 2\n",
       "\n",
       "| <!--/--> | Var1 &lt;fct&gt; | Freq &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| 1 | 3   | 5 |\n",
       "| 63 | 110 | 5 |\n",
       "| 122 | 189 | 5 |\n",
       "\n"
      ],
      "text/plain": [
       "    Var1 Freq\n",
       "1   3    5   \n",
       "63  110  5   \n",
       "122 189  5   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "distribution[distribution$Freq ==max(distribution$Freq),]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 id='5'>5. Modules for chromosome 12<h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 1 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>nodes</th><th scope=col>communityCount</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>370956</td><td>307385</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       " nodes & communityCount\\\\\n",
       " <int> & <int>\\\\\n",
       "\\hline\n",
       "\t 370956 & 307385\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 2\n",
       "\n",
       "| nodes &lt;int&gt; | communityCount &lt;int&gt; |\n",
       "|---|---|\n",
       "| 370956 | 307385 |\n",
       "\n"
      ],
      "text/plain": [
       "  nodes  communityCount\n",
       "1 370956 307385        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# To know the numbers of modules ----------------\n",
    "query = \"CALL algo.louvain(\n",
    "                        'MATCH (n:chr12)--(m) with id(n) as chr12,id(m) AS m   MATCH (al) where id(al) = m or  id(al) = chr12 return id(al) as id',\n",
    "                        'MATCH (n:chr12)-[r]-(m) RETURN id(n) AS source, id(m) AS target',\n",
    "                        {graph:'cypher',direction: 'both'})\n",
    "YIELD nodes, communityCount\"\n",
    "cypher(graph, query)  #Results： nodes:370956 communityCount:307385"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At here analysis, you need to know 1288086 nodes including all chr12 nodes and all nodes directly related chr12 nodes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Then we filter all nodes are chr12 in their community as following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To know the numbers of chr12 nodes in each modules ----------------\n",
    "query = \"\n",
    "// to make louvain modules for chr12\n",
    "CALL algo.louvain.stream(\n",
    "  'MATCH (n:chr12)--(m) with id(n) as chr12,id(m) AS m   MATCH (al) where id(al) = m or  id(al) = chr12 return id(al) as id',\n",
    "  'MATCH (n:chr12)-[r]-(m) RETURN id(n) AS source, id(m) AS target',\n",
    "  {graph:'cypher',direction: 'both'})\n",
    "YIELD nodeId, community\n",
    "\n",
    "// filering community which the number of nodes is more than one\n",
    "with collect(nodeId) AS nodeId , community\n",
    "where size(nodeId) > 1\n",
    "with nodeId, 'Module'+ community as community\n",
    "\n",
    "// to get communities that contain chromosome 12 nodes. \n",
    "MATCH (chr12:chr12)\n",
    "where id(chr12) in nodeId\n",
    "with collect( chr12.Name )  AS GeneList, community\n",
    "RETURN GeneList, community, size(GeneList) AS Size\n",
    "ORDER BY Size DESC\n",
    "\"\n",
    "module_chr12_total = cypher(graph, query) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>GeneList</th><th scope=col>community</th><th scope=col>Size</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Bin17 , Bin37 , Bin39 , Bin43 , Bin46 , Bin47 , Bin50 , Bin55 , Bin56 , Bin57 , Bin68 , Bin69 , Bin70 , Bin71 , Bin76 , Bin78 , Bin82 , Bin83 , Bin85 , Bin86 , Bin87 , Bin88 , Bin89 , Bin90 , Bin91 , Bin93 , Bin94 , Bin96 , Bin97 , Bin99 , Bin100, Bin101, Bin102, Bin103, Bin104, Bin105, Bin106, Bin107, Bin110, Bin112, Bin114, Bin115, Bin116, Bin117, Bin118, Bin119, Bin121, Bin122, Bin123, Bin142, Bin143, Bin144, Bin145, Bin146, Bin148, Bin149, Bin150, Bin151, Bin166, Bin167, Bin170, Bin172, Bin174, Bin175, Bin177, Bin178, Bin181, Bin183, Bin185, Bin188, Bin189, Bin190, Bin193, Bin194, Bin196, Bin197, Bin200, Bin201, Bin203, Bin204, Bin205, Bin206, Bin207, Bin208, Bin209, Bin210, Bin215, Bin216, Bin217, Bin218, Bin221, Bin222, Bin223, Bin224, Bin228, Bin238, Bin271, Bin276, Bin277, Bin278, Bin291, Bin293, Bin294, Bin297, Bin298, Bin299, Bin300, Bin301, Bin303, Bin304, Bin305, Bin307, Bin308, Bin309, Bin310, Bin311, Bin318, Bin322, Bin333, Bin334, Bin335, Bin336, Bin341, Bin344, Bin345, Bin346, Bin355, Bin356, Bin357, Bin361, Bin362, Bin363, Bin364, Bin365, Bin366, Bin367, Bin368, Bin369, Bin370, Bin374, Bin375, Bin376, Bin387, Bin388, Bin389, Bin390, Bin391, Bin392, Bin393, Bin394, Bin395, Bin396, Bin397, Bin398, Bin402, Bin403, Bin404, Bin405, Bin406, Bin407, Bin408, Bin409, Bin410, Bin411, Bin413, Bin414, Bin416, Bin417, Bin418, Bin419, Bin420, Bin421, Bin422, Bin424, Bin425, Bin426, Bin427, Bin428, Bin429, Bin430, Bin431, Bin432, Bin434, Bin435, Bin436, Bin437, Bin438, Bin442, Bin444, Bin445, Bin446, Bin447, Bin451, Bin452, Bin454, Bin456, Bin460, Bin461, Bin486, Bin487, Bin488, Bin490, Bin491, Bin495, Bin496, Bin497, Bin498, Bin499, Bin504, Bin505, Bin506, Bin512, Bin514, Bin516, Bin517, Bin533, Bin542, Bin552, Bin564, Bin575, Bin576, Bin577, Bin583, Bin584, Bin585, Bin587, Bin588, Bin594, Bin595, Bin622, Bin627, Bin628, Bin629, Bin630, Bin632, Bin633, Bin634, Bin635, Bin637, Bin638, Bin639, Bin640, Bin641, Bin643, Bin646, Bin648, Bin650, Bin653, Bin654</td><td>Module11196</td><td>249</td></tr>\n",
       "\t<tr><td>Bin1  , Bin2  , Bin20 , Bin24 , Bin38 , Bin42 , Bin44 , Bin49 , Bin52 , Bin53 , Bin58 , Bin72 , Bin73 , Bin74 , Bin92 , Bin108, Bin111, Bin113, Bin120, Bin126, Bin156, Bin160, Bin161, Bin165, Bin171, Bin179, Bin191, Bin192, Bin195, Bin199, Bin202, Bin211, Bin220, Bin225, Bin226, Bin243, Bin273, Bin275, Bin302, Bin316, Bin325, Bin332, Bin342, Bin351, Bin353, Bin371, Bin373, Bin377, Bin386, Bin455, Bin457, Bin462, Bin463, Bin472, Bin475, Bin484, Bin485, Bin492, Bin500, Bin503, Bin508, Bin515, Bin518, Bin540, Bin547, Bin549, Bin568, Bin597, Bin645, Bin651, Bin656, Bin657</td><td>Module382  </td><td> 72</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       " GeneList & community & Size\\\\\n",
       " <list> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t Bin17 , Bin37 , Bin39 , Bin43 , Bin46 , Bin47 , Bin50 , Bin55 , Bin56 , Bin57 , Bin68 , Bin69 , Bin70 , Bin71 , Bin76 , Bin78 , Bin82 , Bin83 , Bin85 , Bin86 , Bin87 , Bin88 , Bin89 , Bin90 , Bin91 , Bin93 , Bin94 , Bin96 , Bin97 , Bin99 , Bin100, Bin101, Bin102, Bin103, Bin104, Bin105, Bin106, Bin107, Bin110, Bin112, Bin114, Bin115, Bin116, Bin117, Bin118, Bin119, Bin121, Bin122, Bin123, Bin142, Bin143, Bin144, Bin145, Bin146, Bin148, Bin149, Bin150, Bin151, Bin166, Bin167, Bin170, Bin172, Bin174, Bin175, Bin177, Bin178, Bin181, Bin183, Bin185, Bin188, Bin189, Bin190, Bin193, Bin194, Bin196, Bin197, Bin200, Bin201, Bin203, Bin204, Bin205, Bin206, Bin207, Bin208, Bin209, Bin210, Bin215, Bin216, Bin217, Bin218, Bin221, Bin222, Bin223, Bin224, Bin228, Bin238, Bin271, Bin276, Bin277, Bin278, Bin291, Bin293, Bin294, Bin297, Bin298, Bin299, Bin300, Bin301, Bin303, Bin304, Bin305, Bin307, Bin308, Bin309, Bin310, Bin311, Bin318, Bin322, Bin333, Bin334, Bin335, Bin336, Bin341, Bin344, Bin345, Bin346, Bin355, Bin356, Bin357, Bin361, Bin362, Bin363, Bin364, Bin365, Bin366, Bin367, Bin368, Bin369, Bin370, Bin374, Bin375, Bin376, Bin387, Bin388, Bin389, Bin390, Bin391, Bin392, Bin393, Bin394, Bin395, Bin396, Bin397, Bin398, Bin402, Bin403, Bin404, Bin405, Bin406, Bin407, Bin408, Bin409, Bin410, Bin411, Bin413, Bin414, Bin416, Bin417, Bin418, Bin419, Bin420, Bin421, Bin422, Bin424, Bin425, Bin426, Bin427, Bin428, Bin429, Bin430, Bin431, Bin432, Bin434, Bin435, Bin436, Bin437, Bin438, Bin442, Bin444, Bin445, Bin446, Bin447, Bin451, Bin452, Bin454, Bin456, Bin460, Bin461, Bin486, Bin487, Bin488, Bin490, Bin491, Bin495, Bin496, Bin497, Bin498, Bin499, Bin504, Bin505, Bin506, Bin512, Bin514, Bin516, Bin517, Bin533, Bin542, Bin552, Bin564, Bin575, Bin576, Bin577, Bin583, Bin584, Bin585, Bin587, Bin588, Bin594, Bin595, Bin622, Bin627, Bin628, Bin629, Bin630, Bin632, Bin633, Bin634, Bin635, Bin637, Bin638, Bin639, Bin640, Bin641, Bin643, Bin646, Bin648, Bin650, Bin653, Bin654 & Module11196 & 249\\\\\n",
       "\t Bin1  , Bin2  , Bin20 , Bin24 , Bin38 , Bin42 , Bin44 , Bin49 , Bin52 , Bin53 , Bin58 , Bin72 , Bin73 , Bin74 , Bin92 , Bin108, Bin111, Bin113, Bin120, Bin126, Bin156, Bin160, Bin161, Bin165, Bin171, Bin179, Bin191, Bin192, Bin195, Bin199, Bin202, Bin211, Bin220, Bin225, Bin226, Bin243, Bin273, Bin275, Bin302, Bin316, Bin325, Bin332, Bin342, Bin351, Bin353, Bin371, Bin373, Bin377, Bin386, Bin455, Bin457, Bin462, Bin463, Bin472, Bin475, Bin484, Bin485, Bin492, Bin500, Bin503, Bin508, Bin515, Bin518, Bin540, Bin547, Bin549, Bin568, Bin597, Bin645, Bin651, Bin656, Bin657 & Module382   &  72\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 3\n",
       "\n",
       "| GeneList &lt;list&gt; | community &lt;chr&gt; | Size &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| Bin17 , Bin37 , Bin39 , Bin43 , Bin46 , Bin47 , Bin50 , Bin55 , Bin56 , Bin57 , Bin68 , Bin69 , Bin70 , Bin71 , Bin76 , Bin78 , Bin82 , Bin83 , Bin85 , Bin86 , Bin87 , Bin88 , Bin89 , Bin90 , Bin91 , Bin93 , Bin94 , Bin96 , Bin97 , Bin99 , Bin100, Bin101, Bin102, Bin103, Bin104, Bin105, Bin106, Bin107, Bin110, Bin112, Bin114, Bin115, Bin116, Bin117, Bin118, Bin119, Bin121, Bin122, Bin123, Bin142, Bin143, Bin144, Bin145, Bin146, Bin148, Bin149, Bin150, Bin151, Bin166, Bin167, Bin170, Bin172, Bin174, Bin175, Bin177, Bin178, Bin181, Bin183, Bin185, Bin188, Bin189, Bin190, Bin193, Bin194, Bin196, Bin197, Bin200, Bin201, Bin203, Bin204, Bin205, Bin206, Bin207, Bin208, Bin209, Bin210, Bin215, Bin216, Bin217, Bin218, Bin221, Bin222, Bin223, Bin224, Bin228, Bin238, Bin271, Bin276, Bin277, Bin278, Bin291, Bin293, Bin294, Bin297, Bin298, Bin299, Bin300, Bin301, Bin303, Bin304, Bin305, Bin307, Bin308, Bin309, Bin310, Bin311, Bin318, Bin322, Bin333, Bin334, Bin335, Bin336, Bin341, Bin344, Bin345, Bin346, Bin355, Bin356, Bin357, Bin361, Bin362, Bin363, Bin364, Bin365, Bin366, Bin367, Bin368, Bin369, Bin370, Bin374, Bin375, Bin376, Bin387, Bin388, Bin389, Bin390, Bin391, Bin392, Bin393, Bin394, Bin395, Bin396, Bin397, Bin398, Bin402, Bin403, Bin404, Bin405, Bin406, Bin407, Bin408, Bin409, Bin410, Bin411, Bin413, Bin414, Bin416, Bin417, Bin418, Bin419, Bin420, Bin421, Bin422, Bin424, Bin425, Bin426, Bin427, Bin428, Bin429, Bin430, Bin431, Bin432, Bin434, Bin435, Bin436, Bin437, Bin438, Bin442, Bin444, Bin445, Bin446, Bin447, Bin451, Bin452, Bin454, Bin456, Bin460, Bin461, Bin486, Bin487, Bin488, Bin490, Bin491, Bin495, Bin496, Bin497, Bin498, Bin499, Bin504, Bin505, Bin506, Bin512, Bin514, Bin516, Bin517, Bin533, Bin542, Bin552, Bin564, Bin575, Bin576, Bin577, Bin583, Bin584, Bin585, Bin587, Bin588, Bin594, Bin595, Bin622, Bin627, Bin628, Bin629, Bin630, Bin632, Bin633, Bin634, Bin635, Bin637, Bin638, Bin639, Bin640, Bin641, Bin643, Bin646, Bin648, Bin650, Bin653, Bin654 | Module11196 | 249 |\n",
       "| Bin1  , Bin2  , Bin20 , Bin24 , Bin38 , Bin42 , Bin44 , Bin49 , Bin52 , Bin53 , Bin58 , Bin72 , Bin73 , Bin74 , Bin92 , Bin108, Bin111, Bin113, Bin120, Bin126, Bin156, Bin160, Bin161, Bin165, Bin171, Bin179, Bin191, Bin192, Bin195, Bin199, Bin202, Bin211, Bin220, Bin225, Bin226, Bin243, Bin273, Bin275, Bin302, Bin316, Bin325, Bin332, Bin342, Bin351, Bin353, Bin371, Bin373, Bin377, Bin386, Bin455, Bin457, Bin462, Bin463, Bin472, Bin475, Bin484, Bin485, Bin492, Bin500, Bin503, Bin508, Bin515, Bin518, Bin540, Bin547, Bin549, Bin568, Bin597, Bin645, Bin651, Bin656, Bin657 | Module382   |  72 |\n",
       "\n"
      ],
      "text/plain": [
       "  GeneList                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
       "1 Bin17 , Bin37 , Bin39 , Bin43 , Bin46 , Bin47 , Bin50 , Bin55 , Bin56 , Bin57 , Bin68 , Bin69 , Bin70 , Bin71 , Bin76 , Bin78 , Bin82 , Bin83 , Bin85 , Bin86 , Bin87 , Bin88 , Bin89 , Bin90 , Bin91 , Bin93 , Bin94 , Bin96 , Bin97 , Bin99 , Bin100, Bin101, Bin102, Bin103, Bin104, Bin105, Bin106, Bin107, Bin110, Bin112, Bin114, Bin115, Bin116, Bin117, Bin118, Bin119, Bin121, Bin122, Bin123, Bin142, Bin143, Bin144, Bin145, Bin146, Bin148, Bin149, Bin150, Bin151, Bin166, Bin167, Bin170, Bin172, Bin174, Bin175, Bin177, Bin178, Bin181, Bin183, Bin185, Bin188, Bin189, Bin190, Bin193, Bin194, Bin196, Bin197, Bin200, Bin201, Bin203, Bin204, Bin205, Bin206, Bin207, Bin208, Bin209, Bin210, Bin215, Bin216, Bin217, Bin218, Bin221, Bin222, Bin223, Bin224, Bin228, Bin238, Bin271, Bin276, Bin277, Bin278, Bin291, Bin293, Bin294, Bin297, Bin298, Bin299, Bin300, Bin301, Bin303, Bin304, Bin305, Bin307, Bin308, Bin309, Bin310, Bin311, Bin318, Bin322, Bin333, Bin334, Bin335, Bin336, Bin341, Bin344, Bin345, Bin346, Bin355, Bin356, Bin357, Bin361, Bin362, Bin363, Bin364, Bin365, Bin366, Bin367, Bin368, Bin369, Bin370, Bin374, Bin375, Bin376, Bin387, Bin388, Bin389, Bin390, Bin391, Bin392, Bin393, Bin394, Bin395, Bin396, Bin397, Bin398, Bin402, Bin403, Bin404, Bin405, Bin406, Bin407, Bin408, Bin409, Bin410, Bin411, Bin413, Bin414, Bin416, Bin417, Bin418, Bin419, Bin420, Bin421, Bin422, Bin424, Bin425, Bin426, Bin427, Bin428, Bin429, Bin430, Bin431, Bin432, Bin434, Bin435, Bin436, Bin437, Bin438, Bin442, Bin444, Bin445, Bin446, Bin447, Bin451, Bin452, Bin454, Bin456, Bin460, Bin461, Bin486, Bin487, Bin488, Bin490, Bin491, Bin495, Bin496, Bin497, Bin498, Bin499, Bin504, Bin505, Bin506, Bin512, Bin514, Bin516, Bin517, Bin533, Bin542, Bin552, Bin564, Bin575, Bin576, Bin577, Bin583, Bin584, Bin585, Bin587, Bin588, Bin594, Bin595, Bin622, Bin627, Bin628, Bin629, Bin630, Bin632, Bin633, Bin634, Bin635, Bin637, Bin638, Bin639, Bin640, Bin641, Bin643, Bin646, Bin648, Bin650, Bin653, Bin654\n",
       "2 Bin1  , Bin2  , Bin20 , Bin24 , Bin38 , Bin42 , Bin44 , Bin49 , Bin52 , Bin53 , Bin58 , Bin72 , Bin73 , Bin74 , Bin92 , Bin108, Bin111, Bin113, Bin120, Bin126, Bin156, Bin160, Bin161, Bin165, Bin171, Bin179, Bin191, Bin192, Bin195, Bin199, Bin202, Bin211, Bin220, Bin225, Bin226, Bin243, Bin273, Bin275, Bin302, Bin316, Bin325, Bin332, Bin342, Bin351, Bin353, Bin371, Bin373, Bin377, Bin386, Bin455, Bin457, Bin462, Bin463, Bin472, Bin475, Bin484, Bin485, Bin492, Bin500, Bin503, Bin508, Bin515, Bin518, Bin540, Bin547, Bin549, Bin568, Bin597, Bin645, Bin651, Bin656, Bin657                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        \n",
       "  community   Size\n",
       "1 Module11196 249 \n",
       "2 Module382    72 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "module_chr12_total[1:2,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>49</li>\n",
       "\t<li>3</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 49\n",
       "\\item 3\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 49\n",
       "2. 3\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 49  3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(module_chr12_total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAOVBMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6et2OaysrK9vb2+vr7Hx8fQ0NDZ2dnh4eHp6enw8PD///9TaHlrAAAACXBI\nWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO2diXajPLMA+RUnnpn7ZdP7P+yNbRYtjW1AtFqk\n6pwEm80ykxpE0406DwCb6Wo3AOAIIBJAARAJoACIBFAARAIoACIBFACRAAqASAAFQCSAAiAS\nQAF+r0hdt/W7n13XvRVpy5UtDdr+ZRZ/Vv7NP7rT4r3cm3PqPua2njnss4fhdVzw99R1L/+e\na+MSEGk1bz97+LUifbj8A1+6/xbu5f6c/7oXeVvpw2d2euN9WPDtuisLjH8SRNqyg68yTRn3\nV2XbVZ8lfODPX+v3wr3cn/Pdde9Pbnt/wcWj24Kf//zOP12Jrit+TkKkejsoiHpbhA98nTuB\nPLmDfM5L9/rsh99Z8PXajSK9vnS39ZY09Sns/C0U4HK4/vvpA58+/HRQx/9A/71cOsdfb8Hy\nn6ucl7478n7Z7r3f4OvlcvL/fPvpCLycw9POddZbv3kX/rv999K58/Ch0c6mJoUL8r2PDQ3W\nT9vmv84vP6u83pZ+vA79/Z9535clb5/BZtPisOFzx0KYJbfq8qb/8sMf5M8nX84fZ3lX0adH\nRyr7kPjLnsNz3PTNkyM/dxiGf8dLTy4xbIf/d44m0ul2lD/yf6TbklcXLL/NuvzzX/7NLvy9\nbfDzb/H6c+3cM/1x9mtdVktEui3p/8HinU1NChfke48bGlxoB1t9DVtd/tL+3l6ebtu6tLHB\n4rDhd49FNGumVd0k0tutd/t1uVb8r7teIgm7ij49OlLph8RfdtjlleCbx0d+9jB0/b/j7Sf8\nQ//gjHSfy3F8v3aET8Jfwrf/dzm2n5fJbbn79J/u+m/9eV1+Gv4Bfl5/Xf6b/bz+Y45Xph/j\nJp8+/m/tc1zSZTubmhQuyPc+NHRaf9z3uNXp+pf1L9jbrb9/2ezj+lf4Fm92Wxw1XD4WwqyZ\nVgV/9e+3v/i/lz/vt/GopLuKPv0jPFK5SOGXvbwZv074zSMtZg/D8O94Ob3FIp26JWGR5zia\nSJf/qb/Ff6S+V/IRzLqs/H496G+DTudxgY/+e7/ydrtI/Td1YoIll23+u84TdvYtLUj2Hrbq\nO9h5uFW46vk6//v6v6uwWbg4arh8LIRZM60KF7vrf+0/fbXrX+e3F3cVffrbeNhFkaIv+52H\n16LNsu+ZtvM92izY4uxLczSRpqn0Jpz0r38Ourt18q8Lbv/fXXvmly75S/Q/l7stuf2jRf84\nL+OSLtvZ1IpwQb73vNXjvsetLp/x8fcU7i3/5tFmecPnj4Uwa+5NP+d86dt9DX6m++gn0ae7\n8bCL+42/bHwiGb95NHv2MHTBFVa4xXmHjh0iTW96pgX9Vcz5W95/9I/zcGfZgpm9C0aEW/mP\nU7brvGXpy2ePxXKRPi59u7/hRdXyXaX7nb5s9B1mvvnD75mudOl1Fr1t0X9C+V3Wo6xI/vPt\nNutD3P9GkWb2/kCkS4/o7b8vMyJd+3Yv4Ql6L5HmvvlSkS77KX6B5H+5SFOHQ+ofXHi/BFPH\njsCyrl3epHBBvvc5kV7izxnjHBa6dpd+0vsYDEz30U9ccqSnI5XvNz5E1253/s0jke517bIF\n1+BfwWyUoKl77LQW6XG8/It93PlLuVwCvw/3uj+SP7R0p/6JYMO/6zxpZ9mCfO9zIuW7+7r/\nMflmabBBOBbCrOzwCSLdotKXrtLrGGxIdxV9+jkMy+QfEh2i7266Ixt+8+hfaPYwiCKd9kgP\nun7CLnutRHQcX6//bh/uzl+K+xiCsrcY6k2HYS+3IGkYoH5/EP7+CMPf8c6yBfne50QKt3q5\n/E1+noK9jXFfebPb4vck/C0ci3SWdPiCld3wQdegyeXFOToq4ST69OhGQf4h4ZeNIpXhN3fB\nkZ8/DJJI790uF0j+yCL1V/Onub+ULroh29/VC880X/29PTcd+ejOYvr/4vhh0s6yBfne50QK\nt+q/1K2fE9+JnNtMuiErHIt0lnT4gpUvF3hvw6dcrznexxuy6a7mb8gKHxJ82ct/M2MAO/zm\n44en3/OhSP11afRPV4bjinT97+v0717fJUwRivJMbrOu2SYuTxF6fY8+7cYtLaafl+8sW5Dv\nfVakcKuPawLN1+0P6eP2DfOP8dO6YYrQ63uwzgORpMMXvnntu12Xi53vfuFZ3nt82KJkKuFD\ngi8bpQiF3/y1C7PwZg6DJJJDpDbYqwdul+mm6Vvx2zOnfeICu4BIRehuV7znvvvyi3gbo8kf\ni8oonuA7TDi0DiIV4e/QY/hlJ6TrZcvwZlFh3xPMFvZZBJHK8H65ChivSH4NrnNv42nos/B/\nI6c0G9EyiARQAEQCKAAiARQAkQAKgEgABUAkgAIgEkABEAmgAIgEUAAbIv0fQJOMf8JGRKrd\nAIA1NCbSnwSdRgE8ApEACoBIAAVYIZJLpun8ZJH7QZoOa03v4/aIIBLYZI1ILpwE8+dfunw6\n7MDF6yMSNEk9kZwvI1L5p1cALGZV127yIOivpbb0i/Iz0TgtIlJn5eoOfjXbRIrFid7Pd+n8\njEj/u7BcJM5IYIByIvl5kVwwvZ29+h35RDDOSNAo66J27sFpx916dq7v4vlw2q/iEQkOxH4i\nJZsIU0SCw7C7SHeulYjawWFYeUO2/5m7ELrvWFmRAAywTaQ0/B2GwecyGjZlNgDYpLFcOwCb\nNCYS/TqwCSIBFACRAAqgVkYR3qCNQ+OryyiebTDA7qhlfw9vwvC3S5cjEjSKqkixOM4XEImc\nVTCBWhnFMHHTbqLXWXtEcpGebTXAnqiVUfjpPOSiZeOSxWUUnhMSWEGtjMKPsYVg/rhHoT0i\nmUick8AEemUU+eVRAZGWtBtgPxTLKNL+3Rj3DlfljARNolZGkYjkkzl5e0SEqN2zzQbYEbUy\nilmRIo+4jwRtoldGMdoTZzYEXby4PQAN0ViuHYBNGhPpT4ZOswDug0gABUAkgALUG40if4ZD\n3B4RRAKbVHuIfh4eT9ojgkhgkyOI1JG6CrXRHY0ifV/sjIRIUBfF0SjkYV4mkdaMRtGLhEdQ\nGb3RKJKsovCM9XxmAyKBTdTKKGbPROEoFStFwiOojVoZxd1rI0SCxqk2GkXJYAMiQW2qjUZR\nNmoHUJd6o1GsymwAsEljuXYANmlMpLxrR/8OLIBIAAVAJIACbBApvwmUvQ8X7VRGgUhggS0i\nuXASzJ9/uUP4G5HAAscRiVIKqMimrp2bfi8alaJkitAgEs9chZqUEUnKcJjJIvIz79eWUSAS\nWKC8SFLOXeCYD+qSwvdZe0QQCWyyLWq3blSKsctXsmvHUElQk/1FSja5O+WMBI2iJhJROzgy\nG2/I9j9SsGFhOUXWHpE7IgFUpIxIJcop0vYANERjuXYANmlMJLpzYBNEAigAIgEUQH80imHp\n1jKKZ5sLoID6Q/THxKBsftIeEUQCm+iL5HxBkWz0RwF0R6PoFfTy/Lg9ImQzgE0UR6PwD0Ra\nWEZxbTIigQ30RqPo30+/N45GQZ4qGEJ9NAofvgyWZ+0RiUViqD6wgvpoFD4VKZwSbIBGUR+N\nwiciEbWDI6A+GoUvKhKADfRHo0iCDZRRwBFoLNcOwCaNiUS/DmyCSAAFQCSAAtQvo4jXRiRo\nkuplFMmOEAmapHoZhXdrz0ikB4EdqpVReOn9EpFIWQVDVCuj8Nn7JWUUC5sMsDO7l1EI5RJu\n2lP83i+7RsIjMMPuZRShUOkZaWvUDpPACmrZ39ObLH7ONRI0T7UyivzElLRHhKgd2KRaGUUW\nckjbI8J9JLBJtTKKsRR9kUgANmks1w7AJo2J9Oc+Om0EyEAkgAIgEkABNg59GU7T+cmiIkNf\nIhLYZItILpwE8+df3g+Lx+0RQSSwySFFunwVbteCJpu6dm76vaycwifTsiJ1HQlEoEwZkaQM\nh9ksokGsWLDFo1E8OCMhEmhSXiQp5y48Mw3ZRUGW0arRKB6JhEegyLao3YpyiuzaaO1oFIgE\nhthfpGnt+SDDHsGGpV8HYAPVRdopasfQSaDKxhuywYVOGmx4UE6xp0gA2pQRaU05BaNRwIFo\nLNcOwCaNiUSHDmyCSAAFQCSAAuiPRjE/KkXcHhFEApuoP0R/Cneno1Ik7RFBJLBJPZHSUSnS\n9ojMicTdV6hLjdEoxk5d/Ctpj8iMSBRNQGVqjEYxI9LqMgpEguqojkYxiBROBgez9ojQtQOb\nqI1GMSdSgTIKzkhQnYoP0Q/0QSRonAoiOZ8sImoHzaM/GkVyHiojEkBd1EejyEahoIwCDkBj\nuXYANmlMJDp0YBNEAijA7tnf4/2kZLr5IfrPNhdAgd2TVqP3wVS6UesRCRpFRSTBo0Ii2eiP\nAuye/d3PEafhjdusPSKJSNyHBSPsnv3t73jUJ0n0wi3M/r42GZHABvtnf98/IfkpNTxpjwgZ\ndmCT3bO//SORojnLROK5xGCF3ZNWH3hEsAEOQUWRgtMVIkHj7J79PStSOOE+EjTO/tnfc2ei\nNMMhbY8IIoFNGsu1A7AJIgEUoDGRqIwFmyASQAHUyyiyPax/iP6zDQbYnRplFKNPmzIbEAkM\nUaGMwkXTjSKRJAQm0C+jSFcLV1gh0rOtBtgT/TKKWZEWP0Tfc0ICK6iXUbjkZ1uwgUoKsIF2\nGUW65vZgAyKBAbSzv10vmBvD4uEqnJGgUSqUUcinqKw9IkLU7tlmA+xIhTKKoiIBmKBCGYXQ\n2ZPaI4JIYJPGcu0AbIJIAAVoTKQ/CTqNAngEIgEUQK2MIggquGlBME3bI4JIYBO1MoogHD4X\nzIvbI4JIYBO1MoogJ6isSNySBQOolVGkms10EBeLRJIQWECtjGIadSJIugumK8oo/vQPAH+2\n4QC7oVZGEYw6EWUPbUgR4owEZlAro0jFil4hEjSOWvY3IsGRqSZSqa4dUTuwgFoZhfQy/Mna\nI8J9JLCJXhmFlNkQz4/bI4JIYJPGcu0AbIJIAAVoTKS0a0f/DmyASAAF0CujGGanQYZNwQZE\nAhtojkYRjT5R5j4SIoENFEejiEefKCsSd2WhLnqjUQyzg9dxBkTSHpEZkcgTgsrojUYxK9KW\nMgpEAhvolVGEP+GqcY4QXTtoEq0yiiD1Lu8FbhaJMxJURiv724VhcESCo6E4GsWdYANdO2gc\nxdEo9hQJoC6Ko1EEnTxpmrZHBJHAJo3l2gHYBJEACtCYSPTrwCaIBFCADSK5ZJrOT17OlVPE\nqyESNMkWkVw4CeaLL4P3UTlFsgdEgiZRFWkIhyc5Q1vPSB3P/4babOrauen34nKKcYYr0bVD\nJKhLGZGkDIckiyj8LYu0roxiXdsBylJeJCnnLhBp7NYFrzefkRAJKrMtarelnMKnniXtEZFF\nwiOozf4ihev7qJzCJe/j9oggEthEW6TglbQEkaBNNt6QDaMFSbDhQTnFTJSc+0jQJGVEWlFO\nkWU6pO0RQSSwSWO5dgA2QSSAAjQmEh07sAkiARRArYzCD8GFos9sWNpmgH3Qy/4OI3TCLaes\nPSKIBDZpXyRKKMAAamUUUUewoEg8ZBUsoFZG4UKhRJHWlFFcq/qWNhygOGplFKlrpYINnJHA\nApplFGNXj64dHA3F7G8fXSghEhwJNZHurELUDppHrYwiWLusSAAG0CujeDRN2yOCSGCTxnLt\nAGyCSAAFaEwkenVgE0QCKIBeGUW6RVBVsepxXEubC7Ajmg/RH591EobBk/UQCZpEUaRh1Il0\nNIpotYUi2eiPAuiNRpHehO1fB5O4PSKJSGQ1gBH0RqOYF6l/u7CM4tpkRAIbqJZRRGlC0zTs\nAC4RiXxVMINWGYWYWrdVJIbqAytoZX8Po06ko1H42COCDdAmqvVI8ZlIvA+FSNAkiqNRpDdk\nw/0I7RHhPhLYRHE0ilikqYuHSNA+jeXaAdgEkQAK0JhIf7aj8z3gl4FIAAVYIZIYto7fO2FJ\n8oyGMdiw6ZkNiAQ2WCOSCyfBfPHlsOZcakMSuEMkaBIFkeKyifS+UiWRrrlFJBhBKVZ17dz0\n+/nyif51OA0TWLP2iJQSqetTxzEJyrBNJCmjIckauiNSv8pwjbRmNIq1Il2/LymvUIxyIkk5\ndi4862QZDv25zAem6V0jdZRhQEHWRe2Wl0+Mm6aTWsEGRIKS7CdSv3I2rEuiU61gg0ckKMje\nIg1bCKefmiLdSgK5RoJSrLwhG0YRkmDDTPmEi15MC2t17QBKsk2kJeUT2QMhg86f1B4RRAKb\nNJZrB2ATRAIoQGMi0asDmyASQAH0yiiCBAcnTdP2iCAS2EStjGKYzN16ytojgkhgE7UyCh+m\nC1UVqb8Na6MrCwdBs4wiz2uoIVKfGERWA5REs4xiODHNVMzuX0YRfFUGsoCiKJZRDAkOXnQv\na49IEZFI/IbiaJZRRNdIaW8wa49IMZEYyAKKollGEYuUbJK1R6SESN3C7wvwGLUyCp+IVC/Y\n0J+LEAkKolZG4a2IBLADemUUw4TMBjggjeXaAdikMZE4x4BNEAmgAIgEUIBqo1GUCTY822aA\nfak2GkWZ8PezbQbYl2qjURQTiVQfMIBaGYUo0LTLvD0iuUikn4IF1MoocpGCa6PrizVlFJ4h\nJcAEamUUmXPTzE3BBs5IYAG1MorphVA+QdcOGketjCIVB5HgSKiPRkHUDo6I+mgUxUUCMID6\naBRkNsARaSzXDsAmjYmUnpFaQefgQT0QSQWdgwf1QCQVdA4e1KOASPnNoOx9FAuv98yGeqw8\ntNAMJURy4SSYP/+yWPi7FdYeW2gFRFLh2vjbE/W4gXxIinTt3PT7cVlFMGt7ilArTAeZlKZj\nUlYkKdMhySbywSqxSPuPRlGP6Bgj0gHZTyQp9248W/k88yFtj0htIdYSHmM8OiJlonbPl1UE\nm/goFy9rj0htIdYSHmJEOiJ6ImWb+GFmuOj4IuHRIVEXKY/WRYodXiQGZjomhW7I9j9SsMHd\ndyy5j3tokeC4lBXpmbKKZBpdOyESNEpjuXYANmlMJP6vB5sgEkABEAmgAPuPRvHg2Q1bgw3P\nthpgT3Z/iH5449V5Ifwdr49I0CSKIsmjUiSrrxGJe5xQnf1Ho+hnjbODVQqJtKTxALuw/2gU\nw3RepH6tdWUUy9oOsA+7j0bh+7KJOP9BEilpjwgigU32Ho1i2CT2qrBIeAS1UXmIvnfTqBQ+\nWTVZE5GgSXYXKQ0q5NdIiATts/toFBoiAdRm/9EoHo1KgUhwABrLtQOwSWMicTICmyASQAEQ\nCaAA+mUU423YzWUUzzYYYHfUyyiGSRT5RiRoHPUyijhdKFt9qUhUUIAJ1MsovHAvdotIz7Ya\nYE/Uyyh8LlI/e3EZheeEBFZQL6PwwSIhtWGpSJyTwAT6ZRRB1M5vFmlJuwH2o0YZhQt3tSHX\nznNGAiuol1FEszaLxIisYAP1MgpfViQAE+iXUQTXSPklEiJBmzSWawdgk8ZE+iOj0zaAWRAJ\noACIBFAAvTKKcBoGI9bfkEUkMMO8SN/nl647/c03WZf9PeUFxfOTHSESNMmsSF+uu+K+0k0K\ni+QKnZG4NQsVmRXp1J1+FPo6dW/pJivLKIJFPp5fRCSShaAmsyL1/8F/5//RryyjGBblgvVv\n141GgUhggVmRXrvv2+xTusm6MgoflFNIKUNZe0QQCWwyH2x4O31eunan/BppsuieSC7PDUqn\nRaN21PhBRe507ULCJU+KlGwiTKU4OmckaJLdRZpbJfCHqB00j1oZRbTqsJtph3l7RO6IBFAR\nvTKKZCpWzCISNIos0k83Se7WKbYHoCEaE4lTEdik8exvnUYBPAKRAAowL9LZPejaSbeBkvfh\nosdBh7g9IogENpkV6fzwGmlbFriYi4dI0CizIrlOKEWK1ygmUrDacpG4EQsGeJT9fYcC5RQF\nRCI1CCxwp2v3/WDTjeUU0/t+7poyCnJVwQbzwYbXPO87Zm05RZgXMb7N2yPCGQlsMndDdiZh\nNeTJ5FWhnCLKKkIkOAD7i5Rs4sNZSfaqRyRolA03ZJeJ5OJf2fusPSJE7cAm20QafqRgw9yo\nFFH3zvvNIgEYYF6kfz9/4B+dm7+b5MKfp8spxmum7aNRIBKYYVakfz9dpuuz7R7dl92rPQAN\nMSvSS/fx8/Pvs3Nek3WFfU+i8xXgN3Ivs+G9e1G/lkckaJI7uXZfb93n5SqpUntEEAlsMivS\n38tzvy8npHO6SZrvk85PFiVBhjTYsK2MApHABnfqkTr3/vN3nnm0MuvbJ1NxPiJBo6x5itBq\nkdJcu+3Z32tE6q/6uI8LBVn1OC43/V5UPmFCpD6niIQIKMmKUvPV5RN3s4U2jUax6IzUu4RI\nUJAVpeZLyyfEE5IgUtIekSIidaS6QmnuhL//zWziJovuiRRE4wyKpP7APjg2K0rNnxQp2iCY\nhPZE/b+kPSIlROrE7wuwhYcDjeUsFUmIffuKIg3nIkSCgtwZjHmu1NxNP1KwISmfmDwxIhLA\nDjwxPlK6iQt/HpZPTNdKyZmpSBkFIoENVoik1B6AhuDZ3wAFQCSAAsyL9H1+6bqX86PHRO7W\nHhFEApvcidrdrpBcFrtLw9np/GTRgyEwo5w8RIJGmRXprbuEv79O3Vu6SZmH56fTrD0iiAQ2\neZjZIIS/bYlEqg8YYI1IU9LqilEofLCK0EvkSavQJGu6dpNIUmZDkj3kZ6b5ckajgHZZFWwQ\nRZJy7YYkBh9O0yDDtmADZySwwIrwt5ssuidSXkYRpd3RtYMjsarU/CmRog2iKSLB4VAQKenY\nzWyStUeEqB3YZG58pOeyv6Vgw1wZRdjNKykSgAG2ibSgjCIbhYLMBjgQjSWtAtikMZHSMxKn\nJbDBw8yGLBNoXxAJmkQWyXXd/DWSUntEEAlsIov0L/BofLxdftMnex8uSu/HCsGF7cEGRAIb\nLHiu3dqH5wePPonD3c5F6yXtEUEksMmCYMNKkZyfEcn5ciJxVxbqcm98pOQaaXX5hJwaFGY/\nbBWJPCGozIKH6G8eheKBSKtHo0AkqM6dh+h/nrqv71P3Mc6RRZJy7KIowziJromm7KESwQa6\ndlCXe8GGv927/+5Owxw3WXRPpPlRKMJV40sjunbQOPdEer+EvsOu3VMiTbvI4+WBSNEgzIgE\njXNnNIr/vroX/7FepPh8I4S/idrBYZgV6WLQ6RJrGJ/Z4KYfKdgwOwqFhkgAdZkPf7+/XJ6A\n0p3HGS78WTAKxRRTcOF0WEBmAxyAxrK/AWzSmEich8AmcxWy9rO/ddoD8BSIBFCAJUmryTSd\nnyySns3gZuZL7RFBJLDJ7tnfaTh87hZU1h4RRAKb3BHp3+vlVtLn+N6iSDYiJACzIn2/XK+P\nOjFpdUEZRSSMnxFJao9IIhIJDWCEO6NRnC9/p/+FSaujSFJmQ5I95INVrtPhEimdP2Z/Lyyj\nuDYZkcAGd0vNh58b68oopFWjrKI4nWiJSKSqghkWieRXlFGkZygpq2itSOrBeYAZHnXtzlHS\n6ooyikykZEqwAY7AfLAhG2hsqUhu+n1/E0SC5rkT/v6bDDTmpp/ZC55AjKB7N67qS4sEYIOF\nmQ3jz/NlFHOZDGQ2wIF4LNJfnYb0UEYBTTIj0sdL93K9E/v5YjJpVacxAM8ii/RxjTN8/pyO\nfi6TKrVHBJHAJrJIr935Gvg+dZ1uzw6RoE3m6pG+/XfXnbqXKWc1uwmUzk8WPXpGQxqMSNsj\ngkhgk9kxZK+JA+dw1Y3Z32E2kBAez9ojgkhgk7sivUerbhUpGX1is0ikB4Ed7ooUr7qtjCK8\nKevT6RqRSFkFQ6wTScpsSLKHfLBKLFJww9aH6y8po8ibDFCT7SJJuXZBty1LERKyjFZeI+ER\nmGHBU4TcZNE9keQyCpe8D3bp14qESWCG8iJNe4niDINgxUTiGgkMsSxpdUMZxTCZ2zRrjwhR\nO7DJuuxvKdjwoIwimu2iva3u2gFYQa+Mwg/SlCijeOq7AajR2EP0AWzSmEgU9oFNEAmgAIgE\nUADd0SjSqcs2QyRoErXRKLL7RsENJ0SC1jEgkttwRrLRHwVY1LVz0+/lZRTJ+2ydtD0iiUgk\nN4AR1okkZTYk2UM+WEUUaXRxWMxoFNAu20WScu2ElKD0fZoqlLVHJBKJtFUwQ/mk1TujUWS5\ndS5+6xeKxGgUYIUds78DSZz4/rq/UDxPsAEaRbGMIn6fnqmy9oggEthEr4wieR9eG60XCcAG\namUUs6NTeESC9mks1w7AJo2JxMkIbIJIAAVAJIAC1C2jCKdpe0QQCWxSLftbup/kEQkapVWR\nyA0CU6iVUURTn0wXi0S+KthCrYximg6jUKwfjeK59gLooVZGMU5dMvVRb/FpkfAILKFWRjF7\nbTRkGaXtEUEksIlaGcXdIMOKYMOj1gJoolZGUTZqR0kf2EKtjKKsSAC20BuNgswGODCN5doB\n2KQxkf7I6LQNYBZEAigAIgEUYIVI+c2f7L0TFqSjUGwa+hKRwBZrRHLhJJgvvhxWfBQWz9oj\ngkhgk/1FCp+muqdI3KGFiqzq2rnp9zPlFHmqUPq+gEgUVkBNtokkZTgkWUQ+FWm4RAo2vf1a\nOBoFIoEhyokk5dwlIjnBvVLBBkSCmqyL2i0rpxijdlG/LukNZu0RuXuNhEhQjf1EijZIXrlw\nV76ISJyRoCZqIt1ZlagdNM/KG7L9jxRsSMspBJHkYEPSHpE7IgFUZJtIT5RT+OAayd2bpu0R\nQSSwSWO5dgA2aUwkTkBgE0QCKAAiARRgg0jpTaF0frZoPuiwMGq3tKUAe7NFJBdOgvniS5+U\nU0x5QzPtEUEksImeSC6/n7RBJG6/gik2de3c9HtdOYVLPWI0CmiTMiJJGQ5JFpGfE2kUjdEo\noF3KiyTl3AUijQ+QTFbN2iMSpNYtbS7AjmyL2q0opxAFQiRonP1FijYoJhIegSmqi7Sua8do\nFGCLjTdk+x8p2PBEOcW4idQeEe4jgU3KiFSinCJtjwgigU0ay7UDsEljIlHOBzZBJIACIBJA\nARYOfRlO04XXXDsAAA60SURBVPnZojGWMDf6xNpgAyKBNZaI5MJJMF98Ga0o3zeKUoWy9ogg\nEthkP5Gm56r6UJjwZlIBkbgxCyZY1LVz0++nyybSF6M4zhcR6VGrATRYJ5KUyZBkC/lJD7GL\nl4u0eDQKzwkJrLBdJCm3ThQpL5/YHmygwA9ssCxqt2YUCvmMFJ6YtgUbEAkMUF6kaINgEq/q\nBuG2icQZCWygJpIQ/i4TtXvUbAAFFt6QDaIEabBhpmxiZ5EATLBOpCVlE0GsIQ0uLA82ANik\nsVw7AJs0JtKfh+g0EyAGkQAKgEgABVg59GU4Tedni/qw3VRNMb3wC4MNiAQ2WSOSCyfBfPHl\nsGIU5nY+jumtC38jEphhf5GSUSjuvfblRLp8G27WghqrunZu+r1sFAo/zYjvJQntEXlWpK4j\nfQg02SaSlOGQZBH58WUiVvIw/RVlFI/OSIgEapQTScq5C0Vyyfvh1z7Bhm7RVwLYyLqo3YZR\nKMbfySZZe0Q4I4FN9hMp2iC3ySebZO0RWSIS10igRzWR0k2y9ogsOiMRtQM1Vt6Q7X+kYMOD\nUSjCmXGwIWmPyNMiAaiyTaQ1o1AM1qwajQLAJo3l2gHYpDGR6M2BTRAJoACIBFAA3TKKJD7h\n8rURCZpErYzCh9ZE4XGHSNA8amUUsyK5Emck7r1CXdTLKISMhu0ikQ0EldEro0jKJnKRVpdR\nIBJUR6+MIljFiSIl7RGhawc2USujGF8lvUBfQiTOSFAZtezv6BUiwcFQL6PYJdhA1w4qo1pG\ncT/YkLRHZE4kgLrol1FsymwAsEljuXYANmlMJHp1YBNEAigAIgEUYOHQl+E0nZ8tmoIMSVxi\nJugQt0cEkcAmS0Ry4SSYL770D8onhDB43B4RRAKb7CfSg/KJYiJxLxYMsKhr56bfq8snfLBq\nkgERt0ckF4nsILDAOpGkjIYka2j4nY864RORBvHWlFFcTkh4BPXZLpKUY5ekBsVnqkmsq4Oj\nZUl7RDgjgU2WRe02lE9k/by8nCJujwgigU3KixRtEG7qY5Gk+YgEjbK7SI+idUTt4AgsvCHb\n/0jBhoXlE8VEAjDAOpFKlE8wGgUciMZy7QBs0phIadeOnh7YAJEACoBIAAWoPxrFpjIKRAIb\nVB+NYlv4G5HABtVHo9hDpMtX4T4taGJhNIrSInUdmUOgjIXRKPrp6tEoxDMSIoEmVUajiM9Y\n5YMN3aLvA7Ad9dEoFLp2iATqqI9GoSUSHoEm1Uej2Clq11GCDppUH41iD5EAtKk/GgVlFHAA\nGsu1A7BJYyLRqwObIBJAARAJoAB6ZRTDi53KKJ5tPcAe6JVReC/m2pUKfz/beoA90CujGHP0\n/J4iUUEBVVAro4iyHYLeYVmRqKCAOqiVUTwWaVsZxfQtEAn0US2jSMMUU36E0B4RRAKbaJVR\nRD24aDpkF6XtEUEksIlW9rcT6pNctEbaHhFEApuo1iNJHpWP2lFAAfqolVGML5LzUWGRAKqg\nV0YRbOHCayjKKOAANJZrB2CTxkSiLwc2QSSAAiASQAF0R6MIpuuCDYgENlEro0hvNXnpPSJB\no6iVUUgiZblDi0Wy0R8F0CujCBYFb7eJRA4DGEGvjGJcJAz3cv21sIzi2mREAhuolVF4P8QW\nkvnrgw1kp4IZqo1G4cOX67p2HQ/4BitUG40itylpjwjBBrBJtdEoEAmOhOpoFOVFArBBvdEo\nghjE88EGAJs0lmsHYJPGRKJrBzZBJIACIBJAARaI5JJpOj9bFAUZsrKJdH7aHhFEApssEcmF\nk2C++NIn5RPD8jwMHm2HSNAk+4nksltGY9A8FWllrt0zjQZQYVHXzk2/F5dPDBPpTLRaJFLt\nwAjrRJIyGpKsIZ+LJHb1pneUUUC7bBdJyrFLRBIzHNL5aXtEKKMAmyyL2q0snwiFEa6NVotE\nGQVYobxI0QbhNO3fSR4RbIA2URBJuCCaiX17RIJGWXhDtv+Rgg1zo1AkHTwX7ipeIW6PCPeR\nwCbrRFpQPjFdMw1z04yGJSIB2KSxXDsAmzQmEv06sAkiARQAkQAKUG00Cu+kpDxEgjapNxqF\nS1ZI2yOCSGCTeqNRuGSFtD0ipH6DTeqMRuEij9aJRMoqGKLCaBT9NLgfe1thSRnFwiYD7Izi\naBRBjy51UGqPCIl2YBPV0ShCkfoZaZIQZyRoEs3RKFz86zp1654i5LlGAlMoihScnuKuH1E7\naB610SgiaZJro7VdOwArqI1GMfeAyHGFtD0ADdFYrh2ATRoT6U+ITosAngCRAAqw/0P0x2BD\nck0kZIF7RIJG2f0h+j7J/h5mjO/j3SESNMnuD9H32cP0nS8oEo+IBBNoPEQ/Fim8h+RSj5aL\n9KjVABpoPET/vkijeAsfon/LE+KEBCbQeIj+5FAgUepg1h6RTCTOSWACjYfoj1G7+HrJuWSN\nuD0iQrABkcAAig/Rv76Ks703i8QZCWygJlLYhcuCDevPSFwjgQn2f4j+EyIF8nEfCZpk94fo\nj9N7mQ1Se0QQCWzSWK4dgE0QCaAAjYlEnw5sgkgABdAvo0ima54i9KilANqol1GI948QCRpH\nvYxCEmlp+NtzIxaMoV5GEZ+04hX90yKRGgS2UC+juKbb+XDJsN7zZRTPtRdAD7Uyiim4MC3J\nT0jPXyPhEVhCrYzizrURIkHzqGd/52es6Krp6WDDo9YCaFJRpDz27Z8PNlBjDqZQL6Nw4a6i\n1ZL2iHAfCWyiX0aRZTQgErRPY7l2ADZBJIACNCbSn/votBEgA5EACrByxL5wms7PFuVBBydM\n0/aIIBLYZI1ILpwE88WXPimnyDyMd4NI0CT7i5SUU2iIdPkq3LAFTVZ17dz0e105hTDN2iPy\nlEhdR50FKLNNJCnDIcki8qlIs5dIi0ejuHtGQiTQpJxIUs7dE6lCSXJDOZHwCBRZF7XbMiqF\nNAeRoHH2EynaIJqqiIRHoIm6SEpdO+osQJWVN2T7HynY8EQ5hfSTtUfkOZEAtNkmUoFyig2j\nUSASmKGxXDsAmyASQAEaE4luHNgEkQAKUH00itXBhkfNBVCk+mgUq+8jPWougCLVR6PYJpKN\n/iiA2mgUPlhlnBlnQMTtEUlEInsBjKA2GoUfV3HTc/SD94vLKK5NRiSwgd5oFINASVJEODqF\nX3ZGongPzKA2GoWf32S1SGSmghXUHqLv5zch2ADNU200CkSCI1FtNIoyIgHYoNpoFGQ2wJFo\nLNcOwCaIBFCAxkT6A1CHp/9wEQlgnqf/cLVGo5gLPqTtEal9NOHX8sAK/dEoZsoqsvaI1D6a\n8Gt5YEWN0SgQCdpj9GR8snyUn6Y+GoX3gkhSexAJLDFp0o8aFOdMq49GkYtUcDQKgL2Y/gi7\nWwHPXiJJOXcPRLq654IlnJHAKrFHZUR6Knl1vpxijNpxjQTNEIl0LeCJinhqjEbh4l0hEjRA\n4oneGSnaIJ8StYOmGDTpz0NFonbjjxRseFBOgUjQJA+s0B+NIrhGcj5aL26PSO2jCb+WB1Y0\nlmsHYBNEAihAYyItPN8CKIFIAAXYfzQK6RkNYRRiU7DhUZsBdNh9NIoszO3iFbaFvx+1GUCH\n3UejmBfJ+SIi8bBVMMDuo1GIyQ/h6uNKSXtEcpF4ADhYYPfRKDKRRgdTkdaUUXjP47/BAruP\nRnHHOc5IcBh2H41iujZybhQmuH5CJDgCuz9EPxIlEGkQDpHgCKiJlEXvfJkzElE7sMDuo1FI\nIrlkj5tEAjDA/qNR3Bl9In3vEQkapbFcOwCbIBJAARoTiV4d2ASRAAqgNxrFvaEv10Ttnm0u\ngAJqo1HM3k+Kd4NI0CRqo1HsI5KN/iiA2mgUkUDC+6w9IolIZDWAEdRGo5gWuagydrpEWlhG\n4fvHxgIYQG00ikkcL7qXtUeEfFWwidpoFPm1UVRWkbdHJBapo6oPjKD2EP1MJGlKsAEapZpI\noliIBI2iNhrFPZEC6biPBE2iNxrFvcwGqT0iiAQ2aSzXDsAmiARQgMZE+iOj0zaAWRAJoAD6\nZRRzwYe0PSKIBDZRL6NI36++j4RIYAj1Mor0fTGRyBaCiqiXUfhQmzgDIm6PyLxIZLBCTdTL\nKHwm0soyCkQCQ6iXUYy/h7qkOEcIkaBJ1MsofKhN2WskRIJqqGd/+51E4owENakmElE7OBLq\nZRTp+2IiAVREv4wifU9mAxyAxnLtAGyCSAAFMCfSff73YPneVP58vr7djx//hG2I9ID//e7P\n5+s38PGIZP/z+foNfDwi2f98vn4DH49I9j+fr9/AxzchEoB1EAmgAIgEUABEAiiAfZHcwO/8\nfL5+Gx9vXiQnvPpFn8/Xb+TjEcn25/P1G/l4RLL9+Xz9Rj4ekWx/Pl+/kY83L1Lty83an8/X\nb+Pj7YsE0ACIBFAARHqA8IQJ3c+v3LcZ2lDn4yt//QXfHpEeMDzwJR8+QO3j01fan/97v/6C\nb49IDxiPpfm40W6f/3u//oJvj0gP+N1/Sb/86yNSOYZn9/3Ov6Rf/vUXfHtEesz1YtP+nYwd\nW/CLv/6z3x6RAAqASAAFQCSwzNirq3aJyDUStM90E6derOO5j0ck2zhX93K78ue7cWL94xHJ\nOPWidRY+343TqtH3Jz4ekazzq00aP7xyhhIpQtA2bjop2P54RAIoACIBFACRAAqASAAFQCSA\nAiASQAEQCaAAiARQAET6hfyrnS1xQBDpF9Lxr14cDukvBJHKwyE9FF+vnTtfX7x13duXH6S5\n/O66fmnXYVJxOKJH4ttdJHkdXrjvWKTrzDMi7QFH9Eicuzf/cbHk3J28P3XnWKTTt//XObp2\ne8AhPRIv3ffw4qdb99W9JF278VXFNh4UDumRGA0J9Zl7BSXhkB4JRKoGh/RIzHXtvhBpbzik\nR+Lcnf1nHGxw3X/++4RIe8MhPRJf1wD3Sxj+Pl9e/E1FIkeoNIh0KD5P/X3Y8YasP7vub3KN\n9A+RioNIAAVAJIACIBJAARAJoACIBFAARAIoACIBFACRAAqASAAFQCSAAiASQAEQCaAA/w9r\nVJCYeBicPQAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ggplot() + \n",
    "  geom_bar(aes(x = module_chr12_total$community, y = module_chr12_total$Size),\n",
    "           stat=\"identity\", fill = \"lightblue\")+ \n",
    "  labs(x=\"Relationship\",y = \"count\")+\n",
    "  ggtitle(\"numbers of genes in each community(module) at chr12\") + backgroup+\n",
    "  geom_text(aes(x = module_chr12_total$community, \n",
    "                y = module_chr12_total$Size,\n",
    "                label=module_chr12_total$Size), hjust=-0.5, size = 2)+\n",
    "  coord_flip() #将条形图坐标轴转换过来变成横向条形图"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 id='6'>6. Annotate all modules with the genes they contain in their bins<h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='6.1'> learn how to get the genes </h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before we run to get genes, we should learn string function `split()` and `substring()`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Example is a chr12 node, we use `split()` to get a list of details property of this node.   \n",
    "\n",
    "`split()` returns a list of strings resulting from the splitting of the original string around matches of the given delimiter.\n",
    "\n",
    "Syntax: `split(original, splitDelimiter)`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 1 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Details</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>3791047                          , 3873448                          , gene_id ENSG00000111224.13       ,  gene_type protein_coding        ,  gene_name PARP11                ,  level 2                         ,  havana_gene OTTHUMG00000156442.5, |3839459                         , 3842525                          , gene_id ENSG00000118976.5        ,  gene_type processed_pseudogene  ,  gene_name HIN1L                 ,  level 1                         ,  tag pseudo_consens              ,  havana_gene OTTHUMG00000168119.1, |3871734                         , 3910338                          , gene_id ENSG00000256862.1        ,  gene_type antisense             ,  gene_name RP11-664D1.1          ,  level 2                         ,  havana_gene OTTHUMG00000168120.1,                                  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 1\n",
       "\\begin{tabular}{r|l}\n",
       " Details\\\\\n",
       " <list>\\\\\n",
       "\\hline\n",
       "\t 3791047                          , 3873448                          , gene\\_id ENSG00000111224.13       ,  gene\\_type protein\\_coding        ,  gene\\_name PARP11                ,  level 2                         ,  havana\\_gene OTTHUMG00000156442.5, \\textbar{}3839459                         , 3842525                          , gene\\_id ENSG00000118976.5        ,  gene\\_type processed\\_pseudogene  ,  gene\\_name HIN1L                 ,  level 1                         ,  tag pseudo\\_consens              ,  havana\\_gene OTTHUMG00000168119.1, \\textbar{}3871734                         , 3910338                          , gene\\_id ENSG00000256862.1        ,  gene\\_type antisense             ,  gene\\_name RP11-664D1.1          ,  level 2                         ,  havana\\_gene OTTHUMG00000168120.1,                                  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 1\n",
       "\n",
       "| Details &lt;list&gt; |\n",
       "|---|\n",
       "| 3791047                          , 3873448                          , gene_id ENSG00000111224.13       ,  gene_type protein_coding        ,  gene_name PARP11                ,  level 2                         ,  havana_gene OTTHUMG00000156442.5, |3839459                         , 3842525                          , gene_id ENSG00000118976.5        ,  gene_type processed_pseudogene  ,  gene_name HIN1L                 ,  level 1                         ,  tag pseudo_consens              ,  havana_gene OTTHUMG00000168119.1, |3871734                         , 3910338                          , gene_id ENSG00000256862.1        ,  gene_type antisense             ,  gene_name RP11-664D1.1          ,  level 2                         ,  havana_gene OTTHUMG00000168120.1,                                   |\n",
       "\n"
      ],
      "text/plain": [
       "  Details                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
       "1 3791047                          , 3873448                          , gene_id ENSG00000111224.13       ,  gene_type protein_coding        ,  gene_name PARP11                ,  level 2                         ,  havana_gene OTTHUMG00000156442.5, |3839459                         , 3842525                          , gene_id ENSG00000118976.5        ,  gene_type processed_pseudogene  ,  gene_name HIN1L                 ,  level 1                         ,  tag pseudo_consens              ,  havana_gene OTTHUMG00000168119.1, |3871734                         , 3910338                          , gene_id ENSG00000256862.1        ,  gene_type antisense             ,  gene_name RP11-664D1.1          ,  level 2                         ,  havana_gene OTTHUMG00000168120.1,                                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query = \"MATCH( chr12:chr12{Name: 'Bin20'})\n",
    "RETURN split(chr12.Details, ';') as Details\"\n",
    "cypher(graph,query)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to you can easy to see that this result is list; we use `UNWIND` to  transform this list back into individual rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 23 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>details</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>3791047                          </td></tr>\n",
       "\t<tr><td>3873448                          </td></tr>\n",
       "\t<tr><td>gene_id ENSG00000111224.13       </td></tr>\n",
       "\t<tr><td> gene_type protein_coding        </td></tr>\n",
       "\t<tr><td> gene_name PARP11                </td></tr>\n",
       "\t<tr><td> level 2                         </td></tr>\n",
       "\t<tr><td> havana_gene OTTHUMG00000156442.5</td></tr>\n",
       "\t<tr><td>|3839459                         </td></tr>\n",
       "\t<tr><td>3842525                          </td></tr>\n",
       "\t<tr><td>gene_id ENSG00000118976.5        </td></tr>\n",
       "\t<tr><td> gene_type processed_pseudogene  </td></tr>\n",
       "\t<tr><td> gene_name HIN1L                 </td></tr>\n",
       "\t<tr><td> level 1                         </td></tr>\n",
       "\t<tr><td> tag pseudo_consens              </td></tr>\n",
       "\t<tr><td> havana_gene OTTHUMG00000168119.1</td></tr>\n",
       "\t<tr><td>|3871734                         </td></tr>\n",
       "\t<tr><td>3910338                          </td></tr>\n",
       "\t<tr><td>gene_id ENSG00000256862.1        </td></tr>\n",
       "\t<tr><td> gene_type antisense             </td></tr>\n",
       "\t<tr><td> gene_name RP11-664D1.1          </td></tr>\n",
       "\t<tr><td> level 2                         </td></tr>\n",
       "\t<tr><td> havana_gene OTTHUMG00000168120.1</td></tr>\n",
       "\t<tr><td>                                 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 23 × 1\n",
       "\\begin{tabular}{r|l}\n",
       " details\\\\\n",
       " <chr>\\\\\n",
       "\\hline\n",
       "\t 3791047                          \\\\\n",
       "\t 3873448                          \\\\\n",
       "\t gene\\_id ENSG00000111224.13       \\\\\n",
       "\t  gene\\_type protein\\_coding        \\\\\n",
       "\t  gene\\_name PARP11                \\\\\n",
       "\t  level 2                         \\\\\n",
       "\t  havana\\_gene OTTHUMG00000156442.5\\\\\n",
       "\t \\textbar{}3839459                         \\\\\n",
       "\t 3842525                          \\\\\n",
       "\t gene\\_id ENSG00000118976.5        \\\\\n",
       "\t  gene\\_type processed\\_pseudogene  \\\\\n",
       "\t  gene\\_name HIN1L                 \\\\\n",
       "\t  level 1                         \\\\\n",
       "\t  tag pseudo\\_consens              \\\\\n",
       "\t  havana\\_gene OTTHUMG00000168119.1\\\\\n",
       "\t \\textbar{}3871734                         \\\\\n",
       "\t 3910338                          \\\\\n",
       "\t gene\\_id ENSG00000256862.1        \\\\\n",
       "\t  gene\\_type antisense             \\\\\n",
       "\t  gene\\_name RP11-664D1.1          \\\\\n",
       "\t  level 2                         \\\\\n",
       "\t  havana\\_gene OTTHUMG00000168120.1\\\\\n",
       "\t                                  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 23 × 1\n",
       "\n",
       "| details &lt;chr&gt; |\n",
       "|---|\n",
       "| 3791047                           |\n",
       "| 3873448                           |\n",
       "| gene_id ENSG00000111224.13        |\n",
       "|  gene_type protein_coding         |\n",
       "|  gene_name PARP11                 |\n",
       "|  level 2                          |\n",
       "|  havana_gene OTTHUMG00000156442.5 |\n",
       "| |3839459                          |\n",
       "| 3842525                           |\n",
       "| gene_id ENSG00000118976.5         |\n",
       "|  gene_type processed_pseudogene   |\n",
       "|  gene_name HIN1L                  |\n",
       "|  level 1                          |\n",
       "|  tag pseudo_consens               |\n",
       "|  havana_gene OTTHUMG00000168119.1 |\n",
       "| |3871734                          |\n",
       "| 3910338                           |\n",
       "| gene_id ENSG00000256862.1         |\n",
       "|  gene_type antisense              |\n",
       "|  gene_name RP11-664D1.1           |\n",
       "|  level 2                          |\n",
       "|  havana_gene OTTHUMG00000168120.1 |\n",
       "| <!----> |\n",
       "\n"
      ],
      "text/plain": [
       "   details                          \n",
       "1  3791047                          \n",
       "2  3873448                          \n",
       "3  gene_id ENSG00000111224.13       \n",
       "4   gene_type protein_coding        \n",
       "5   gene_name PARP11                \n",
       "6   level 2                         \n",
       "7   havana_gene OTTHUMG00000156442.5\n",
       "8  |3839459                         \n",
       "9  3842525                          \n",
       "10 gene_id ENSG00000118976.5        \n",
       "11  gene_type processed_pseudogene  \n",
       "12  gene_name HIN1L                 \n",
       "13  level 1                         \n",
       "14  tag pseudo_consens              \n",
       "15  havana_gene OTTHUMG00000168119.1\n",
       "16 |3871734                         \n",
       "17 3910338                          \n",
       "18 gene_id ENSG00000256862.1        \n",
       "19  gene_type antisense             \n",
       "20  gene_name RP11-664D1.1          \n",
       "21  level 2                         \n",
       "22  havana_gene OTTHUMG00000168120.1\n",
       "23                                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query = \"MATCH( chr12:chr12{Name: 'Bin20'})\n",
    "with split(chr12.Details, ';') as Details\n",
    "UNWIND Details AS details\n",
    "RETURN details\n",
    "\"\n",
    "cypher(graph,query)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we use `substring()` to get gene names. In Details property, gene names always use `gene_name ` for tell use what its genes contained in this nodes.\n",
    "\n",
    "`substring()` returns a substring of the original string, beginning with a 0-based index start and length.\n",
    "\n",
    "Syntax: `substring(original, start [, length])`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 3 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>PARP11      </td></tr>\n",
       "\t<tr><td>HIN1L       </td></tr>\n",
       "\t<tr><td>RP11-664D1.1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 1\n",
       "\\begin{tabular}{r|l}\n",
       " genes\\\\\n",
       " <chr>\\\\\n",
       "\\hline\n",
       "\t PARP11      \\\\\n",
       "\t HIN1L       \\\\\n",
       "\t RP11-664D1.1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 1\n",
       "\n",
       "| genes &lt;chr&gt; |\n",
       "|---|\n",
       "| PARP11       |\n",
       "| HIN1L        |\n",
       "| RP11-664D1.1 |\n",
       "\n"
      ],
      "text/plain": [
       "  genes       \n",
       "1 PARP11      \n",
       "2 HIN1L       \n",
       "3 RP11-664D1.1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query = \"MATCH( chr12:chr12{Name: 'Bin20'})\n",
    "with split(chr12.Details, ';') as Details\n",
    "UNWIND Details AS details\n",
    "with details as details\n",
    "where details contains 'gene_name'\n",
    "with details as genes\n",
    "return substring(genes,11) as genes\n",
    "\"\n",
    "cypher(graph,query)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='6.2'> Get genes in each modules </h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, the node chr12 `Bin200` includes 3 genes.\n",
    "\n",
    "###### Now, we can use that functions to do our tasks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"\n",
    "// to make louvain modules for chr12\n",
    "CALL algo.louvain.stream(\n",
    "  'MATCH (n:chr12)--(m) with id(n) as chr12,id(m) AS m   MATCH (al) where id(al) = m or  id(al) = chr12 return id(al) as id',\n",
    "  'MATCH (n:chr12)-[r]-(m) RETURN id(n) AS source, id(m) AS target',\n",
    "  {graph:'cypher',direction: 'both'})\n",
    "YIELD nodeId, community\n",
    "\n",
    "// filering community which the number of nodes is more than one\n",
    "with collect(nodeId) AS nodeId , community\n",
    "where size(nodeId) > 1\n",
    "with nodeId, 'Module'+ community AS community\n",
    "\n",
    "// to get communities that contain chromosome 12 nodes. \n",
    "MATCH (chr12:chr12)\n",
    "where id(chr12) in nodeId\n",
    "\n",
    "// split details\n",
    "with split(chr12.Details, ';') as ann, community\n",
    "\n",
    "// then ann is list, we need unlist ann for strings search\n",
    "UNWIND (ann) as ann1\n",
    "with  ann1 as ann1, community\n",
    "\n",
    "// now we want to find out which one terms including gene_name in ann1\n",
    "where ann1 contains 'gene_name'\n",
    "with ann1 as genelist, community\n",
    "\n",
    "// And we can extract the text about real gene names\n",
    "return distinct substring(genelist, 11) as genelist, community\n",
    "\"\n",
    "\n",
    "module_chr12 = cypher(graph, query) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>genelist</th><th scope=col>community</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>DDX11L8           </td><td>Module382</td></tr>\n",
       "\t<tr><td>ABC7-42389800N19.1</td><td>Module382</td></tr>\n",
       "\t<tr><td>RP11-598F7.1      </td><td>Module382</td></tr>\n",
       "\t<tr><td>IQSEC3            </td><td>Module382</td></tr>\n",
       "\t<tr><td>RP11-598F7.3      </td><td>Module382</td></tr>\n",
       "\t<tr><td>RP11-598F7.4      </td><td>Module382</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       " genelist & community\\\\\n",
       " <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t DDX11L8            & Module382\\\\\n",
       "\t ABC7-42389800N19.1 & Module382\\\\\n",
       "\t RP11-598F7.1       & Module382\\\\\n",
       "\t IQSEC3             & Module382\\\\\n",
       "\t RP11-598F7.3       & Module382\\\\\n",
       "\t RP11-598F7.4       & Module382\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| genelist &lt;chr&gt; | community &lt;chr&gt; |\n",
       "|---|---|\n",
       "| DDX11L8            | Module382 |\n",
       "| ABC7-42389800N19.1 | Module382 |\n",
       "| RP11-598F7.1       | Module382 |\n",
       "| IQSEC3             | Module382 |\n",
       "| RP11-598F7.3       | Module382 |\n",
       "| RP11-598F7.4       | Module382 |\n",
       "\n"
      ],
      "text/plain": [
       "  genelist           community\n",
       "1 DDX11L8            Module382\n",
       "2 ABC7-42389800N19.1 Module382\n",
       "3 RP11-598F7.1       Module382\n",
       "4 IQSEC3             Module382\n",
       "5 RP11-598F7.3       Module382\n",
       "6 RP11-598F7.4       Module382"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(module_chr12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>3218</li>\n",
       "\t<li>2</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 3218\n",
       "\\item 2\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 3218\n",
       "2. 2\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 3218    2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(module_chr12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(module_chr12, file = \"module_chr12.csv\" ,sep=\",\",quote = F, col.names = TRUE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get `module_chr12.csv` at [here](module_chr12.csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or we process the results showing as list of modules with R."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "module_genelist <- aggregate(module_chr12[1], by = module_chr12[2], FUN = as.character)\n",
    "module_genelist <- as.list(module_genelist$genelist)\n",
    "names(module_genelist) <- paste(\"module\",1:length(module_genelist))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"\n",
    "// to make louvain modules for chr12\n",
    "CALL algo.louvain.stream(\n",
    "  'MATCH (n:chr12)--(m) with id(n) as chr12,id(m) AS m   MATCH (al) where id(al) = m or  id(al) = chr12 return id(al) as id',\n",
    "  'MATCH (n:chr12)-[r]-(m) RETURN id(n) AS source, id(m) AS target',\n",
    "  {graph:'cypher',direction: 'both'})\n",
    "YIELD nodeId, community\n",
    "\n",
    "// filering community which the number of nodes is more than one\n",
    "with collect(nodeId) AS nodeId , community\n",
    "where size(nodeId) > 1\n",
    "with nodeId, 'Module'+ community AS community\n",
    "\n",
    "// to get communities that contain chromosome 12 nodes. \n",
    "MATCH (chr12:chr12)\n",
    "where id(chr12) in nodeId\n",
    "\n",
    "// split details\n",
    "with split(chr12.Details, ';') as ann, community\n",
    "\n",
    "// then ann is list, we need unlist ann for strings search\n",
    "UNWIND (ann) as ann1\n",
    "with  ann1 as ann1, community\n",
    "\n",
    "// now we want to find out which one terms including gene_name in ann1\n",
    "where ann1 contains 'gene_name'\n",
    "with ann1 as genelist, community\n",
    "\n",
    "// And we can extract the text about real gene names\n",
    "return collect( distinct substring(genelist, 11) ) as genelist, community\n",
    "\"\n",
    "\n",
    "module_chr12_list = RNeo4j::cypher(graph, query)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='6.3'> enriched modules in some KEGG pathway or GO term.</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We use `module_chr12_list` to enrich GO and KEGG."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3 id='6.3.1'> Enrich GO </h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Registered S3 method overwritten by 'enrichplot':\n",
      "  method               from\n",
      "  fortify.enrichResult DOSE\n",
      "clusterProfiler v3.12.0  For help: https://guangchuangyu.github.io/software/clusterProfiler\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "Guangchuang Yu, Li-Gen Wang, Yanyan Han, Qing-Yu He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology. 2012, 16(5):284-287.\n",
      "Loading required package: org.Hs.eg.db\n",
      "Loading required package: AnnotationDbi\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "The following object is masked from 'package:grDevices':\n",
      "\n",
      "    windows\n",
      "\n",
      "\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"Warning message in `[<-`(`*tmp*`, i, value = enrichGO(module_gene, \"org.Hs.eg.db\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\""
     ]
    }
   ],
   "source": [
    "library(clusterProfiler)\n",
    "module_chr12_list_enrichGO <- list()\n",
    "for (i in 1:nrow(module_chr12_list)){\n",
    "  module_gene <- unlist(module_chr12_list$genelist[i])\n",
    "  module_chr12_list_enrichGO[i] <- enrichGO(module_gene, \n",
    "                                            'org.Hs.eg.db',\n",
    "                                            keyType = 'SYMBOL', \n",
    "                                            pvalueCutoff = 0.05)\n",
    "}\n",
    "names(module_chr12_list_enrichGO) <- module_chr12_list$community"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>Description</th><th scope=col>GeneRatio</th><th scope=col>BgRatio</th><th scope=col>pvalue</th><th scope=col>p.adjust</th><th scope=col>qvalue</th><th scope=col>geneID</th><th scope=col>Count</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>GO:0045505</th><td>GO:0045505</td><td>dynein intermediate chain binding             </td><td>1/10</td><td>10/16969 </td><td>0.005879051</td><td>0.09862894</td><td>0.05956882</td><td>RAN     </td><td>1</td></tr>\n",
       "\t<tr><th scope=row>GO:0000287</th><td>GO:0000287</td><td>magnesium ion binding                         </td><td>2/10</td><td>209/16969</td><td>0.006365894</td><td>0.09862894</td><td>0.05956882</td><td>WNK1/RAN</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>GO:0005086</th><td>GO:0005086</td><td>ARF guanyl-nucleotide exchange factor activity</td><td>1/10</td><td>16/16969 </td><td>0.009391532</td><td>0.09862894</td><td>0.05956882</td><td>IQSEC3  </td><td>1</td></tr>\n",
       "\t<tr><th scope=row>GO:0017081</th><td>GO:0017081</td><td>chloride channel regulator activity           </td><td>1/10</td><td>16/16969 </td><td>0.009391532</td><td>0.09862894</td><td>0.05956882</td><td>WNK1    </td><td>1</td></tr>\n",
       "\t<tr><th scope=row>GO:0140103</th><td>GO:0140103</td><td>catalytic activity, acting on a glycoprotein  </td><td>1/10</td><td>20/16969 </td><td>0.011726977</td><td>0.09862894</td><td>0.05956882</td><td>B4GALNT3</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>GO:0032452</th><td>GO:0032452</td><td>histone demethylase activity                  </td><td>1/10</td><td>22/16969 </td><td>0.012892840</td><td>0.09862894</td><td>0.05956882</td><td>KDM5A   </td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & ID & Description & GeneRatio & BgRatio & pvalue & p.adjust & qvalue & geneID & Count\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\tGO:0045505 & GO:0045505 & dynein intermediate chain binding              & 1/10 & 10/16969  & 0.005879051 & 0.09862894 & 0.05956882 & RAN      & 1\\\\\n",
       "\tGO:0000287 & GO:0000287 & magnesium ion binding                          & 2/10 & 209/16969 & 0.006365894 & 0.09862894 & 0.05956882 & WNK1/RAN & 2\\\\\n",
       "\tGO:0005086 & GO:0005086 & ARF guanyl-nucleotide exchange factor activity & 1/10 & 16/16969  & 0.009391532 & 0.09862894 & 0.05956882 & IQSEC3   & 1\\\\\n",
       "\tGO:0017081 & GO:0017081 & chloride channel regulator activity            & 1/10 & 16/16969  & 0.009391532 & 0.09862894 & 0.05956882 & WNK1     & 1\\\\\n",
       "\tGO:0140103 & GO:0140103 & catalytic activity, acting on a glycoprotein   & 1/10 & 20/16969  & 0.011726977 & 0.09862894 & 0.05956882 & B4GALNT3 & 1\\\\\n",
       "\tGO:0032452 & GO:0032452 & histone demethylase activity                   & 1/10 & 22/16969  & 0.012892840 & 0.09862894 & 0.05956882 & KDM5A    & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | Description &lt;chr&gt; | GeneRatio &lt;chr&gt; | BgRatio &lt;chr&gt; | pvalue &lt;dbl&gt; | p.adjust &lt;dbl&gt; | qvalue &lt;dbl&gt; | geneID &lt;chr&gt; | Count &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| GO:0045505 | GO:0045505 | dynein intermediate chain binding              | 1/10 | 10/16969  | 0.005879051 | 0.09862894 | 0.05956882 | RAN      | 1 |\n",
       "| GO:0000287 | GO:0000287 | magnesium ion binding                          | 2/10 | 209/16969 | 0.006365894 | 0.09862894 | 0.05956882 | WNK1/RAN | 2 |\n",
       "| GO:0005086 | GO:0005086 | ARF guanyl-nucleotide exchange factor activity | 1/10 | 16/16969  | 0.009391532 | 0.09862894 | 0.05956882 | IQSEC3   | 1 |\n",
       "| GO:0017081 | GO:0017081 | chloride channel regulator activity            | 1/10 | 16/16969  | 0.009391532 | 0.09862894 | 0.05956882 | WNK1     | 1 |\n",
       "| GO:0140103 | GO:0140103 | catalytic activity, acting on a glycoprotein   | 1/10 | 20/16969  | 0.011726977 | 0.09862894 | 0.05956882 | B4GALNT3 | 1 |\n",
       "| GO:0032452 | GO:0032452 | histone demethylase activity                   | 1/10 | 22/16969  | 0.012892840 | 0.09862894 | 0.05956882 | KDM5A    | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "           ID         Description                                    GeneRatio\n",
       "GO:0045505 GO:0045505 dynein intermediate chain binding              1/10     \n",
       "GO:0000287 GO:0000287 magnesium ion binding                          2/10     \n",
       "GO:0005086 GO:0005086 ARF guanyl-nucleotide exchange factor activity 1/10     \n",
       "GO:0017081 GO:0017081 chloride channel regulator activity            1/10     \n",
       "GO:0140103 GO:0140103 catalytic activity, acting on a glycoprotein   1/10     \n",
       "GO:0032452 GO:0032452 histone demethylase activity                   1/10     \n",
       "           BgRatio   pvalue      p.adjust   qvalue     geneID   Count\n",
       "GO:0045505 10/16969  0.005879051 0.09862894 0.05956882 RAN      1    \n",
       "GO:0000287 209/16969 0.006365894 0.09862894 0.05956882 WNK1/RAN 2    \n",
       "GO:0005086 16/16969  0.009391532 0.09862894 0.05956882 IQSEC3   1    \n",
       "GO:0017081 16/16969  0.009391532 0.09862894 0.05956882 WNK1     1    \n",
       "GO:0140103 20/16969  0.011726977 0.09862894 0.05956882 B4GALNT3 1    \n",
       "GO:0032452 22/16969  0.012892840 0.09862894 0.05956882 KDM5A    1    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(module_chr12_list_enrichGO[[1]]@result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3 id='6.3.2'> Enrich KEGG</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "--> No gene can be mapped....\n",
      "--> Expected input gene ID: 54658,10720,55276,5106,124,219\n",
      "--> return NULL...\n",
      "'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:many mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\"'select()' returned 1:1 mapping between keys and columns\n",
      "Warning message in `[<-`(`*tmp*`, i, value = enrichKEGG(geneIDselect$ENTREZID, organism = \"hsa\", :\n",
      "\"implicit list embedding of S4 objects is deprecated\""
     ]
    }
   ],
   "source": [
    "#-------------------------------\n",
    "\n",
    "module_chr12_list_enrichKEGG <- list()\n",
    "library(org.Hs.eg.db)\n",
    "library(clusterProfiler)\n",
    "for (i in 1:nrow(module_chr12_list)){\n",
    "module_gene <- unlist(module_chr12_list$genelist[i])\n",
    "geneIDselect <-select(org.Hs.eg.db, #.db是这个芯片数据对应的注释包\n",
    "                      keys= module_gene,\n",
    "                      columns=c(\"ENTREZID\",\"ENSEMBL\",\"GENENAME\"), #clolumns参数是你要转换的ID类型是什么，这里选择三个。\n",
    "                      keytype=\"SYMBOL\" )#函数里面的keytype与keys参数是对应的，keys是你输入的那些数据，keytype是指这些数据是属于什么类型的数据。 \n",
    "module_chr12_list_enrichKEGG[i] = enrichKEGG(geneIDselect$ENTREZID, organism = 'hsa', keyType = \"kegg\", pvalueCutoff = 0.05)\n",
    "}\n",
    "names(module_chr12_list_enrichKEGG) <- module_chr12_list$community\n",
    "#--------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>Description</th><th scope=col>GeneRatio</th><th scope=col>BgRatio</th><th scope=col>pvalue</th><th scope=col>p.adjust</th><th scope=col>qvalue</th><th scope=col>geneID</th><th scope=col>Count</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>hsa04721</th><td>hsa04721</td><td>Synaptic vesicle cycle                   </td><td>3/6</td><td>78/7926 </td><td>1.795361e-05</td><td>0.0001436288</td><td>7.559413e-05</td><td>6539/6540/2054</td><td>3</td></tr>\n",
       "\t<tr><th scope=row>hsa04727</th><td>hsa04727</td><td>GABAergic synapse                        </td><td>2/6</td><td>89/7926 </td><td>1.816203e-03</td><td>0.0072648125</td><td>3.823586e-03</td><td>6539/6540     </td><td>2</td></tr>\n",
       "\t<tr><th scope=row>hsa04130</th><td>hsa04130</td><td>SNARE interactions in vesicular transport</td><td>1/6</td><td>34/7926 </td><td>2.547158e-02</td><td>0.0583427598</td><td>3.070672e-02</td><td>2054          </td><td>1</td></tr>\n",
       "\t<tr><th scope=row>hsa00513</th><td>hsa00513</td><td>Various types of N-glycan biosynthesis   </td><td>1/6</td><td>39/7926 </td><td>2.917138e-02</td><td>0.0583427598</td><td>3.070672e-02</td><td>283358        </td><td>1</td></tr>\n",
       "\t<tr><th scope=row>hsa03008</th><td>hsa03008</td><td>Ribosome biogenesis in eukaryotes        </td><td>1/6</td><td>105/7926</td><td>7.692228e-02</td><td>0.1230756510</td><td>6.477666e-02</td><td>5901          </td><td>1</td></tr>\n",
       "\t<tr><th scope=row>hsa03013</th><td>hsa03013</td><td>RNA transport                            </td><td>1/6</td><td>165/7926</td><td>1.186179e-01</td><td>0.1581572420</td><td>8.324065e-02</td><td>5901          </td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & ID & Description & GeneRatio & BgRatio & pvalue & p.adjust & qvalue & geneID & Count\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\thsa04721 & hsa04721 & Synaptic vesicle cycle                    & 3/6 & 78/7926  & 1.795361e-05 & 0.0001436288 & 7.559413e-05 & 6539/6540/2054 & 3\\\\\n",
       "\thsa04727 & hsa04727 & GABAergic synapse                         & 2/6 & 89/7926  & 1.816203e-03 & 0.0072648125 & 3.823586e-03 & 6539/6540      & 2\\\\\n",
       "\thsa04130 & hsa04130 & SNARE interactions in vesicular transport & 1/6 & 34/7926  & 2.547158e-02 & 0.0583427598 & 3.070672e-02 & 2054           & 1\\\\\n",
       "\thsa00513 & hsa00513 & Various types of N-glycan biosynthesis    & 1/6 & 39/7926  & 2.917138e-02 & 0.0583427598 & 3.070672e-02 & 283358         & 1\\\\\n",
       "\thsa03008 & hsa03008 & Ribosome biogenesis in eukaryotes         & 1/6 & 105/7926 & 7.692228e-02 & 0.1230756510 & 6.477666e-02 & 5901           & 1\\\\\n",
       "\thsa03013 & hsa03013 & RNA transport                             & 1/6 & 165/7926 & 1.186179e-01 & 0.1581572420 & 8.324065e-02 & 5901           & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | Description &lt;chr&gt; | GeneRatio &lt;chr&gt; | BgRatio &lt;chr&gt; | pvalue &lt;dbl&gt; | p.adjust &lt;dbl&gt; | qvalue &lt;dbl&gt; | geneID &lt;chr&gt; | Count &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| hsa04721 | hsa04721 | Synaptic vesicle cycle                    | 3/6 | 78/7926  | 1.795361e-05 | 0.0001436288 | 7.559413e-05 | 6539/6540/2054 | 3 |\n",
       "| hsa04727 | hsa04727 | GABAergic synapse                         | 2/6 | 89/7926  | 1.816203e-03 | 0.0072648125 | 3.823586e-03 | 6539/6540      | 2 |\n",
       "| hsa04130 | hsa04130 | SNARE interactions in vesicular transport | 1/6 | 34/7926  | 2.547158e-02 | 0.0583427598 | 3.070672e-02 | 2054           | 1 |\n",
       "| hsa00513 | hsa00513 | Various types of N-glycan biosynthesis    | 1/6 | 39/7926  | 2.917138e-02 | 0.0583427598 | 3.070672e-02 | 283358         | 1 |\n",
       "| hsa03008 | hsa03008 | Ribosome biogenesis in eukaryotes         | 1/6 | 105/7926 | 7.692228e-02 | 0.1230756510 | 6.477666e-02 | 5901           | 1 |\n",
       "| hsa03013 | hsa03013 | RNA transport                             | 1/6 | 165/7926 | 1.186179e-01 | 0.1581572420 | 8.324065e-02 | 5901           | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "         ID       Description                               GeneRatio BgRatio \n",
       "hsa04721 hsa04721 Synaptic vesicle cycle                    3/6       78/7926 \n",
       "hsa04727 hsa04727 GABAergic synapse                         2/6       89/7926 \n",
       "hsa04130 hsa04130 SNARE interactions in vesicular transport 1/6       34/7926 \n",
       "hsa00513 hsa00513 Various types of N-glycan biosynthesis    1/6       39/7926 \n",
       "hsa03008 hsa03008 Ribosome biogenesis in eukaryotes         1/6       105/7926\n",
       "hsa03013 hsa03013 RNA transport                             1/6       165/7926\n",
       "         pvalue       p.adjust     qvalue       geneID         Count\n",
       "hsa04721 1.795361e-05 0.0001436288 7.559413e-05 6539/6540/2054 3    \n",
       "hsa04727 1.816203e-03 0.0072648125 3.823586e-03 6539/6540      2    \n",
       "hsa04130 2.547158e-02 0.0583427598 3.070672e-02 2054           1    \n",
       "hsa00513 2.917138e-02 0.0583427598 3.070672e-02 283358         1    \n",
       "hsa03008 7.692228e-02 0.1230756510 6.477666e-02 5901           1    \n",
       "hsa03013 1.186179e-01 0.1581572420 8.324065e-02 5901           1    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(module_chr12_list_enrichKEGG[[1]]@result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 id='7'>7. Get All genes names</h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a function to get all genes names in neo4j.\n",
    "\n",
    "As we know, GREG only `Details` property have gene names informations in each chromosome nodes.\n",
    "\n",
    "So, the first step, we get all nodes which had `Details` property.  \n",
    "Second, we use `split` to split the `Details` information, at this step, we get a list of `Details` information.  \n",
    "Thrid, we use command `UNWIND` to change the list format.  \n",
    "Forth, we use `contains` to filter those `Details` information which includes `gene_name`. \n",
    "At the end, we use `substring` to get real gene names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    }
   ],
   "source": [
    "#----------------------------------------------------------------------\n",
    "#---建一个函数来获取数据库里面所有gene names\n",
    "#该基因gene name 存在于 node的Details属性上\n",
    "#返回数据框，包含基因名称，节点的id号和节点label， 节点Name\n",
    "\n",
    "getAllGeneName = function(){\n",
    "  #连接neo4j数据库\n",
    "  \n",
    "  library(RNeo4j)\n",
    "  graph = startGraph(\"http://localhost:7474/db/data/\", username=\"neo4j\", password=\"xiaowei\")\n",
    "  \n",
    "  #cypher语言查询\n",
    "  query = \"\n",
    "        //get all nodes which had `Details` property\n",
    "        //查找有Details属性的节点\n",
    "      MATCH (n)\n",
    "      where exists(n.Details)\n",
    "\n",
    "        //split the Details information\n",
    "        //利用split函数拆分Details属性值\n",
    "      with split(n.Details, ';') AS Details,id(n) AS NodeID, labels(n) AS NodeLabel, n.Name AS NodeNames\n",
    "\n",
    "        //change the list format\n",
    "        //利用UNWIND将Details列表换成表格形式\n",
    "      UNWIND Details AS details\n",
    "      with details, NodeID, NodeLabel, NodeNames\n",
    "\n",
    "        //filter those Details information which includes gene_name\n",
    "        // 筛选出含有gene_name的信息\n",
    "      where details contains 'gene_name'\n",
    "      with details as genes, NodeID, NodeLabel, NodeNames\n",
    "\n",
    "        // use `substring` to get real gene names\n",
    "        // 然后利用substring函数截取基因名称， 返回基因名称，节点的id号和节点label\n",
    "      return substring(genes,11) as geneName, NodeID, NodeLabel, NodeNames\n",
    "  \"\n",
    "  #进行查询\n",
    "  allGenes <- RNeo4j::cypher(graph, query) \n",
    "  \n",
    "  #返回查询结果\n",
    "  return(allGenes)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>geneName</th><th scope=col>NodeID</th><th scope=col>NodeLabel</th><th scope=col>NodeNames</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>DDX11L1     </td><td>0</td><td>chr1</td><td>Bin1</td></tr>\n",
       "\t<tr><td>WASH7P      </td><td>0</td><td>chr1</td><td>Bin1</td></tr>\n",
       "\t<tr><td>MIR6859-1   </td><td>0</td><td>chr1</td><td>Bin1</td></tr>\n",
       "\t<tr><td>RP11-34P13.3</td><td>0</td><td>chr1</td><td>Bin1</td></tr>\n",
       "\t<tr><td>MIR1302-2   </td><td>0</td><td>chr1</td><td>Bin1</td></tr>\n",
       "\t<tr><td>FAM138A     </td><td>0</td><td>chr1</td><td>Bin1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       " geneName & NodeID & NodeLabel & NodeNames\\\\\n",
       " <chr> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t DDX11L1      & 0 & chr1 & Bin1\\\\\n",
       "\t WASH7P       & 0 & chr1 & Bin1\\\\\n",
       "\t MIR6859-1    & 0 & chr1 & Bin1\\\\\n",
       "\t RP11-34P13.3 & 0 & chr1 & Bin1\\\\\n",
       "\t MIR1302-2    & 0 & chr1 & Bin1\\\\\n",
       "\t FAM138A      & 0 & chr1 & Bin1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| geneName &lt;chr&gt; | NodeID &lt;int&gt; | NodeLabel &lt;chr&gt; | NodeNames &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| DDX11L1      | 0 | chr1 | Bin1 |\n",
       "| WASH7P       | 0 | chr1 | Bin1 |\n",
       "| MIR6859-1    | 0 | chr1 | Bin1 |\n",
       "| RP11-34P13.3 | 0 | chr1 | Bin1 |\n",
       "| MIR1302-2    | 0 | chr1 | Bin1 |\n",
       "| FAM138A      | 0 | chr1 | Bin1 |\n",
       "\n"
      ],
      "text/plain": [
       "  geneName     NodeID NodeLabel NodeNames\n",
       "1 DDX11L1      0      chr1      Bin1     \n",
       "2 WASH7P       0      chr1      Bin1     \n",
       "3 MIR6859-1    0      chr1      Bin1     \n",
       "4 RP11-34P13.3 0      chr1      Bin1     \n",
       "5 MIR1302-2    0      chr1      Bin1     \n",
       "6 FAM138A      0      chr1      Bin1     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "AllgeneName <- getAllGeneName()\n",
    "head(AllgeneName )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>67418</li>\n",
       "\t<li>4</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 67418\n",
       "\\item 4\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 67418\n",
       "2. 4\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 67418     4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(AllgeneName)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 id='8'>  Find the modules that contain the gene \"NANOG\" and report all the nodes and relationships inside such modules. Also, report if they are enriched in some KEGG pathway or GO term. </h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='8.1'> Step1： Find out those nodes about target genes</h2>\n",
    "Find out those nodes contains the gene `\"NOANOG\"` strings, \n",
    "\n",
    "result is a data.frame including `GeneName`, `NodeID`, `NodeLabel` and `NodeNames`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    }
   ],
   "source": [
    "#接下来我们来尝试一下这一步\n",
    "#Find the modules that contain the gene \"NANOG\" \n",
    "#and report all the nodes and relationships inside such modules\n",
    "\n",
    "#首先，我们来得到哪些基因里面包含`NANOG` 的节点信息\n",
    "#This is for those gene names contain 'NANOG' strings.\n",
    "#TargetNodes <- AllgeneName[grep('NANOG',AllgeneName$geneName, fixed = FALSE), ]\n",
    "\n",
    "#Find out node id of `NANOG` as gene name \n",
    "TargetNodes <- AllgeneName[which(AllgeneName$geneName == 'NANOG'), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 1 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>geneName</th><th scope=col>NodeID</th><th scope=col>NodeLabel</th><th scope=col>NodeNames</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>39711</th><td>NANOG</td><td>9759</td><td>chr12</td><td>Bin39</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & geneName & NodeID & NodeLabel & NodeNames\\\\\n",
       "  & <chr> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t39711 & NANOG & 9759 & chr12 & Bin39\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 4\n",
       "\n",
       "| <!--/--> | geneName &lt;chr&gt; | NodeID &lt;int&gt; | NodeLabel &lt;chr&gt; | NodeNames &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 39711 | NANOG | 9759 | chr12 | Bin39 |\n",
       "\n"
      ],
      "text/plain": [
       "      geneName NodeID NodeLabel NodeNames\n",
       "39711 NANOG    9759   chr12     Bin39    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "TargetNodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='8.2'> Step2: To get modules which includes target genes</h2>\n",
    "\n",
    "we use `{}` to input arguments to cypher commands. But at here we need to use cypher command `MATCH (n)   where id(n) in {TargetNodes_NodeID}  with id(n) AS TargetNodes_ID` to get `TargetNodeID` again, then we can follow to match those modules including `TagetNodeID`.\n",
    "\n",
    "For now, we can not find out what's problems at this step if we directly use  `{}` as match arguments in command `where {TargetNodes_ID} in nodeId`.\n",
    "\n",
    "Because sometimes we only one gene as targetNode,we should add `if` as conditions to make query comands efective."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    }
   ],
   "source": [
    "#第二步，我们利用{}来传递TargetNode的ID号，并找出含有这些TargetNode的模块\n",
    "\n",
    "\n",
    "queryLouvainForChr12_TargetNodesInModules =function(TargetNodes){\n",
    "    query = paste(\"\n",
    "        // to make louvain modules for chr12\n",
    "        CALL algo.louvain.stream(\n",
    "          'MATCH (n:chr12)--(m) with id(n) as chr12,id(m) AS m   MATCH (al) where id(al) = m or  id(al) = chr12 return id(al) as id',\n",
    "          'MATCH (n:chr12)-[r]-(m) RETURN id(n) AS source, id(m) AS target',\n",
    "          {graph:'cypher',direction: 'both'})\n",
    "        YIELD nodeId, community\n",
    "\n",
    "        // filtering community which the number of nodes is more than one\n",
    "        with collect(nodeId) AS nodeId , community\n",
    "        where size(nodeId) > 1\n",
    "        with nodeId, 'Module'+ community AS community\n",
    "\n",
    "        // filtering community that had target nodes\n",
    "        MATCH (n)\",\n",
    "\n",
    "\n",
    "        if (length(TargetNodes) != 1){\"where id(n) in {TargetNodes_NodeID}\" }\n",
    "        else{\"where id(n) = {TargetNodes_NodeID}\"},\n",
    "\n",
    "        \"with id(n) AS TargetNodes_ID, n.Name AS TargetNodes_Name, nodeId, community\n",
    "        where TargetNodes_ID in nodeId\n",
    "        return nodeId, community,collect(TargetNodes_ID) AS TargetNodes_ID ,collect(TargetNodes_Name) AS TargetNodes_Name\n",
    "        \"\n",
    "        ,sep = ' ')\n",
    "    return(query)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = queryLouvainForChr12_TargetNodesInModules(TargetNodes$NodeID)\n",
    "chr12Mod <- cypher(graph, query,TargetNodes_NodeID = TargetNodes$NodeID)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The result is a data.frame including all nodes ids in each community, community, TargetNodes_ID and TargetNodes_Name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#chr12Mod[1,]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='8.3'> Step3: build a function to get relationships during nodes in each community(module)</h2>\n",
    "We build a function to get relationships during nodes in each community(module). The input argument is the result we got at the step2. Results includes two nodes' id, labels, properties and relationship id, type, and properties."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    }
   ],
   "source": [
    "#----------------建一个函数来获取每一个模块节点之间的关系--------------------------------------\n",
    "#\n",
    "#The argument is result we get  from step2\n",
    "#module就是我们前面获取的结果\n",
    "Module_Relationship = function(Module){\n",
    "  \n",
    "    #Module_Relationship as an object for save our results\n",
    "    #Module_Relationship作为list来存储数据\n",
    "  Module_Relationship = list()\n",
    "    \n",
    "    #To get relationships during nodes in the same modules\n",
    "    #获取节点之间的关系，返回的结果尽量不要去改变。\n",
    "    #The results include nodes' id ,label and properties, and relationships' type, id and properties.\n",
    "    #返回的结果包括节点的id号，label，还有就是节点的属性；以及关系的id号和关系的类型、关系的属性\n",
    "  for( i in 1:nrow(Module) ){\n",
    "      #for the vector about all nodes id in the same modules\n",
    "    Module_NodeID = unlist(Module[[1]][[i]])\n",
    "    query = \"\n",
    "            // the first we should get all node in the same modules, result is a list\n",
    "            MATCH (n)\n",
    "            where id(n) in {Module_NodeID}\n",
    "            with collect( id(n) ) AS Module_Node_ID\n",
    "        \n",
    "        // get relationships \n",
    "        MATCH (A)-[r]->(B)\n",
    "        where id(A) in Module_Node_ID and id(B) in Module_Node_ID\n",
    "        \n",
    "        //at here, properties() function for all properties at an node or a relationship    \n",
    "        RETURN id(A) AS fromNodeID, \n",
    "              labels(A) AS fromNodeLabel, \n",
    "              \n",
    "              id(B) AS toNodeID,\n",
    "              labels(B) AS toNodeLabel,\n",
    "              \n",
    "              id(r) AS Relationship_ID,\n",
    "              type(r) AS Relationship,\n",
    "          \n",
    "              properties(A) AS AProperties,\n",
    "              properties(B) AS BProperties,\n",
    "              properties(r) AS RProperties\n",
    "        \"\n",
    "     #run the query for results   \n",
    "    #Module_Relationship[[i]]= cypher(graph, query, Module_NodeID = Module_NodeID )\n",
    "    ABRelationship= cypher(graph, query, Module_NodeID = Module_NodeID )\n",
    "    \n",
    "    #Now, we split all properties of nodes using unnest_nodes() function in {neo4r}\n",
    "      #这里解析属性使用的neo4r的unnest_nodes()来拆分\n",
    "    #解析节点A的属性\n",
    "    Ap = ABRelationship[\"AProperties\"]\n",
    "    names(Ap)[1] = \"properties\"\n",
    "    App =  neo4r::unnest_nodes(Ap, what = \"properties\")\n",
    "    names(App) = paste0(\"fromNode_\", names(App) )\n",
    "    \n",
    "    #解析节点B的属性\n",
    "    Bp = ABRelationship[\"BProperties\"]\n",
    "    names(Bp)[1] = \"properties\"\n",
    "    Bpp =  neo4r::unnest_nodes(Bp, what = \"properties\")\n",
    "    names(Bpp) = paste0(\"toNode_\", names(Bpp) )\n",
    "    \n",
    "    \n",
    "    #split properties of relationships,we need make the same formats. If not, there show some wrong and we can not fixed.\n",
    "    #解析关系的属性\n",
    "    #这里需要说明的是数据框的列名称和列的位置不能发生变化，否则容易出错，在解析完关系后我们将它的列名还原过来\n",
    "    Rp = ABRelationship[,c(\"Relationship_ID\",\"Relationship\",\"fromNodeID\",\"toNodeID\",\"RProperties\")]\n",
    "    names(Rp) = c(\"id\", \"type\", \"stratNode\", \"endNode\", \"properties\")\n",
    "    Rpp =  neo4r::unnest_relationships(Rp)\n",
    "    names(Rpp)[1:4] = c(\"Relationship_ID\",\"Relationship\",\"fromNodeID\",\"toNodeID\")\n",
    "    names(Rpp)[c(-(1:4))] = paste0(\"R_\", names(Rpp)[c(-(1:4))] )\n",
    "    \n",
    "    #union all data\n",
    "    #最后，合并每一列\n",
    "    alldataRelationship = cbind(Rpp,App,Bpp, ABRelationship[,c(\"fromNodeLabel\",\"toNodeLabel\")])\n",
    "    ColNames = c(\"fromNodeID\",\"fromNodeLabel\",\"toNodeID\",\"toNodeLabel\",\"Relationship_ID\",\"Relationship\",\n",
    "                 names(App),\n",
    "                 names(Bpp),\n",
    "                 names(Rpp)[c(-(1:4))])\n",
    "    Module_Relationship[[i]] = alldataRelationship[,ColNames]\n",
    "  }\n",
    "  \n",
    "  names(Module_Relationship) = Module$community\n",
    "  \n",
    "  return(Module_Relationship)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='8.4'> step4: get relationships during nodes in each community(module)</h2>\n",
    "Now, we get relationships during nodes in each community(module), result is a list including data.frame data that have nodeID and their relationship."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windowsWarning message:\n",
      "\"The `.drop` argument of `unnest()` is deprecated as of tidyr 1.0.0.\n",
      "All list-columns are now preserved.\n",
      "\u001b[90mThis warning is displayed once per session.\u001b[39m\n",
      "\u001b[90mCall `lifecycle::last_warnings()` to see where this warning was generated.\u001b[39m\""
     ]
    }
   ],
   "source": [
    "#---------获取Module里的node的关系-------------------------------------------------------------\n",
    "Module_Relationships <- Module_Relationship(chr12Mod)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's us see what we get.**It includes all properties about nodes and relationships.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 26</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>fromNodeID</th><th scope=col>fromNodeLabel</th><th scope=col>toNodeID</th><th scope=col>toNodeLabel</th><th scope=col>Relationship_ID</th><th scope=col>Relationship</th><th scope=col>fromNode_Start</th><th scope=col>fromNode_End</th><th scope=col>fromNode_community</th><th scope=col>fromNode_Details</th><th scope=col>...</th><th scope=col>R_PubMedID</th><th scope=col>R_CellType</th><th scope=col>R_SourceDB</th><th scope=col>R_Method</th><th scope=col>R_value</th><th scope=col>R_GEO</th><th scope=col>R_Start</th><th scope=col>R_OtherGEO</th><th scope=col>R_End</th><th scope=col>R_Information</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>...</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>834688</td><td>chr12_Range</td><td>834161</td><td>chr12_Range</td><td>4762133</td><td>Interaction</td><td>5387590</td><td>5399902</td><td>15900</td><td>NA</td><td>...</td><td>24141950</td><td>IMR90</td><td>CistromeDB</td><td>Hi-C</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>834678</td><td>chr12_Range</td><td>860049</td><td>chr12_Range</td><td>4762119</td><td>Interaction</td><td>5382566</td><td>5399902</td><td>15900</td><td>NA</td><td>...</td><td>24141950</td><td>IMR90</td><td>CistromeDB</td><td>Hi-C</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>834677</td><td>chr12_Range</td><td>834618</td><td>chr12_Range</td><td>4762117</td><td>Interaction</td><td>5382566</td><td>5387589</td><td>15900</td><td>NA</td><td>...</td><td>24141950</td><td>IMR90</td><td>CistromeDB</td><td>Hi-C</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>834677</td><td>chr12_Range</td><td>834389</td><td>chr12_Range</td><td>4762116</td><td>Interaction</td><td>5382566</td><td>5387589</td><td>15900</td><td>NA</td><td>...</td><td>24141950</td><td>IMR90</td><td>CistromeDB</td><td>Hi-C</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>834662</td><td>chr12_Range</td><td>834115</td><td>chr12_Range</td><td>4762098</td><td>Interaction</td><td>5376790</td><td>5387589</td><td>15900</td><td>NA</td><td>...</td><td>24141950</td><td>IMR90</td><td>CistromeDB</td><td>Hi-C</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>834662</td><td>chr12_Range</td><td>833788</td><td>chr12_Range</td><td>4762097</td><td>Interaction</td><td>5376790</td><td>5387589</td><td>15900</td><td>NA</td><td>...</td><td>24141950</td><td>IMR90</td><td>CistromeDB</td><td>Hi-C</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 26\n",
       "\\begin{tabular}{r|llllllllllllllllllllllllll}\n",
       " fromNodeID & fromNodeLabel & toNodeID & toNodeLabel & Relationship\\_ID & Relationship & fromNode\\_Start & fromNode\\_End & fromNode\\_community & fromNode\\_Details & fromNode\\_Name & toNode\\_Start & toNode\\_End & toNode\\_community & toNode\\_Details & toNode\\_Name & R\\_PubMedID & R\\_CellType & R\\_SourceDB & R\\_Method & R\\_value & R\\_GEO & R\\_Start & R\\_OtherGEO & R\\_End & R\\_Information\\\\\n",
       " <int> & <chr> & <int> & <chr> & <int> & <chr> & <chr> & <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <lgl> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 834688 & chr12\\_Range & 834161 & chr12\\_Range & 4762133 & Interaction & 5387590 & 5399902 & 15900 & NA & NA & 5298457 & 5302447 & 15900 & NA & NA & 24141950 & IMR90 & CistromeDB & Hi-C & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 834678 & chr12\\_Range & 860049 & chr12\\_Range & 4762119 & Interaction & 5382566 & 5399902 & 15900 & NA & NA & 5248002 & 5260657 & 15900 & NA & NA & 24141950 & IMR90 & CistromeDB & Hi-C & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 834677 & chr12\\_Range & 834618 & chr12\\_Range & 4762117 & Interaction & 5382566 & 5387589 & 15900 & NA & NA & 5370303 & 5376789 & 15900 & NA & NA & 24141950 & IMR90 & CistromeDB & Hi-C & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 834677 & chr12\\_Range & 834389 & chr12\\_Range & 4762116 & Interaction & 5382566 & 5387589 & 15900 & NA & NA & 5337153 & 5344314 & 15900 & NA & NA & 24141950 & IMR90 & CistromeDB & Hi-C & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 834662 & chr12\\_Range & 834115 & chr12\\_Range & 4762098 & Interaction & 5376790 & 5387589 & 15900 & NA & NA & 5292209 & 5298456 & 15900 & NA & NA & 24141950 & IMR90 & CistromeDB & Hi-C & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 834662 & chr12\\_Range & 833788 & chr12\\_Range & 4762097 & Interaction & 5376790 & 5387589 & 15900 & NA & NA & 5238181 & 5244322 & 15900 & NA & NA & 24141950 & IMR90 & CistromeDB & Hi-C & NA & NA & NA & NA & NA & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 26\n",
       "\n",
       "| fromNodeID &lt;int&gt; | fromNodeLabel &lt;chr&gt; | toNodeID &lt;int&gt; | toNodeLabel &lt;chr&gt; | Relationship_ID &lt;int&gt; | Relationship &lt;chr&gt; | fromNode_Start &lt;chr&gt; | fromNode_End &lt;chr&gt; | fromNode_community &lt;int&gt; | fromNode_Details &lt;chr&gt; | ... ... | R_PubMedID &lt;chr&gt; | R_CellType &lt;chr&gt; | R_SourceDB &lt;chr&gt; | R_Method &lt;chr&gt; | R_value &lt;lgl&gt; | R_GEO &lt;chr&gt; | R_Start &lt;chr&gt; | R_OtherGEO &lt;chr&gt; | R_End &lt;chr&gt; | R_Information &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 834688 | chr12_Range | 834161 | chr12_Range | 4762133 | Interaction | 5387590 | 5399902 | 15900 | NA | ... | 24141950 | IMR90 | CistromeDB | Hi-C | NA | NA | NA | NA | NA | NA |\n",
       "| 834678 | chr12_Range | 860049 | chr12_Range | 4762119 | Interaction | 5382566 | 5399902 | 15900 | NA | ... | 24141950 | IMR90 | CistromeDB | Hi-C | NA | NA | NA | NA | NA | NA |\n",
       "| 834677 | chr12_Range | 834618 | chr12_Range | 4762117 | Interaction | 5382566 | 5387589 | 15900 | NA | ... | 24141950 | IMR90 | CistromeDB | Hi-C | NA | NA | NA | NA | NA | NA |\n",
       "| 834677 | chr12_Range | 834389 | chr12_Range | 4762116 | Interaction | 5382566 | 5387589 | 15900 | NA | ... | 24141950 | IMR90 | CistromeDB | Hi-C | NA | NA | NA | NA | NA | NA |\n",
       "| 834662 | chr12_Range | 834115 | chr12_Range | 4762098 | Interaction | 5376790 | 5387589 | 15900 | NA | ... | 24141950 | IMR90 | CistromeDB | Hi-C | NA | NA | NA | NA | NA | NA |\n",
       "| 834662 | chr12_Range | 833788 | chr12_Range | 4762097 | Interaction | 5376790 | 5387589 | 15900 | NA | ... | 24141950 | IMR90 | CistromeDB | Hi-C | NA | NA | NA | NA | NA | NA |\n",
       "\n"
      ],
      "text/plain": [
       "  fromNodeID fromNodeLabel toNodeID toNodeLabel Relationship_ID Relationship\n",
       "1 834688     chr12_Range   834161   chr12_Range 4762133         Interaction \n",
       "2 834678     chr12_Range   860049   chr12_Range 4762119         Interaction \n",
       "3 834677     chr12_Range   834618   chr12_Range 4762117         Interaction \n",
       "4 834677     chr12_Range   834389   chr12_Range 4762116         Interaction \n",
       "5 834662     chr12_Range   834115   chr12_Range 4762098         Interaction \n",
       "6 834662     chr12_Range   833788   chr12_Range 4762097         Interaction \n",
       "  fromNode_Start fromNode_End fromNode_community fromNode_Details ...\n",
       "1 5387590        5399902      15900              NA               ...\n",
       "2 5382566        5399902      15900              NA               ...\n",
       "3 5382566        5387589      15900              NA               ...\n",
       "4 5382566        5387589      15900              NA               ...\n",
       "5 5376790        5387589      15900              NA               ...\n",
       "6 5376790        5387589      15900              NA               ...\n",
       "  R_PubMedID R_CellType R_SourceDB R_Method R_value R_GEO R_Start R_OtherGEO\n",
       "1 24141950   IMR90      CistromeDB Hi-C     NA      NA    NA      NA        \n",
       "2 24141950   IMR90      CistromeDB Hi-C     NA      NA    NA      NA        \n",
       "3 24141950   IMR90      CistromeDB Hi-C     NA      NA    NA      NA        \n",
       "4 24141950   IMR90      CistromeDB Hi-C     NA      NA    NA      NA        \n",
       "5 24141950   IMR90      CistromeDB Hi-C     NA      NA    NA      NA        \n",
       "6 24141950   IMR90      CistromeDB Hi-C     NA      NA    NA      NA        \n",
       "  R_End R_Information\n",
       "1 NA    NA           \n",
       "2 NA    NA           \n",
       "3 NA    NA           \n",
       "4 NA    NA           \n",
       "5 NA    NA           \n",
       "6 NA    NA           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(Module_Relationships[[1]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 id='9'>Select the modules that include NANOG AND CTCF AND EP300 AND POLR2A AND YY1 AND cohesin (RAD21, SMC3, OR STAG1) AND mediator (MED1 or MED12).</h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='9.1'> Step1: Get nodes in all gene names in chromosomes,TF and LncRNA.</h2>\n",
    "For this term, its include TF genes, so we expand nodes of TF and LncRNA labels as following codes:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------TF----------------------\n",
    "query = \"MATCH (n:TF)\n",
    "        RETURN n.Name AS geneName, id(n) AS NodeID, labels(n) AS NodeLabel, n.Name AS NodeNames\"\n",
    "TF = cypher(graph, query)\n",
    "#-----------LncRNA----------------\n",
    "query = \"MATCH (n:LncRNA)\n",
    "        RETURN n.Name AS geneName, id(n) AS NodeID, labels(n) AS NodeLabel, n.Name AS NodeNames\"\n",
    "LncRNA = cypher(graph, query)\n",
    "#------------AllgeneName_TF_LncRNA--------------------------------------\n",
    "AllgeneName_TF_LncRNA = rbind(TF, LncRNA, AllgeneName)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So `AllgeneName_TF_LncRNA` includes all genes in all chromosomes, all transcription factors(TFs) and all LncRNA in GREG."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>geneName</th><th scope=col>NodeID</th><th scope=col>NodeLabel</th><th scope=col>NodeNames</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AFF4  </td><td>15448</td><td>TF</td><td>AFF4  </td></tr>\n",
       "\t<tr><td>AHR   </td><td>15449</td><td>TF</td><td>AHR   </td></tr>\n",
       "\t<tr><td>ARID3A</td><td>15450</td><td>TF</td><td>ARID3A</td></tr>\n",
       "\t<tr><td>ARNT  </td><td>15451</td><td>TF</td><td>ARNT  </td></tr>\n",
       "\t<tr><td>ARRB1 </td><td>15452</td><td>TF</td><td>ARRB1 </td></tr>\n",
       "\t<tr><td>ATF1  </td><td>15453</td><td>TF</td><td>ATF1  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       " geneName & NodeID & NodeLabel & NodeNames\\\\\n",
       " <chr> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t AFF4   & 15448 & TF & AFF4  \\\\\n",
       "\t AHR    & 15449 & TF & AHR   \\\\\n",
       "\t ARID3A & 15450 & TF & ARID3A\\\\\n",
       "\t ARNT   & 15451 & TF & ARNT  \\\\\n",
       "\t ARRB1  & 15452 & TF & ARRB1 \\\\\n",
       "\t ATF1   & 15453 & TF & ATF1  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| geneName &lt;chr&gt; | NodeID &lt;int&gt; | NodeLabel &lt;chr&gt; | NodeNames &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| AFF4   | 15448 | TF | AFF4   |\n",
       "| AHR    | 15449 | TF | AHR    |\n",
       "| ARID3A | 15450 | TF | ARID3A |\n",
       "| ARNT   | 15451 | TF | ARNT   |\n",
       "| ARRB1  | 15452 | TF | ARRB1  |\n",
       "| ATF1   | 15453 | TF | ATF1   |\n",
       "\n"
      ],
      "text/plain": [
       "  geneName NodeID NodeLabel NodeNames\n",
       "1 AFF4     15448  TF        AFF4     \n",
       "2 AHR      15449  TF        AHR      \n",
       "3 ARID3A   15450  TF        ARID3A   \n",
       "4 ARNT     15451  TF        ARNT     \n",
       "5 ARRB1    15452  TF        ARRB1    \n",
       "6 ATF1     15453  TF        ATF1     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(AllgeneName_TF_LncRNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>67714</li>\n",
       "\t<li>4</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 67714\n",
       "\\item 4\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 67714\n",
       "2. 4\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 67714     4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(AllgeneName_TF_LncRNA)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='9.2'> Step2: A vector about out genes.</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "#----------Find the gene nodes------------------------------------\n",
    "gene <- c('NANOG', 'CTCF', 'EP300', 'POLR2A', 'YY1', 'RAD21', 'SMC3', 'STAG1', 'MED1', 'MED12')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='9.3'> Step3: Create a function `geneMatchIndex` for find out those nodes that genes we want.</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we create a function for match those genes by full match or fuzzy match. Full match means your input is full name of genes, that you only get this nodes that gene names is your input.  \n",
    "\n",
    "This is as defaults. please set argument `FullMatch = 0` If you want use fuzzy match for those gene names contain strings of your input. \n",
    "\n",
    "The argument `by` must one of `\"geneName\"`,  `\"NodeID\"` ,   `\"NodeLabel\"` and `\"NodeNames\"`. It means the type of your input data.\n",
    "\n",
    "Result is a list includes a data frame of `TargetNodes` about genes and a vector `NoGeneIndex` about those genes not in GREG. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    }
   ],
   "source": [
    "#-----------grep或者which---------------------\n",
    "#-----------------------------------------\n",
    "#a function geneMatchIndex for find out those nodes that genes we want\n",
    "#一个函数用来查找gene在所有基因数据框内的对应位置和没有匹配的基因\n",
    "\n",
    "#This is as defaults. please set argument `FullMatch = 0` If you want use fuzzy match for those gene names contain strings of your input.\n",
    "#FullMatch默认为1，也就是说默认使用的是完全匹配的方式，找到的基因名称跟输入的是一模一样的\n",
    "#FullMatch如果是其他数字，则用grep的方式查找含有该字符串的基因位置\n",
    "\n",
    "#Result is a list includes a data frame of `TargetNodes` about genes and a vector `NoGeneIndex` about those genes not in GREG.\n",
    "#返回的是两个向量的list，geneIndex是对应基因的位置；NoGeneIndex是该基因在数据库中没有找到\n",
    "\n",
    "#The argument `by` must one of `\"geneName\"`,  `\"NodeID\"` ,   `\"NodeLabel\"` and `\"NodeNames\"`. It means the type of your input data.\n",
    "#第三个参数是指匹配的是哪个类型， 是属于基因名，还是节点id，节点的label和节点名称\n",
    "#第三个参数 即 \"geneName\"  \"NodeID\"    \"NodeLabel\" \"NodeNames\"\n",
    "geneMatchIndex = function(gene, FullMatch = 1, by = \"geneName\"){\n",
    "  geneIndex = c() \n",
    "  NoGeneIndex =  c()\n",
    "  Ag = AllgeneName_TF_LncRNA[by]\n",
    "  if (FullMatch == 1){\n",
    "    #----------which-----------------\n",
    "    for (i in 1:length(gene) ) {\n",
    "      geneIndex1 = which(Ag == gene[i])\n",
    "      if (length(geneIndex1) == 0){NoGeneIndex = c(gene[i], NoGeneIndex)}\n",
    "      else{geneIndex = c(geneIndex,geneIndex1)}\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  else{\n",
    "    #----------grep---------------\n",
    "    for (i in 1:length(gene) ) {\n",
    "      geneIndex1 = grep(gene[i],Ag, fixed = TRUE)\n",
    "      if (length(geneIndex1) == 0){NoGeneIndex = c(gene[i], NoGeneIndex)}\n",
    "      else{geneIndex = c(geneIndex,geneIndex1)}\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  TargetNodes <- AllgeneName_TF_LncRNA[geneIndex, ]\n",
    "  geneInGREG = list(TargetNodes, NoGeneIndex)\n",
    "  names(geneInGREG) = c(\"TargetNodes\", \"NoGeneIndex\")\n",
    "  #返回匹配上的TargetNodes和没有找到基因NoGeneIndex\n",
    "  return (geneInGREG )\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='9.4'> Step4: Find out those nodes that genes we want.</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 24 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>geneName</th><th scope=col>NodeID</th><th scope=col>NodeLabel</th><th scope=col>NodeNames</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>257</th><td>NANOG </td><td>1288081</td><td>TF   </td><td>NANOG </td></tr>\n",
       "\t<tr><th scope=row>40007</th><td>NANOG </td><td>   9759</td><td>chr12</td><td>Bin39 </td></tr>\n",
       "\t<tr><th scope=row>40</th><td>CTCF  </td><td>  15487</td><td>TF   </td><td>CTCF  </td></tr>\n",
       "\t<tr><th scope=row>51743</th><td>CTCF  </td><td>  12342</td><td>chr16</td><td>Bin338</td></tr>\n",
       "\t<tr><th scope=row>51744</th><td>CTCF  </td><td>  12343</td><td>chr16</td><td>Bin339</td></tr>\n",
       "\t<tr><th scope=row>58</th><td>EP300 </td><td>  15505</td><td>TF   </td><td>EP300 </td></tr>\n",
       "\t<tr><th scope=row>64105</th><td>EP300 </td><td>  14330</td><td>chr22</td><td>Bin206</td></tr>\n",
       "\t<tr><th scope=row>155</th><td>POLR2A</td><td>  15602</td><td>TF   </td><td>POLR2A</td></tr>\n",
       "\t<tr><th scope=row>52860</th><td>POLR2A</td><td>  12494</td><td>chr17</td><td>Bin38 </td></tr>\n",
       "\t<tr><th scope=row>240</th><td>YY1   </td><td>  15687</td><td>TF   </td><td>YY1   </td></tr>\n",
       "\t<tr><th scope=row>46700</th><td>YY1   </td><td>  11461</td><td>chr14</td><td>Bin502</td></tr>\n",
       "\t<tr><th scope=row>167</th><td>RAD21 </td><td>  15614</td><td>TF   </td><td>RAD21 </td></tr>\n",
       "\t<tr><th scope=row>30285</th><td>RAD21 </td><td>   7542</td><td>chr8 </td><td>Bin585</td></tr>\n",
       "\t<tr><th scope=row>194</th><td>SMC3  </td><td>  15641</td><td>TF   </td><td>SMC3  </td></tr>\n",
       "\t<tr><th scope=row>35620</th><td>SMC3  </td><td>   8928</td><td>chr10</td><td>Bin553</td></tr>\n",
       "\t<tr><th scope=row>35621</th><td>SMC3  </td><td>   8929</td><td>chr10</td><td>Bin554</td></tr>\n",
       "\t<tr><th scope=row>205</th><td>STAG1 </td><td>  15652</td><td>TF   </td><td>STAG1 </td></tr>\n",
       "\t<tr><th scope=row>13536</th><td>STAG1 </td><td>   3137</td><td>chr3 </td><td>Bin682</td></tr>\n",
       "\t<tr><th scope=row>13537</th><td>STAG1 </td><td>   3138</td><td>chr3 </td><td>Bin683</td></tr>\n",
       "\t<tr><th scope=row>13539</th><td>STAG1 </td><td>   3139</td><td>chr3 </td><td>Bin684</td></tr>\n",
       "\t<tr><th scope=row>125</th><td>MED1  </td><td>  15572</td><td>TF   </td><td>MED1  </td></tr>\n",
       "\t<tr><th scope=row>54100</th><td>MED1  </td><td>  12654</td><td>chr17</td><td>Bin198</td></tr>\n",
       "\t<tr><th scope=row>126</th><td>MED12 </td><td>  15573</td><td>TF   </td><td>MED12 </td></tr>\n",
       "\t<tr><th scope=row>65605</th><td>MED12 </td><td>  14735</td><td>chrX </td><td>Bin356</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 24 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & geneName & NodeID & NodeLabel & NodeNames\\\\\n",
       "  & <chr> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t257 & NANOG  & 1288081 & TF    & NANOG \\\\\n",
       "\t40007 & NANOG  &    9759 & chr12 & Bin39 \\\\\n",
       "\t40 & CTCF   &   15487 & TF    & CTCF  \\\\\n",
       "\t51743 & CTCF   &   12342 & chr16 & Bin338\\\\\n",
       "\t51744 & CTCF   &   12343 & chr16 & Bin339\\\\\n",
       "\t58 & EP300  &   15505 & TF    & EP300 \\\\\n",
       "\t64105 & EP300  &   14330 & chr22 & Bin206\\\\\n",
       "\t155 & POLR2A &   15602 & TF    & POLR2A\\\\\n",
       "\t52860 & POLR2A &   12494 & chr17 & Bin38 \\\\\n",
       "\t240 & YY1    &   15687 & TF    & YY1   \\\\\n",
       "\t46700 & YY1    &   11461 & chr14 & Bin502\\\\\n",
       "\t167 & RAD21  &   15614 & TF    & RAD21 \\\\\n",
       "\t30285 & RAD21  &    7542 & chr8  & Bin585\\\\\n",
       "\t194 & SMC3   &   15641 & TF    & SMC3  \\\\\n",
       "\t35620 & SMC3   &    8928 & chr10 & Bin553\\\\\n",
       "\t35621 & SMC3   &    8929 & chr10 & Bin554\\\\\n",
       "\t205 & STAG1  &   15652 & TF    & STAG1 \\\\\n",
       "\t13536 & STAG1  &    3137 & chr3  & Bin682\\\\\n",
       "\t13537 & STAG1  &    3138 & chr3  & Bin683\\\\\n",
       "\t13539 & STAG1  &    3139 & chr3  & Bin684\\\\\n",
       "\t125 & MED1   &   15572 & TF    & MED1  \\\\\n",
       "\t54100 & MED1   &   12654 & chr17 & Bin198\\\\\n",
       "\t126 & MED12  &   15573 & TF    & MED12 \\\\\n",
       "\t65605 & MED12  &   14735 & chrX  & Bin356\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 24 × 4\n",
       "\n",
       "| <!--/--> | geneName &lt;chr&gt; | NodeID &lt;int&gt; | NodeLabel &lt;chr&gt; | NodeNames &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 257 | NANOG  | 1288081 | TF    | NANOG  |\n",
       "| 40007 | NANOG  |    9759 | chr12 | Bin39  |\n",
       "| 40 | CTCF   |   15487 | TF    | CTCF   |\n",
       "| 51743 | CTCF   |   12342 | chr16 | Bin338 |\n",
       "| 51744 | CTCF   |   12343 | chr16 | Bin339 |\n",
       "| 58 | EP300  |   15505 | TF    | EP300  |\n",
       "| 64105 | EP300  |   14330 | chr22 | Bin206 |\n",
       "| 155 | POLR2A |   15602 | TF    | POLR2A |\n",
       "| 52860 | POLR2A |   12494 | chr17 | Bin38  |\n",
       "| 240 | YY1    |   15687 | TF    | YY1    |\n",
       "| 46700 | YY1    |   11461 | chr14 | Bin502 |\n",
       "| 167 | RAD21  |   15614 | TF    | RAD21  |\n",
       "| 30285 | RAD21  |    7542 | chr8  | Bin585 |\n",
       "| 194 | SMC3   |   15641 | TF    | SMC3   |\n",
       "| 35620 | SMC3   |    8928 | chr10 | Bin553 |\n",
       "| 35621 | SMC3   |    8929 | chr10 | Bin554 |\n",
       "| 205 | STAG1  |   15652 | TF    | STAG1  |\n",
       "| 13536 | STAG1  |    3137 | chr3  | Bin682 |\n",
       "| 13537 | STAG1  |    3138 | chr3  | Bin683 |\n",
       "| 13539 | STAG1  |    3139 | chr3  | Bin684 |\n",
       "| 125 | MED1   |   15572 | TF    | MED1   |\n",
       "| 54100 | MED1   |   12654 | chr17 | Bin198 |\n",
       "| 126 | MED12  |   15573 | TF    | MED12  |\n",
       "| 65605 | MED12  |   14735 | chrX  | Bin356 |\n",
       "\n"
      ],
      "text/plain": [
       "      geneName NodeID  NodeLabel NodeNames\n",
       "257   NANOG    1288081 TF        NANOG    \n",
       "40007 NANOG       9759 chr12     Bin39    \n",
       "40    CTCF       15487 TF        CTCF     \n",
       "51743 CTCF       12342 chr16     Bin338   \n",
       "51744 CTCF       12343 chr16     Bin339   \n",
       "58    EP300      15505 TF        EP300    \n",
       "64105 EP300      14330 chr22     Bin206   \n",
       "155   POLR2A     15602 TF        POLR2A   \n",
       "52860 POLR2A     12494 chr17     Bin38    \n",
       "240   YY1        15687 TF        YY1      \n",
       "46700 YY1        11461 chr14     Bin502   \n",
       "167   RAD21      15614 TF        RAD21    \n",
       "30285 RAD21       7542 chr8      Bin585   \n",
       "194   SMC3       15641 TF        SMC3     \n",
       "35620 SMC3        8928 chr10     Bin553   \n",
       "35621 SMC3        8929 chr10     Bin554   \n",
       "205   STAG1      15652 TF        STAG1    \n",
       "13536 STAG1       3137 chr3      Bin682   \n",
       "13537 STAG1       3138 chr3      Bin683   \n",
       "13539 STAG1       3139 chr3      Bin684   \n",
       "125   MED1       15572 TF        MED1     \n",
       "54100 MED1       12654 chr17     Bin198   \n",
       "126   MED12      15573 TF        MED12    \n",
       "65605 MED12      14735 chrX      Bin356   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#------------------------------------ \n",
    "geneIndex = geneMatchIndex(gene)\n",
    "TargetNodes <- geneIndex$TargetNodes\n",
    "TargetNodes "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='9.5'> Step5: To get modules which includes target genes</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At here, we use the query command at part`8.2` for modules nodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    }
   ],
   "source": [
    "#---------------------------\n",
    "#找到包含了这些基因节点的模块\n",
    "query = queryLouvainForChr12_TargetNodesInModules(TargetNodes$NodeID)\n",
    "TargetNodesInModules <- cypher(graph, query, TargetNodes_NodeID = TargetNodes$NodeID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4"
      ],
      "text/latex": [
       "4"
      ],
      "text/markdown": [
       "4"
      ],
      "text/plain": [
       "[1] 4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(TargetNodesInModules)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`TargetNodesInModules` has 4 modules."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='9.6'> Step6: Build a function to judge which modules is we want.</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`real_TargetNodes_In_Modules` function use to chose those modules includes nodes/genes as we designs(ask).\n",
    "\n",
    "argument `TargetNodesInModules` is like the last result.\n",
    "\n",
    "argument `gene` is a vector which includes all genes we want.\n",
    "\n",
    "argument `MustIn` is a vertor which includes genes we ask must in this modules.\n",
    "\n",
    "Argument `OrIn` is a vertor which includes genes some in this modules, and some may not in, but always at least one genes in. \n",
    "\n",
    "`OrIn2` is the same of `OrIn`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    }
   ],
   "source": [
    "#-------------------------------------------------------------------------------------------------------\n",
    "#real_TargetNodes_In_Modules function use to chose those modules includes nodes/genes as we designs(ask).\n",
    "#这个real_TargetNodes_In_Modules函数是用来选择那些符合要求的基因模块列表\n",
    "#argument TargetNodesInModules is the last result\n",
    "#参数有TargetNodesInModules就是我们上一步得出的结果TargetNodesInModules，结构必须是相同的\n",
    "#gene is a vector which includes all genes we want.\n",
    "#参数gene也是之前使用的\n",
    "# MustIn is a vertor which includes genes we ask must in this modules\n",
    "#参数MustIn就是我们选择的gene一定要出现在模块中\n",
    "#OrIn is a vertor which includes genes some in this modules, and some may not in, but always at least one genes in \n",
    "#参数OrIn就是我们选择的gene中至少有一个要出现在模块中\n",
    "#OrIn2 is the sames of OrIn\n",
    "#参数OrIn2也就是我们选择的gene中至少有一个要出现在模块中\n",
    "real_TargetNodes_In_Modules <- function(TargetNodesInModules,gene,MustIn,OrIn = NULL,OrIn2 = NULL){\n",
    "    index = c()\n",
    "    diffGene = list()\n",
    "    for (i in 1:nrow(TargetNodesInModules)){\n",
    "        #At first, we need to match those genes names about nodes in modules\n",
    "        #找对节点id号对应的基因名称\n",
    "        Mod_TargetNodes = unlist(TargetNodesInModules$TargetNodes_ID[i])\n",
    "        Mod_TargetNodes = geneMatchIndex(Mod_TargetNodes, by = \"NodeID\")\n",
    "        \n",
    "        #find out genes those are not in this module\n",
    "        #找到在对应的模块里面不存在的输入的基因\n",
    "        diff = setdiff(gene, Mod_TargetNodes$TargetNodes$geneName)\n",
    "        \n",
    "        #To judge whather module is we want \n",
    "        #判断diff是属于哪些基因的，是属于必需的，还是属于可选的，还是不要的\n",
    "          #-----------------------\n",
    "          #get an intersection between diff and MustIn, if geneIn was not null, then it judge that this module not as we want \n",
    "          #diff和MustIn的交集geneIn，如果有，则那这个模块不能用\n",
    "          geneIn = intersect(diff,MustIn)\n",
    "          #即如果geneIn的数量大于0， 则说明diff中有基因是跟MustIn相同的,则那这个模块不能用\n",
    "          case1 <- length(geneIn) > 0 \n",
    "          #-----------------------\n",
    "           #get an intersection between diff and OrIn, if geneOrIn have the same number of OrIn, we can not chose the modules\n",
    "          if (length(OrIn) > 1 ) {\n",
    "            #如果diff和OrIn的交集geneOrIn跟OrIn一样多，则说明OrIn中的基因都是diff，那这个模块不能用\n",
    "            geneOrIn = intersect(OrIn,diff)\n",
    "            #如果diff和OrIn的交集geneOrIn跟OrIn一样多，则说明OrIn中的基因都是diff，那这个模块不能用\n",
    "            case2 <- length(geneOrIn) == length(OrIn) }\n",
    "          else{ case2 = FALSE }\n",
    "          #-----------------------\n",
    "          if (length(OrIn2) > 1 ){\n",
    "          geneOrIn2 = intersect(diff,OrIn2)\n",
    "          case3 <- length(geneOrIn2) == length(OrIn2)}\n",
    "          else{ case2 = FALSE }\n",
    "        \n",
    "        #If one of three case is TRUE, the module will not be chose.\n",
    "          #有一个是TRUE则这个模块就不可以用了，这里是case1或者case2或者case3\n",
    "          if( case1 | case2 | case3) { \n",
    "              index = c(index, i)\n",
    "          diffGene[i] = diff}\n",
    "    }\n",
    "    names(diffGene) = TargetNodesInModules$community\n",
    "    #符合所选情况的模块（即根据自己定义了的基因都在对应的模块中）\n",
    "    realTargetNodesInModules = TargetNodesInModules[c(-index),]\n",
    "    \n",
    "    judgeResults = list(realTargetNodesInModules,diffGene)\n",
    "    names(judgeResults) = c(\"realTargetNodesInModules\",\"diffGene\")\n",
    "    return(judgeResults)\n",
    "}\n",
    "#---------------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='9.7'> Step7: Select those mudules as we want</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windowsWarning message in diffGene[i] <- diff:\n",
      "\"number of items to replace is not a multiple of replacement length\"Warning message in diffGene[i] <- diff:\n",
      "\"number of items to replace is not a multiple of replacement length\"Warning message in diffGene[i] <- diff:\n",
      "\"number of items to replace is not a multiple of replacement length\"Warning message in diffGene[i] <- diff:\n",
      "\"number of items to replace is not a multiple of replacement length\""
     ]
    }
   ],
   "source": [
    "#查看是否都包含这些基因\n",
    "\n",
    "#现将gene拆分成哪些基因是必需的，哪些是可选的，哪些是不要的\n",
    "#一定要有的基因\n",
    "#Genes must is in modules\n",
    "MustIn = c(\"NANOG\",\"CTCF\", \"EP300\",\"POLR2A\",\"YY1\")\n",
    "\n",
    "#Genes may not in modules, but at least one gene in modules\n",
    "#不一定要有的基因，即可以有，也可以没有，至少有一个在\n",
    "OrIn = c(\"RAD21\",\"SMC3\",\"STAG1\")\n",
    "OrIn2 = c(\"MED1\",\"MED12\")\n",
    "\n",
    "#Select module like we want\n",
    "judgeResults = real_TargetNodes_In_Modules(TargetNodesInModules,\n",
    "                                                       gene,MustIn,OrIn,OrIn2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>0</li>\n",
       "\t<li>4</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0\n",
       "\\item 4\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0\n",
       "2. 4\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0 4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$Module11195</dt>\n",
       "\t\t<dd>'POLR2A'</dd>\n",
       "\t<dt>$Module14229</dt>\n",
       "\t\t<dd>'NANOG'</dd>\n",
       "\t<dt>$Module14779</dt>\n",
       "\t\t<dd>'NANOG'</dd>\n",
       "\t<dt>$Module40033</dt>\n",
       "\t\t<dd>'NANOG'</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$Module11195] 'POLR2A'\n",
       "\\item[\\$Module14229] 'NANOG'\n",
       "\\item[\\$Module14779] 'NANOG'\n",
       "\\item[\\$Module40033] 'NANOG'\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$Module11195\n",
       ":   'POLR2A'\n",
       "$Module14229\n",
       ":   'NANOG'\n",
       "$Module14779\n",
       ":   'NANOG'\n",
       "$Module40033\n",
       ":   'NANOG'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$Module11195\n",
       "[1] \"POLR2A\"\n",
       "\n",
       "$Module14229\n",
       "[1] \"NANOG\"\n",
       "\n",
       "$Module14779\n",
       "[1] \"NANOG\"\n",
       "\n",
       "$Module40033\n",
       "[1] \"NANOG\"\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "realTargetNodesInModules = judgeResults[[1]]\n",
    "\n",
    "diffGene = judgeResults[[2]]\n",
    "\n",
    "dim(realTargetNodesInModules)\n",
    "\n",
    "diffGene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, we didn't get an module as we want. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 id='9.8'> Step8: get relationships during nodes in each community(module)</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### If there have modules as we want, we can get relationships by function `Module_Relationship`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    }
   ],
   "source": [
    "#假设有的话，我们继续将这个模块的所有关系都找出来\n",
    "#realTargetNodesInModules = TargetNodesInModules\n",
    "#Relationship_realTargetNodesInModules <- Module_Relationship(realTargetNodesInModules)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "#head(Relationship_realTargetNodesInModules[[1]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 id='10'> save all results<h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get all results at [here](allResults.Rdata)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "save(labels,  relationshipType,  rel_details,  BindRel,BindRel_TF,  BindRel_LncRNA,InclusionRel, \n",
    "     InteractionRel,  chr12_hubs,  distribution,  module_chr12_total,  module_chr12,module_genelist,\n",
    "     module_chr12_list,  module_chr12_list_enrichGO,  module_chr12_list_enrichKEGG,AllgeneName, \n",
    "     chr12Mod,   Module_Relationships,   AllgeneName_TF_LncRNA,   TargetNodesInModules, \n",
    "      realTargetNodesInModules, diffGene,\n",
    "     file = \"allResults.Rdata\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "206.4px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
